


  
    
  




<head>
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/all.min.css" integrity="sha512-iBBXm8fW90+nuLcSKlbmrPcLa0OT92xO1BIsZ+ywDWZCvqsWgccV3gFoRBv0z+8dLJgyAHIhR35VZc2oM/gI1w==" crossorigin="anonymous" referrerpolicy="no-referrer" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.9.1/css/academicons.min.css" integrity="sha512-b1ASx0WHgVFL5ZQhTgiPWX+68KjS38Jk87jg7pe+qC7q9YkEtFq0z7xCglv7qGIs/68d3mAp+StfC8WKC5SSAg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
  <link rel="shortcut icon" href="../../_static/logo/mda_favicon.ico">
</head>

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>11.3.3. Topology attribute objects — MDAnalysis.core.topologyattrs &mdash; MDAnalysis 2.8.0-dev0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/site.css" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=a9b57af3"></script>
        <script src="../../_static/doctools.js?v=888ff710"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script src="../../_static/js/versions.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="search" type="application/opensearchdescription+xml"
          title="Search within MDAnalysis 2.8.0-dev0 documentation"
          href="../../_static/opensearch.xml"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="11.4.1. Atom selection Hierarchy — MDAnalysis.core.selection" href="selection.html" />
    <link rel="prev" title="11.3.2. Core Topology Objects — MDAnalysis.core.topologyobjects" href="topologyobjects.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >




  




<a href="../../index.html">
  
    <img src="../../_static/logo/mda_logo.png" class="logo" alt="Logo"/>
</a>


  
  
  
    <div class="version">
      2.8.0-dev0
    </div>
  



<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

        </div>
<div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
    <!-- <p class="caption" role="heading"></p> -->
    <ul>
        
        <li class="toctree-l1"><a class="reference internal" href="http://mdanalysis.org">MDAnalysis</a></li>
        
        <li class="toctree-l1"><a class="reference internal" href="http://userguide.mdanalysis.org">User guide</a></li>
        
        <li class="toctree-l1"><a class="reference internal" href="https://mdakits.mdanalysis.org/">MDAKits</a></li>
        
    </ul>
    
        <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../overview.html">1. Overview over MDAnalysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../topology.html">2. The topology system</a></li>
<li class="toctree-l1"><a class="reference internal" href="../selections.html">3. Selection commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../analysis_modules.html">4. Analysis modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../topology_modules.html">5. Topology modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../coordinates_modules.html">6. Coordinates modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../converters.html">7. Converter modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectory_transformations.html">8. Trajectory transformations (“on-the-fly” transformations)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../selections_modules.html">9. Selection exporters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../auxiliary_modules.html">10. Auxiliary modules</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../core_modules.html">11. Core modules</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="init.html">11.1. Core functions of MDAnalysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../core_modules.html#important-objects-for-users">11.2. Important objects for users</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../core_modules.html#topology-system">11.3. Topology system</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="topology.html">11.3.1. Core Topology object — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.core.topology</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="topologyobjects.html">11.3.2. Core Topology Objects — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.core.topologyobjects</span></code></a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">11.3.3. Topology attribute objects — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.core.topologyattrs</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../core_modules.html#selection-system">11.4. Selection system</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../visualization_modules.html">12. Visualization modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../lib_modules.html">13. Library functions — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.lib</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../version.html">14. Version information for MDAnalysis - <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.version</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../units.html">15. Constants and unit conversion — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.units</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../exceptions.html">16. Custom exceptions and warnings — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.exceptions</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">17. References</a></li>
</ul>

</div>

      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">MDAnalysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../core_modules.html"><span class="section-number">11. </span>Core modules</a></li>
      <li class="breadcrumb-item active"><span class="section-number">11.3.3. </span>Topology attribute objects — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.core.topologyattrs</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/documentation_pages/core/topologyattrs.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="topology-attribute-objects-mdanalysis-core-topologyattrs">
<span id="module-MDAnalysis.core.topologyattrs"></span><h1><span class="section-number">11.3.3. </span>Topology attribute objects — <a class="reference internal" href="#module-MDAnalysis.core.topologyattrs" title="MDAnalysis.core.topologyattrs"><code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.core.topologyattrs</span></code></a><a class="headerlink" href="#topology-attribute-objects-mdanalysis-core-topologyattrs" title="Link to this heading"></a></h1>
<p>Common <a class="reference internal" href="#MDAnalysis.core.topologyattrs.TopologyAttr" title="MDAnalysis.core.topologyattrs.TopologyAttr"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyAttr</span></code></a> instances that are used by most topology
parsers.</p>
<p>TopologyAttrs are used to contain attributes such as atom names or resids.
These are usually read by the TopologyParser.</p>
<p class="rubric">References</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AltLocs">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">AltLocs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AltLocs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AltLocs" title="Link to this definition"></a></dt>
<dd><p>AltLocs for each atom</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AltLocs.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AltLocs.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Angles">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Angles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Angles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Angles" title="Link to this definition"></a></dt>
<dd><p>Angles between three atoms</p>
<p>Initialise with a list of 3 long tuples
E.g.,  <cite>[(0, 1, 2), (1, 2, 3), (2, 3, 4)]</cite></p>
<p>These indices refer to the atom indices.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Aromaticities">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Aromaticities</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Aromaticities"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Aromaticities" title="Link to this definition"></a></dt>
<dd><p>Aromaticity</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Aromaticities.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Aromaticities.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AtomAttr">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">AtomAttr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AtomAttr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AtomAttr" title="Link to this definition"></a></dt>
<dd><p>Base class for atom attributes.</p>
<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AtomAttr.get_atoms">
<span class="sig-name descname"><span class="pre">get_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AtomAttr.get_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AtomAttr.get_atoms" title="Link to this definition"></a></dt>
<dd><p>Get atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AtomAttr.get_residues">
<span class="sig-name descname"><span class="pre">get_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AtomAttr.get_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AtomAttr.get_residues" title="Link to this definition"></a></dt>
<dd><p>By default, the values for each atom present in the set of residues
are returned in a single array. This behavior can be overriden in child
attributes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AtomAttr.get_segments">
<span class="sig-name descname"><span class="pre">get_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AtomAttr.get_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AtomAttr.get_segments" title="Link to this definition"></a></dt>
<dd><p>By default, the values for each atom present in the set of residues
are returned in a single array. This behavior can be overriden in child
attributes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AtomAttr.set_atoms">
<span class="sig-name descname"><span class="pre">set_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AtomAttr.set_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AtomAttr.set_atoms" title="Link to this definition"></a></dt>
<dd><p>Set atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AtomAttr.set_residues">
<span class="sig-name descname"><span class="pre">set_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AtomAttr.set_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AtomAttr.set_residues" title="Link to this definition"></a></dt>
<dd><p>Set residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AtomAttr.set_segments">
<span class="sig-name descname"><span class="pre">set_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AtomAttr.set_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AtomAttr.set_segments" title="Link to this definition"></a></dt>
<dd><p>Set segmentattributes for a given SegmentGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AtomStringAttr">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">AtomStringAttr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AtomStringAttr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AtomStringAttr" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.AtomStringAttr.set_atoms">
<span class="sig-name descname"><span class="pre">set_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#AtomStringAttr.set_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.AtomStringAttr.set_atoms" title="Link to this definition"></a></dt>
<dd><p>Set atom attributes for a given AtomGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomids">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Atomids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomids" title="Link to this definition"></a></dt>
<dd><p>ID for each atom.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomids.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomids.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomindices">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Atomindices</span></span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomindices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomindices" title="Link to this definition"></a></dt>
<dd><p>Globally unique indices for each atom in the group.</p>
<p>If the group is an AtomGroup, then this gives the index for each atom in
the group. This is the unambiguous identifier for each atom in the
topology, and it is not alterable.</p>
<p>If the group is a ResidueGroup or SegmentGroup, then this gives the indices
of each atom represented in the group in a 1-D array, in the order of the
elements in that group.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomindices.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomindices.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomindices.get_atoms">
<span class="sig-name descname"><span class="pre">get_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomindices.get_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomindices.get_atoms" title="Link to this definition"></a></dt>
<dd><p>Get atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomindices.get_residues">
<span class="sig-name descname"><span class="pre">get_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomindices.get_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomindices.get_residues" title="Link to this definition"></a></dt>
<dd><p>Get residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomindices.get_segments">
<span class="sig-name descname"><span class="pre">get_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomindices.get_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomindices.get_segments" title="Link to this definition"></a></dt>
<dd><p>Get segment attributes for a given SegmentGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomindices.set_atoms">
<span class="sig-name descname"><span class="pre">set_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomindices.set_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomindices.set_atoms" title="Link to this definition"></a></dt>
<dd><p>Set atom attributes for a given AtomGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Atomnames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomnames"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames" title="Link to this definition"></a></dt>
<dd><p>Name for each atom.</p>
<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames.chi1_selection">
<span class="sig-name descname"><span class="pre">chi1_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cg_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CG</span> <span class="pre">CG1</span> <span class="pre">OG</span> <span class="pre">OG1</span> <span class="pre">SG'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomnames.chi1_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames.chi1_selection" title="Link to this definition"></a></dt>
<dd><p>Select AtomGroup corresponding to the chi1 sidechain dihedral <code class="docutils literal notranslate"><span class="pre">N-CA-CB-*G.</span></code>
The gamma atom is taken to be the heavy atom in the gamma position. If more than one
heavy atom is present (e.g. CG1 and CG2), the one with the lower number is used (CG1).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This numbering of chi1 atoms here in following with the IUPAC 1970 rules.
However, it should be noted that analyses which use dihedral angles may have
different definitions. For example, the
<a class="reference internal" href="../analysis/dihedrals.html#MDAnalysis.analysis.dihedrals.Janin" title="MDAnalysis.analysis.dihedrals.Janin"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.analysis.dihedrals.Janin</span></code></a> class does not incorporate
amino acids where the gamma atom is not carbon, into its chi1 selections.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
<li><p><strong>cb_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the beta-carbon atom</p></li>
<li><p><strong>cg_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the gamma-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>4-atom selection in the correct order. If no CB and/or CG is found
then this method returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup">AtomGroup</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.5.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Added arguments for flexible atom names and refactored code for
faster atom matching with boolean arrays.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames.chi1_selections">
<span class="sig-name descname"><span class="pre">chi1_selections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cg_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CG</span> <span class="pre">CG1</span> <span class="pre">OG</span> <span class="pre">OG1</span> <span class="pre">SG'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomnames.chi1_selections"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames.chi1_selections" title="Link to this definition"></a></dt>
<dd><p>Select list of AtomGroups corresponding to the chi1 sidechain dihedral
N-CA-CB-CG.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
<li><p><strong>cb_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the beta-carbon atom</p></li>
<li><p><strong>cg_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the gamma-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>List of AtomGroups</em> – 4-atom selections in the correct order. If no CB and/or CG is found
then the corresponding item in the list is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionadded:: 1.0.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames.omega_selection">
<span class="sig-name descname"><span class="pre">omega_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomnames.omega_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames.omega_selection" title="Link to this definition"></a></dt>
<dd><p>Select AtomGroup corresponding to the omega protein backbone dihedral
CA-C-N’-CA’.</p>
<p>omega describes the -C-N- peptide bond. Typically, it is trans (180
degrees) although cis-bonds (0 degrees) are also occasionally observed
(especially near Proline).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>AtomGroup</em> – 4-atom selection in the correct order. If no C’ found in the
previous residue (by resid) then this method returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionchanged:: 1.0.0</em> – Added arguments for flexible atom names and refactored code for
faster atom matching with boolean arrays.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames.omega_selections">
<span class="sig-name descname"><span class="pre">omega_selections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomnames.omega_selections"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames.omega_selections" title="Link to this definition"></a></dt>
<dd><p>Select list of AtomGroups corresponding to the omega protein
backbone dihedral CA-C-N’-CA’.</p>
<p>omega describes the -C-N- peptide bond. Typically, it is trans (180
degrees) although cis-bonds (0 degrees) are also occasionally observed
(especially near Proline).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>List of AtomGroups</em> – 4-atom selections in the correct order. If no C’ found in the
previous residue (by resid) then the corresponding item in the
list is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionadded:: 1.0.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames.phi_selection">
<span class="sig-name descname"><span class="pre">phi_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomnames.phi_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames.phi_selection" title="Link to this definition"></a></dt>
<dd><p>Select AtomGroup corresponding to the phi protein backbone dihedral
C’-N-CA-C.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>AtomGroup</em> – 4-atom selection in the correct order. If no C’ found in the
previous residue (by resid) then this method returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionchanged:: 1.0.0</em> – Added arguments for flexible atom names and refactored code for
faster atom matching with boolean arrays.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames.phi_selections">
<span class="sig-name descname"><span class="pre">phi_selections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomnames.phi_selections"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames.phi_selections" title="Link to this definition"></a></dt>
<dd><p>Select list of AtomGroups corresponding to the phi protein
backbone dihedral C’-N-CA-C.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>list of AtomGroups</em> – 4-atom selections in the correct order. If no C’ found in the
previous residue (by resid) then corresponding item in the list
is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionadded:: 1.0.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames.psi_selection">
<span class="sig-name descname"><span class="pre">psi_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomnames.psi_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames.psi_selection" title="Link to this definition"></a></dt>
<dd><p>Select AtomGroup corresponding to the psi protein backbone dihedral
N-CA-C-N’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>AtomGroup</em> – 4-atom selection in the correct order. If no N’ found in the
following residue (by resid) then this method returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionchanged:: 1.0.0</em> – Added arguments for flexible atom names and refactored code for
faster atom matching with boolean arrays.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomnames.psi_selections">
<span class="sig-name descname"><span class="pre">psi_selections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomnames.psi_selections"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomnames.psi_selections" title="Link to this definition"></a></dt>
<dd><p>Select list of AtomGroups corresponding to the psi protein
backbone dihedral N-CA-C-N’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>List of AtomGroups</em> – 4-atom selections in the correct order. If no N’ found in the
following residue (by resid) then the corresponding item in the
list is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionadded:: 1.0.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomtypes">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Atomtypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Atomtypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomtypes" title="Link to this definition"></a></dt>
<dd><p>Type for each atom</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Atomtypes.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Atomtypes.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Bonds">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Bonds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Bonds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Bonds" title="Link to this definition"></a></dt>
<dd><p>Bonds between two atoms</p>
<p>Must be initialised by a list of zero based tuples.
These indices refer to the atom indices.
E.g., ` [(0, 1), (1, 2), (2, 3)]`</p>
<p>Also adds the <cite>bonded_atoms</cite>, <cite>fragment</cite> and <cite>fragments</cite>
attributes.</p>
<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Bonds.bonded_atoms">
<span class="sig-name descname"><span class="pre">bonded_atoms</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Bonds.bonded_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Bonds.bonded_atoms" title="Link to this definition"></a></dt>
<dd><p>An <a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> of all
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> bonded to this
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Bonds.fragindex">
<span class="sig-name descname"><span class="pre">fragindex</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Bonds.fragindex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Bonds.fragindex" title="Link to this definition"></a></dt>
<dd><p>The index (ID) of the
<a class="reference internal" href="#MDAnalysis.core.topologyattrs.Bonds.fragment" title="MDAnalysis.core.topologyattrs.Bonds.fragment"><code class="xref py py-class docutils literal notranslate"><span class="pre">fragment</span></code></a> this
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> is part of.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Bonds.fragindices">
<span class="sig-name descname"><span class="pre">fragindices</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Bonds.fragindices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Bonds.fragindices" title="Link to this definition"></a></dt>
<dd><p>The
<a class="reference internal" href="#MDAnalysis.core.topologyattrs.Bonds.fragindex" title="MDAnalysis.core.topologyattrs.Bonds.fragindex"><code class="xref py py-class docutils literal notranslate"><span class="pre">fragment</span> <span class="pre">indices</span></code></a>
of all <a class="reference internal" href="groups.html#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in this
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p>A <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> with
<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.shape.html#numpy.ndarray.shape" title="(in NumPy v1.26)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">shape</span></code></a><code class="docutils literal notranslate"><span class="pre">=(</span></code><code class="xref py py-attr docutils literal notranslate"><span class="pre">n_atoms</span></code><code class="docutils literal notranslate"><span class="pre">,)</span></code>
and <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.dtype.html#numpy.ndarray.dtype" title="(in NumPy v1.26)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dtype</span></code></a><code class="docutils literal notranslate"><span class="pre">=numpy.int64</span></code>.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Bonds.fragment">
<span class="sig-name descname"><span class="pre">fragment</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Bonds.fragment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Bonds.fragment" title="Link to this definition"></a></dt>
<dd><p>An <a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> representing the
fragment this <a class="reference internal" href="groups.html#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> is part of.</p>
<p>A fragment is a
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">group</span> <span class="pre">of</span> <span class="pre">atoms</span></code></a> which are
interconnected by <a class="reference internal" href="#MDAnalysis.core.topologyattrs.Bonds" title="MDAnalysis.core.topologyattrs.Bonds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bonds</span></code></a>, i.e.,
there exists a path along one
or more <a class="reference internal" href="#MDAnalysis.core.topologyattrs.Bonds" title="MDAnalysis.core.topologyattrs.Bonds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bonds</span></code></a> between any pair
of <a class="reference internal" href="groups.html#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
within a fragment. Thus, a fragment typically corresponds to a molecule.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Bonds.fragments">
<span class="sig-name descname"><span class="pre">fragments</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Bonds.fragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Bonds.fragments" title="Link to this definition"></a></dt>
<dd><p>Read-only <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a> of
<a class="reference internal" href="#MDAnalysis.core.topologyattrs.Bonds.fragment" title="MDAnalysis.core.topologyattrs.Bonds.fragment"><code class="xref py py-class docutils literal notranslate"><span class="pre">fragments</span></code></a>.</p>
<p>Contains all fragments that
any <a class="reference internal" href="groups.html#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> in this
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is part of.</p>
<p>A fragment is a
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">group</span> <span class="pre">of</span> <span class="pre">atoms</span></code></a> which are
interconnected by <a class="reference internal" href="#MDAnalysis.core.topologyattrs.Bonds" title="MDAnalysis.core.topologyattrs.Bonds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bonds</span></code></a>, i.e.,
there exists a path along one
or more <a class="reference internal" href="#MDAnalysis.core.topologyattrs.Bonds" title="MDAnalysis.core.topologyattrs.Bonds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bonds</span></code></a> between any pair
of <a class="reference internal" href="groups.html#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
within a fragment. Thus, a fragment typically corresponds to a molecule.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>The contents of the fragments may extend beyond the contents of this
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p></li>
</ul>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Bonds.n_fragments">
<span class="sig-name descname"><span class="pre">n_fragments</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Bonds.n_fragments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Bonds.n_fragments" title="Link to this definition"></a></dt>
<dd><p>The number of unique
<a class="reference internal" href="#MDAnalysis.core.topologyattrs.Bonds.fragments" title="MDAnalysis.core.topologyattrs.Bonds.fragments"><code class="xref py py-class docutils literal notranslate"><span class="pre">fragments</span></code></a> the
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> of this
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> are part of.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.CMaps">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">CMaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#CMaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.CMaps" title="Link to this definition"></a></dt>
<dd><p>A connection between five atoms
.. versionadded:: 1.0.0</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ChainIDs">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">ChainIDs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ChainIDs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ChainIDs" title="Link to this definition"></a></dt>
<dd><p>ChainID per atom</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is an attribute of the Atom, not Residue or Segment</p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ChainIDs.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ChainIDs.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Charges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Charges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges.center_of_charge">
<span class="sig-name descname"><span class="pre">center_of_charge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Charges.center_of_charge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges.center_of_charge" title="Link to this definition"></a></dt>
<dd><p>Center of (absolute) charge of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \vert q_i \vert \boldsymbol r_i}
                     {\sum_i \vert q_i \vert}\]</div>
<p>where <span class="math notranslate nohighlight">\(q_i\)</span> is the charge and <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> the
position of atom <span class="math notranslate nohighlight">\(i\)</span> in the given
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup</span></code></a>.
Centers of charge per <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the charges of a compound sum up to zero, the
center of mass coordinates of that compound will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of mass of all atoms in the group will
be returned as a single position vector. Otherwise, the centers of
mass of each <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the center(s) of charge of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where
<code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method can only be accessed if the underlying topology has
information about atomic charges.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.2.0.</span></p>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges.dipole_moment">
<span class="sig-name descname"><span class="pre">dipole_moment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Charges.dipole_moment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges.dipole_moment" title="Link to this definition"></a></dt>
<dd><p>Dipole moment of the group or compounds in a group.</p>
<div class="math notranslate nohighlight">
\[\mu = |\boldsymbol{\mu}| = \sqrt{ \sum_{i=1}^{D} \mu^2 }\]</div>
<p>Where <span class="math notranslate nohighlight">\(D\)</span> is the number of dimensions, in this case 3.</p>
<p>Computes the dipole moment of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> in the group.
Dipole per <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is a net charge, the magnitude of the dipole
moment is dependent on the <cite>center</cite> chosen.
See <a class="reference internal" href="#MDAnalysis.core.topologyattrs.Charges.dipole_vector" title="MDAnalysis.core.topologyattrs.Charges.dipole_vector"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dipole_vector()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single dipole vector returns. Otherwise, an
array of each <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dipole moment(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges.dipole_vector">
<span class="sig-name descname"><span class="pre">dipole_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mass'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Charges.dipole_vector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges.dipole_vector" title="Link to this definition"></a></dt>
<dd><p>Dipole vector of the group.</p>
<div class="math notranslate nohighlight">
\[\boldsymbol{\mu} = \sum_{i=1}^{N} q_{i} ( \mathbf{r}_{i} -
\mathbf{r}_{COM} )\]</div>
<p>Computes the dipole vector of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> in the group.
Dipole vector per <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that the magnitude of the dipole moment is independent of the
<code class="docutils literal notranslate"><span class="pre">center</span></code> chosen unless the species has a net charge. In the case of
a charged group the dipole moment can be later adjusted  with:</p>
<div class="math notranslate nohighlight">
\[\boldsymbol{\mu}_{COC} = \boldsymbol{\mu}_{COM} +
q_{ag}\mathbf{r}_{COM} - q_{ag}\boldsymbol{r}_{COC}\]</div>
<p>Where <span class="math notranslate nohighlight">\(\mathbf{r}_{COM}\)</span> is the center of mass and
<span class="math notranslate nohighlight">\(\mathbf{r}_{COC}\)</span> is the center of charge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single dipole vector returns. Otherwise, an
array of each <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dipole vector(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,3)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges.get_residues">
<span class="sig-name descname"><span class="pre">get_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Charges.get_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges.get_residues" title="Link to this definition"></a></dt>
<dd><p>By default, the values for each atom present in the set of residues
are returned in a single array. This behavior can be overriden in child
attributes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges.get_segments">
<span class="sig-name descname"><span class="pre">get_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Charges.get_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges.get_segments" title="Link to this definition"></a></dt>
<dd><p>By default, the values for each atom present in the set of residues
are returned in a single array. This behavior can be overriden in child
attributes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges.quadrupole_moment">
<span class="sig-name descname"><span class="pre">quadrupole_moment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Charges.quadrupole_moment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges.quadrupole_moment" title="Link to this definition"></a></dt>
<dd><p>Quadrupole moment of the group according to <a class="footnote-reference brackets" href="#footcite-gray1984" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p>
<div class="math notranslate nohighlight">
\[Q = \sqrt{\frac{2}{3}{\hat{\mathsf{Q}}}:{\hat{\mathsf{Q}}}}\]</div>
<p>where the quadrupole moment is calculated from the tensor double
contraction of the traceless quadropole tensor <span class="math notranslate nohighlight">\(\hat{\mathsf{Q}}\)</span></p>
<p>Computes the quadrupole moment of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> in the group.
Quadrupole per <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is an unsymmetrical plane in the molecule or
group, the magnitude of the quadrupole moment is dependant on the
<code class="docutils literal notranslate"><span class="pre">center</span></code> chosen and cannot be translated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single quadrupole value returns. Otherwise, an
array of each <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 1d array.
Note that, in any case, <em>only</em> the positions of
<code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Quadrupole moment(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges.quadrupole_tensor">
<span class="sig-name descname"><span class="pre">quadrupole_tensor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mass'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Charges.quadrupole_tensor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges.quadrupole_tensor" title="Link to this definition"></a></dt>
<dd><p>Traceless quadrupole tensor of the group or compounds.</p>
<p>This tensor is first computed as the outer product of vectors from
a reference point to some atom, and multiplied by the atomic charge.
The tensor of each atom is then summed to produce the quadrupole
tensor of the group:</p>
<div class="math notranslate nohighlight">
\[\mathsf{Q} = \sum_{i=1}^{N} q_{i} ( \mathbf{r}_{i} -
\mathbf{r}_{COM} ) \otimes ( \mathbf{r}_{i} - \mathbf{r}_{COM} )\]</div>
<p>The traceless quadrupole tensor, <span class="math notranslate nohighlight">\(\hat{\mathsf{Q}}\)</span>, is then
taken from:</p>
<div class="math notranslate nohighlight">
\[\hat{\mathsf{Q}} = \frac{3}{2} \mathsf{Q} - \frac{1}{2}
tr(\mathsf{Q})\]</div>
<p>Computes the quadrupole tensor of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> in the group.
Tensor per <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is an unsymmetrical plane in the molecule or
group, the magnitude of the quadrupole tensor is dependent on the
<code class="docutils literal notranslate"><span class="pre">center</span></code> (e.g., <span class="math notranslate nohighlight">\(\mathbf{r}_{COM}\)</span>) chosen and cannot be translated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single quadrupole value returns. Otherwise, an
array of each <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 1d array.
Note that, in any case, <em>only</em> the positions of
<code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Quadrupole tensor(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
tensor of shape <code class="docutils literal notranslate"><span class="pre">(3,3)</span></code>. Otherwise, the output will be a 1d array
of shape <code class="docutils literal notranslate"><span class="pre">(n,3,3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Charges.total_charge">
<span class="sig-name descname"><span class="pre">total_charge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Charges.total_charge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Charges.total_charge" title="Link to this definition"></a></dt>
<dd><p>Total charge of (compounds of) the group.</p>
<p>Computes the total charge of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> in the group.
Total charges per <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If ‘group’, the total charge of all atoms in the group will
be returned as a single value. Otherwise, the total charges per
<code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, molecule, or fragment
will be returned as a 1d array.
Note that, in any case, <em>only</em> the charges of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code>
<em>belonging to the group</em> will be taken into account.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Total charge of (compounds of) the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>compound</cite> parameter</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Dihedrals">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Dihedrals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Dihedrals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Dihedrals" title="Link to this definition"></a></dt>
<dd><p>A connection between four sequential atoms</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Elements">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Elements</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Elements"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Elements" title="Link to this definition"></a></dt>
<dd><p>Element for each atom</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Elements.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Elements.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Epsilon14s">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Epsilon14s</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Epsilon14s"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Epsilon14s" title="Link to this definition"></a></dt>
<dd><p>The epsilon LJ parameter for 1-4 interactions</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Epsilon14s.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Epsilon14s.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Epsilons">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Epsilons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Epsilons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Epsilons" title="Link to this definition"></a></dt>
<dd><p>The epsilon LJ parameter</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Epsilons.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Epsilons.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.FormalCharges">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">FormalCharges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#FormalCharges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.FormalCharges" title="Link to this definition"></a></dt>
<dd><p>Formal charge on each atom</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.FormalCharges.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.FormalCharges.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.GBScreens">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">GBScreens</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#GBScreens"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.GBScreens" title="Link to this definition"></a></dt>
<dd><p>Generalized Born screening factor</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.GBScreens.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.GBScreens.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ICodes">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">ICodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ICodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ICodes" title="Link to this definition"></a></dt>
<dd><p>Insertion code for Atoms</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ICodes.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ICodes.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Impropers">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Impropers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Impropers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Impropers" title="Link to this definition"></a></dt>
<dd><p>An imaginary dihedral between four atoms</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Masses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.align_principal_axis">
<span class="sig-name descname"><span class="pre">align_principal_axis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">axis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.align_principal_axis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.align_principal_axis" title="Link to this definition"></a></dt>
<dd><p>Align principal axis with index <cite>axis</cite> with <cite>vector</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>{0</em><em>, </em><em>1</em><em>, </em><em>2}</em>) – Index of the principal axis (0, 1, or 2), as produced by
<a class="reference internal" href="#MDAnalysis.core.topologyattrs.Masses.principal_axes" title="MDAnalysis.core.topologyattrs.Masses.principal_axes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">principal_axes()</span></code></a>.</p></li>
<li><p><strong>vector</strong> (<em>array_like</em>) – Vector to align principal axis with.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>To align the long axis of a channel (the first principal axis, i.e.
<em>axis</em> = 0) with the z-axis:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">align_principal_axis</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;aligned.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.asphericity">
<span class="sig-name descname"><span class="pre">asphericity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.asphericity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.asphericity" title="Link to this definition"></a></dt>
<dd><p>Asphericity.</p>
<p>See <a class="reference internal" href="../references.html#dima2004b" id="id2"><span>[Dima2004b]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <em>unwrap</em> and <em>compound</em> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.5.0: </span>Added calculation for any <cite>compound</cite> type</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.center_of_mass">
<span class="sig-name descname"><span class="pre">center_of_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.center_of_mass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.center_of_mass" title="Link to this definition"></a></dt>
<dd><p>Center of mass of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i m_i \boldsymbol r_i}{\sum m_i}\]</div>
<p>where <span class="math notranslate nohighlight">\(m_i\)</span> is the mass and <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> the
position of atom <span class="math notranslate nohighlight">\(i\)</span> in the given
<a class="reference internal" href="groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup</span></code></a>.
Centers of mass per <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the masses of a compound sum up to zero, the
center of mass coordinates of that compound will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of mass of all atoms in the group will
be returned as a single position vector. Otherwise, the centers of
mass of each <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, molecule, or
fragment will be returned as an array of position vectors, i.e. a 2d
array.
Note that, in any case, <em>only</em> the positions of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the center(s) of mass of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code>
is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method can only be accessed if the underlying topology has
information about atomic masses.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> compounds;
added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.get_residues">
<span class="sig-name descname"><span class="pre">get_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.get_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.get_residues" title="Link to this definition"></a></dt>
<dd><p>By default, the values for each atom present in the set of residues
are returned in a single array. This behavior can be overriden in child
attributes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.get_segments">
<span class="sig-name descname"><span class="pre">get_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.get_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.get_segments" title="Link to this definition"></a></dt>
<dd><p>By default, the values for each atom present in the set of residues
are returned in a single array. This behavior can be overriden in child
attributes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.gyration_moments">
<span class="sig-name descname"><span class="pre">gyration_moments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.gyration_moments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.gyration_moments" title="Link to this definition"></a></dt>
<dd><p>Moments of the gyration tensor.</p>
<p>The moments are defined as the eigenvalues of the gyration
tensor.</p>
<div class="math notranslate nohighlight">
\[\mathsf{T} = \frac{1}{N} \sum_{i=1}^{N} (\mathbf{r}_\mathrm{i} -
    \mathbf{r}_\mathrm{COM})(\mathbf{r}_\mathrm{i} - \mathbf{r}_\mathrm{COM})\]</div>
<p>Where <span class="math notranslate nohighlight">\(\mathbf{r}_\mathrm{COM}\)</span> is the center of mass.</p>
<p>See <a class="reference internal" href="../references.html#dima2004a" id="id3"><span>[Dima2004a]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>principle_moments_of_gyration</strong> – Gyration vector(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(Å^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
vector of length 3. Otherwise, the output will be a 2D array of shape
<code class="docutils literal notranslate"><span class="pre">(n,3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.5.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.moment_of_inertia">
<span class="sig-name descname"><span class="pre">moment_of_inertia</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.moment_of_inertia"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.moment_of_inertia" title="Link to this definition"></a></dt>
<dd><p>Moment of inertia tensor relative to center of mass.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers and tensor of inertia.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
<cite>compound</cite> determines the behavior of <cite>wrap</cite>.
Note that, in any case, <em>only</em> the positions of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>moment_of_inertia</strong> – Moment of inertia tensor as a 3 x 3 numpy array.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The moment of inertia tensor <span class="math notranslate nohighlight">\(\mathsf{I}\)</span> is calculated for a group of
<span class="math notranslate nohighlight">\(N\)</span> atoms with coordinates <span class="math notranslate nohighlight">\(\mathbf{r}_i,\ 1 \le i \le N\)</span>
relative to its center of mass from the relative coordinates</p>
<div class="math notranslate nohighlight">
\[\mathbf{r}'_i = \mathbf{r}_i - \frac{1}{\sum_{i=1}^{N} m_i} \sum_{i=1}^{N} m_i \mathbf{r}_i\]</div>
<p>as</p>
<div class="math notranslate nohighlight">
\[\mathsf{I} = \sum_{i=1}^{N} m_i \Big[(\mathbf{r}'_i\cdot\mathbf{r}'_i) \sum_{\alpha=1}^{3}
      \hat{\mathbf{e}}_\alpha \otimes \hat{\mathbf{e}}_\alpha - \mathbf{r}'_i \otimes \mathbf{r}'_i\Big]\]</div>
<p>where <span class="math notranslate nohighlight">\(\hat{\mathbf{e}}_\alpha\)</span> are Cartesian unit vectors, or in Cartesian coordinates</p>
<div class="math notranslate nohighlight">
\[I_{\alpha,\beta} = \sum_{k=1}^{N} m_k
      \Big(\big(\sum_{\gamma=1}^3 (x'^{(k)}_{\gamma})^2 \big)\delta_{\alpha,\beta}
      - x'^{(k)}_{\alpha} x'^{(k)}_{\beta} \Big).\]</div>
<p>where <span class="math notranslate nohighlight">\(x'^{(k)}_{\alpha}\)</span> are the Cartesian coordinates of the
relative coordinates <span class="math notranslate nohighlight">\(\mathbf{r}'_k\)</span>.</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.principal_axes">
<span class="sig-name descname"><span class="pre">principal_axes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.principal_axes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.principal_axes" title="Link to this definition"></a></dt>
<dd><p>Calculate the principal axes from the moment of inertia.</p>
<p>e1,e2,e3 = AtomGroup.principal_axes()</p>
<p>The eigenvectors are sorted by eigenvalue, i.e. the first one
corresponds to the highest eigenvalue and is thus the first principal
axes.</p>
<p>The eigenvectors form a right-handed coordinate system.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>axis_vectors</strong> – 3 x 3 array with <code class="docutils literal notranslate"><span class="pre">v[0]</span></code> as first, <code class="docutils literal notranslate"><span class="pre">v[1]</span></code> as second, and
<code class="docutils literal notranslate"><span class="pre">v[2]</span></code> as third eigenvector.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Always return principal axes in right-hand convention.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.radius_of_gyration">
<span class="sig-name descname"><span class="pre">radius_of_gyration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.radius_of_gyration"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.radius_of_gyration" title="Link to this definition"></a></dt>
<dd><p>Radius of gyration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.shape_parameter">
<span class="sig-name descname"><span class="pre">shape_parameter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.shape_parameter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.shape_parameter" title="Link to this definition"></a></dt>
<dd><p>Shape parameter.</p>
<p>See <a class="reference internal" href="../references.html#dima2004a" id="id4"><span>[Dima2004a]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.
Superfluous kwargs were removed.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.5.0: </span>Added calculation for any <cite>compound</cite> type</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Masses.total_mass">
<span class="sig-name descname"><span class="pre">total_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Masses.total_mass"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Masses.total_mass" title="Link to this definition"></a></dt>
<dd><p>Total mass of (compounds of) the group.</p>
<p>Computes the total mass of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code> in the group.
Total masses per <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the total mass of all atoms in the group will be
returned as a single value. Otherwise, the total masses per
<code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code>, molecule, or fragment will be
returned as a 1d array.
Note that, in any case, <em>only</em> the masses of <code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code>
<em>belonging to the group</em> will be taken into account.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Total mass of (compounds of) the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>compound</cite> parameter</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Molnums">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Molnums</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Molnums"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Molnums" title="Link to this definition"></a></dt>
<dd><p>Index of molecule from 0</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Molnums.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Molnums.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">int64</span></code></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Moltypes">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Moltypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Moltypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Moltypes" title="Link to this definition"></a></dt>
<dd><p>Name of the molecule type</p>
<p>Two molecules that share a molecule type share a common template topology.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Moltypes.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Moltypes.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.NonbondedIndices">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">NonbondedIndices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#NonbondedIndices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.NonbondedIndices" title="Link to this definition"></a></dt>
<dd><p>Nonbonded index (AMBER)</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.NonbondedIndices.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.NonbondedIndices.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Occupancies">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Occupancies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Occupancies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Occupancies" title="Link to this definition"></a></dt>
<dd><dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Occupancies.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Occupancies.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.RMin14s">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">RMin14s</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#RMin14s"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.RMin14s" title="Link to this definition"></a></dt>
<dd><p>The Rmin/2 LJ parameter for 1-4 interactions</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.RMin14s.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.RMin14s.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.RMins">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">RMins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#RMins"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.RMins" title="Link to this definition"></a></dt>
<dd><p>The Rmin/2 LJ parameter</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.RMins.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.RMins.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.RSChirality">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">RSChirality</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#RSChirality"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.RSChirality" title="Link to this definition"></a></dt>
<dd><p>R/S chirality</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Radii">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Radii</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Radii"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Radii" title="Link to this definition"></a></dt>
<dd><p>Radii for each atom</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Radii.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Radii.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.RecordTypes">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">RecordTypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#RecordTypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.RecordTypes" title="Link to this definition"></a></dt>
<dd><p>For PDB-like formats, indicates if ATOM or HETATM</p>
<p>Defaults to ‘ATOM’</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Now stores array of dtype object rather than boolean mapping</p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.RecordTypes.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.RecordTypes.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resids">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Resids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Resids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resids" title="Link to this definition"></a></dt>
<dd><p>Residue ID</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resids.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resids.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ResidueAttr">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">ResidueAttr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ResidueAttr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ResidueAttr" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ResidueAttr.get_atoms">
<span class="sig-name descname"><span class="pre">get_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ResidueAttr.get_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ResidueAttr.get_atoms" title="Link to this definition"></a></dt>
<dd><p>Get atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ResidueAttr.get_residues">
<span class="sig-name descname"><span class="pre">get_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ResidueAttr.get_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ResidueAttr.get_residues" title="Link to this definition"></a></dt>
<dd><p>Get residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ResidueAttr.get_segments">
<span class="sig-name descname"><span class="pre">get_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ResidueAttr.get_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ResidueAttr.get_segments" title="Link to this definition"></a></dt>
<dd><p>By default, the values for each residue present in the set of
segments are returned in a single array. This behavior can be overriden
in child attributes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ResidueAttr.set_atoms">
<span class="sig-name descname"><span class="pre">set_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ResidueAttr.set_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ResidueAttr.set_atoms" title="Link to this definition"></a></dt>
<dd><p>Set atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ResidueAttr.set_residues">
<span class="sig-name descname"><span class="pre">set_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ResidueAttr.set_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ResidueAttr.set_residues" title="Link to this definition"></a></dt>
<dd><p>Set residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ResidueAttr.set_segments">
<span class="sig-name descname"><span class="pre">set_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ResidueAttr.set_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ResidueAttr.set_segments" title="Link to this definition"></a></dt>
<dd><p>Set segmentattributes for a given SegmentGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ResidueStringAttr">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">ResidueStringAttr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ResidueStringAttr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ResidueStringAttr" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.ResidueStringAttr.set_residues">
<span class="sig-name descname"><span class="pre">set_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#ResidueStringAttr.set_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.ResidueStringAttr.set_residues" title="Link to this definition"></a></dt>
<dd><p>Set residue attributes for a given ResidueGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resindices">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Resindices</span></span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Resindices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resindices" title="Link to this definition"></a></dt>
<dd><p>Globally unique resindices for each residue in the group.</p>
<p>If the group is an AtomGroup, then this gives the resindex for each atom in
the group. This unambiguously determines each atom’s residue membership.
Resetting these values changes the residue membership of the atoms.</p>
<p>If the group is a ResidueGroup or SegmentGroup, then this gives the
resindices of each residue represented in the group in a 1-D array, in the
order of the elements in that group.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resindices.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resindices.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resindices.get_atoms">
<span class="sig-name descname"><span class="pre">get_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Resindices.get_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resindices.get_atoms" title="Link to this definition"></a></dt>
<dd><p>Get atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resindices.get_residues">
<span class="sig-name descname"><span class="pre">get_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Resindices.get_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resindices.get_residues" title="Link to this definition"></a></dt>
<dd><p>Get residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resindices.get_segments">
<span class="sig-name descname"><span class="pre">get_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Resindices.get_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resindices.get_segments" title="Link to this definition"></a></dt>
<dd><p>Get segment attributes for a given SegmentGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resindices.set_residues">
<span class="sig-name descname"><span class="pre">set_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Resindices.set_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resindices.set_residues" title="Link to this definition"></a></dt>
<dd><p>Set residue attributes for a given ResidueGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resnames">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Resnames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Resnames"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resnames" title="Link to this definition"></a></dt>
<dd><dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resnames.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resnames.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resnames.sequence">
<span class="sig-name descname"><span class="pre">sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Resnames.sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resnames.sequence" title="Link to this definition"></a></dt>
<dd><p>Returns the amino acid sequence.</p>
<p>The format of the sequence is selected with the keyword <em>format</em>:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p><em>format</em></p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>‘SeqRecord’</p></td>
<td><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code> (default)</p></td>
</tr>
<tr class="row-odd"><td><p>‘Seq’</p></td>
<td><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code></p></td>
</tr>
<tr class="row-even"><td><p>‘string’</p></td>
<td><p>string</p></td>
</tr>
</tbody>
</table>
<p>The sequence is returned by default (keyword <code class="docutils literal notranslate"><span class="pre">format</span> <span class="pre">=</span> <span class="pre">'SeqRecord'</span></code>)
as a <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code> instance, which can then be
further processed. In this case, all keyword arguments (such as the
<em>id</em> string or the <em>name</em> or the <em>description</em>) are directly passed to
<code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code>.</p>
<p>If the keyword <em>format</em> is set to <code class="docutils literal notranslate"><span class="pre">'Seq'</span></code>, all <em>kwargs</em> are ignored
and a <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code> instance is returned. The difference to the
record is that the record also contains metadata and can be directly
used as an input for other functions in <code class="xref py py-mod docutils literal notranslate"><span class="pre">Bio</span></code>.</p>
<p>If the keyword <em>format</em> is set to <code class="docutils literal notranslate"><span class="pre">'string'</span></code>, all <em>kwargs</em> are
ignored and a Python string is returned.</p>
<p class="rubric">Example: Write FASTA file</p>
<p>Use <code class="xref py py-func docutils literal notranslate"><span class="pre">Bio.SeqIO.write()</span></code>, which takes sequence records:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">Bio.SeqIO</span>

<span class="c1"># get the sequence record of a protein component of a Universe</span>
<span class="n">protein</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s2">&quot;protein&quot;</span><span class="p">)</span>
<span class="n">record</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="s2">&quot;myseq1&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;myprotein&quot;</span><span class="p">)</span>

<span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="s2">&quot;single.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>A FASTA file with multiple entries can be written with</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="n">record1</span><span class="p">,</span> <span class="n">record2</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span> <span class="s2">&quot;multi.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>format</strong> (<em>string</em><em>, </em><em>optional</em>) – <ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;string&quot;</span></code>: return sequence as a string of 1-letter codes</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;Seq&quot;</span></code>: return a <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code> instance</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;SeqRecord&quot;</span></code>: return a <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code>
instance</p></li>
</ul>
<blockquote>
<div><p>Default is <code class="docutils literal notranslate"><span class="pre">&quot;SeqRecord&quot;</span></code></p>
</div></blockquote>
</p></li>
<li><p><strong>id</strong> (<em>optional</em>) – Sequence ID for SeqRecord (should be different for different
sequences)</p></li>
<li><p><strong>name</strong> (<em>optional</em>) – Name of the protein.</p></li>
<li><p><strong>description</strong> (<em>optional</em>) – Short description of the sequence.</p></li>
<li><p><strong>kwargs</strong> (<em>optional</em>) – Any other keyword arguments that are understood by
class:<cite>Bio.SeqRecord.SeqRecord</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.12)"><strong>ValueError</strong></a> – </p></li>
<li><p><strong>1-letter IUPAC protein amino acid code; make sure to only</strong> – </p></li>
<li><p><strong>select protein residues.</strong> – </p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.12)"><strong>TypeError</strong></a> – </p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ImportError" title="(in Python v3.12)"><strong>ImportError</strong></a> – </p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.7.0: </span>Biopython is now an optional dependency</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resnums">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Resnums</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Resnums"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resnums" title="Link to this definition"></a></dt>
<dd><dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Resnums.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Resnums.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Segids">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Segids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Segids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Segids" title="Link to this definition"></a></dt>
<dd><dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Segids.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Segids.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Segindices">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Segindices</span></span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Segindices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Segindices" title="Link to this definition"></a></dt>
<dd><p>Globally unique segindices for each segment in the group.</p>
<p>If the group is an AtomGroup, then this gives the segindex for each atom in
the group. This unambiguously determines each atom’s segment membership.
It is not possible to set these, since membership in a segment is an
attribute of each atom’s residue.</p>
<p>If the group is a ResidueGroup or SegmentGroup, then this gives the
segindices of each segment represented in the group in a 1-D array, in the
order of the elements in that group.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Segindices.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Segindices.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Segindices.get_atoms">
<span class="sig-name descname"><span class="pre">get_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Segindices.get_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Segindices.get_atoms" title="Link to this definition"></a></dt>
<dd><p>Get atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Segindices.get_residues">
<span class="sig-name descname"><span class="pre">get_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Segindices.get_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Segindices.get_residues" title="Link to this definition"></a></dt>
<dd><p>Get residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Segindices.get_segments">
<span class="sig-name descname"><span class="pre">get_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Segindices.get_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Segindices.get_segments" title="Link to this definition"></a></dt>
<dd><p>Get segment attributes for a given SegmentGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Segindices.set_segments">
<span class="sig-name descname"><span class="pre">set_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Segindices.set_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Segindices.set_segments" title="Link to this definition"></a></dt>
<dd><p>Set segmentattributes for a given SegmentGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SegmentAttr">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">SegmentAttr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SegmentAttr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SegmentAttr" title="Link to this definition"></a></dt>
<dd><p>Base class for segment attributes.</p>
<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SegmentAttr.get_atoms">
<span class="sig-name descname"><span class="pre">get_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SegmentAttr.get_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SegmentAttr.get_atoms" title="Link to this definition"></a></dt>
<dd><p>Get atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SegmentAttr.get_residues">
<span class="sig-name descname"><span class="pre">get_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SegmentAttr.get_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SegmentAttr.get_residues" title="Link to this definition"></a></dt>
<dd><p>Get residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SegmentAttr.get_segments">
<span class="sig-name descname"><span class="pre">get_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SegmentAttr.get_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SegmentAttr.get_segments" title="Link to this definition"></a></dt>
<dd><p>Get segment attributes for a given SegmentGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SegmentAttr.set_atoms">
<span class="sig-name descname"><span class="pre">set_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SegmentAttr.set_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SegmentAttr.set_atoms" title="Link to this definition"></a></dt>
<dd><p>Set atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SegmentAttr.set_residues">
<span class="sig-name descname"><span class="pre">set_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SegmentAttr.set_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SegmentAttr.set_residues" title="Link to this definition"></a></dt>
<dd><p>Set residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SegmentAttr.set_segments">
<span class="sig-name descname"><span class="pre">set_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SegmentAttr.set_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SegmentAttr.set_segments" title="Link to this definition"></a></dt>
<dd><p>Set segmentattributes for a given SegmentGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SegmentStringAttr">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">SegmentStringAttr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SegmentStringAttr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SegmentStringAttr" title="Link to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SegmentStringAttr.set_segments">
<span class="sig-name descname"><span class="pre">set_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SegmentStringAttr.set_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SegmentStringAttr.set_segments" title="Link to this definition"></a></dt>
<dd><p>Set segmentattributes for a given SegmentGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SolventRadii">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">SolventRadii</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#SolventRadii"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SolventRadii" title="Link to this definition"></a></dt>
<dd><p>Intrinsic solvation radius</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.SolventRadii.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.SolventRadii.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Tempfactors">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">Tempfactors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Tempfactors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Tempfactors" title="Link to this definition"></a></dt>
<dd><p>Tempfactor for atoms</p>
<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Tempfactors.bfactor">
<span class="sig-name descname"><span class="pre">bfactor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Tempfactors.bfactor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Tempfactors.bfactor" title="Link to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Tempfactors.bfactor_setter">
<span class="sig-name descname"><span class="pre">bfactor_setter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Tempfactors.bfactor_setter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Tempfactors.bfactor_setter" title="Link to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Tempfactors.bfactors">
<span class="sig-name descname"><span class="pre">bfactors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Tempfactors.bfactors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Tempfactors.bfactors" title="Link to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Tempfactors.bfactors_setter">
<span class="sig-name descname"><span class="pre">bfactors_setter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#Tempfactors.bfactors_setter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Tempfactors.bfactors_setter" title="Link to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Tempfactors.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Tempfactors.dtype" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.Tempfactors.group">
<span class="sig-name descname"><span class="pre">group</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.Tempfactors.group" title="Link to this definition"></a></dt>
<dd><p>alias of <a class="reference internal" href="groups.html#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">TopologyAttr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#TopologyAttr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr" title="Link to this definition"></a></dt>
<dd><p>Base class for Topology attributes.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This class is intended to be subclassed, and mostly amounts to
a skeleton. The methods here should be present in all
<a class="reference internal" href="#MDAnalysis.core.topologyattrs.TopologyAttr" title="MDAnalysis.core.topologyattrs.TopologyAttr"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyAttr</span></code></a> child classes, but by default they raise
appropriate exceptions.</p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.attrname">
<span class="sig-name descname"><span class="pre">attrname</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.attrname" title="Link to this definition"></a></dt>
<dd><p>the name used for the attribute when attached to a <code class="docutils literal notranslate"><span class="pre">Topology</span></code> object</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.singular">
<span class="sig-name descname"><span class="pre">singular</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.singular" title="Link to this definition"></a></dt>
<dd><p>name for the attribute on a singular object (Atom/Residue/Segment)</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.per_object">
<span class="sig-name descname"><span class="pre">per_object</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.per_object" title="Link to this definition"></a></dt>
<dd><p>If there is a strict mapping between Component and Attribute</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.dtype" title="Link to this definition"></a></dt>
<dd><p>Type to coerce this attribute to be.  For string use ‘object’</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int/float/object</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.top">
<span class="sig-name descname"><span class="pre">top</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.top" title="Link to this definition"></a></dt>
<dd><p>handle for the Topology object TopologyAttr is associated with</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="topology.html#MDAnalysis.core.topology.Topology" title="MDAnalysis.core.topology.Topology">Topology</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#TopologyAttr.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.copy" title="Link to this definition"></a></dt>
<dd><p>Return a deepcopy of this attribute</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.from_blank">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_blank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_atoms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_residues</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_segments</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#TopologyAttr.from_blank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.from_blank" title="Link to this definition"></a></dt>
<dd><p>Create a blank version of this TopologyAttribute</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_atoms</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – Size of the TopologyAttribute atoms</p></li>
<li><p><strong>n_residues</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – Size of the TopologyAttribute residues</p></li>
<li><p><strong>n_segments</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – Size of the TopologyAttribute segments</p></li>
<li><p><strong>values</strong> (<em>optional</em>) – Initial values for the TopologyAttribute</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.get_atoms">
<span class="sig-name descname"><span class="pre">get_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#TopologyAttr.get_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.get_atoms" title="Link to this definition"></a></dt>
<dd><p>Get atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.get_residues">
<span class="sig-name descname"><span class="pre">get_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#TopologyAttr.get_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.get_residues" title="Link to this definition"></a></dt>
<dd><p>Get residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.get_segments">
<span class="sig-name descname"><span class="pre">get_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#TopologyAttr.get_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.get_segments" title="Link to this definition"></a></dt>
<dd><p>Get segment attributes for a given SegmentGroup</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.is_guessed">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_guessed</span></span><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.is_guessed" title="Link to this definition"></a></dt>
<dd><p>Bool of if the source of this information is a guess</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.set_atoms">
<span class="sig-name descname"><span class="pre">set_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#TopologyAttr.set_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.set_atoms" title="Link to this definition"></a></dt>
<dd><p>Set atom attributes for a given AtomGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.set_residues">
<span class="sig-name descname"><span class="pre">set_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#TopologyAttr.set_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.set_residues" title="Link to this definition"></a></dt>
<dd><p>Set residue attributes for a given ResidueGroup</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.TopologyAttr.set_segments">
<span class="sig-name descname"><span class="pre">set_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#TopologyAttr.set_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.TopologyAttr.set_segments" title="Link to this definition"></a></dt>
<dd><p>Set segmentattributes for a given SegmentGroup</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.topologyattrs.UreyBradleys">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.topologyattrs.</span></span><span class="sig-name descname"><span class="pre">UreyBradleys</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">guessed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/topologyattrs.html#UreyBradleys"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.topologyattrs.UreyBradleys" title="Link to this definition"></a></dt>
<dd><p>Angles between two atoms</p>
<p>Initialise with a list of 2 long tuples</p>
<p>These indices refer to the atom indices.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.0.0.</span></p>
</div>
</dd></dl>



           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="topologyobjects.html" class="btn btn-neutral float-left" title="11.3.2. Core Topology Objects — MDAnalysis.core.topologyobjects" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="selection.html" class="btn btn-neutral float-right" title="11.4.1. Atom selection Hierarchy — MDAnalysis.core.selection" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2005-2024, Naveen Michaud-Agrawal, Elizabeth J. Denning, Christian Beckstein (logo), Joshua L. Adelman, Henok Ademtew, Shobhit Agarwal, Aya M. Alaa, Irfan Alibay, Kazi Shudipto Amin, Anshul Angaria, Luís Pedro Borges Araújo, Balasubramanian, Utkarsh Bansal, Patricio Barletta, Leonardo Barneschi, Jonathan Barnoud, Estefania Barreto-Ojeda, Tone Bengtsen, Alejandro Bernardin, Ninad Bhat, Mateusz Bieniek, Kavya Bisht, Wouter Boomsma, Jose Borreguero, Cédric Bouysset, Kevin Boyd, Meet Brijwani, Bart Bruininks, Sébastien Buchoux, Sören von Bülow, Yantong Cai, David Caplan, Yuanyu Chang, Pratham Chauhan, Matthieu Chavent, Haochuan Chen, Xu Hong Chen, Kathleen Clark, Jennifer A Clark, Orion Cohen, Charlie Cook, Ruggero Cortini, Nicholas Craven, Ramon Crehuet, Davide Cruz, Robert Delgado, John Detlefs, Xavier Deupi, Bradley Dice, Jan Domanski, David L. Dotson, Mark D. Driver, Ali Ehlen, Daniel J. Evans, Shujie Fan, Bjarne Feddersen, Lennard van der Feltz, Jake Fennick, Philip Fowler, Guillaume Fraux, Anirvinya G, Michael Gecht, Ahmed Salah Ghoneim, Mikhail Glagolev, William Glass, Jenna M. Swarthout Goddard, Joseph Goose, Alexander Gorfer, Richard J. Gowers, Lukas Grossar, Abhinav Gupta, Akshay Gupta, Pratik Gupta, Sumit Gupta, Benjamin Hall, Ameya Harmalkar, Ivan Hristov, Eugen Hruska, Kyle J. Huston, Siddharth Jain, Edis Jakupovic, Joe Jordan, Henrik Jäger, Uma D Kadam, Aditya Kamath, Jon Kapla, Ian M. Kenney, Aditya Keshari, Haleema Khan, Navya Khare, Utsav Khatu, Andrew William King, Henry Kobin, Abhishek A. Kognole, Kosuke Kudo, Atharva Kulkarni, Manish Kumar, Mohit Kumar, Shubham Kumar, Alia Lescoulie, Zhenbo Li, Max Linke, Philip Loche, Jinju Lu, Hugo MacDermott-Opeskin, Shaivi Malik, Egor Marin, Domenico Marson, Micaela Matta, Andrew R. McCluskey, Robert McGibbon, Rocco Meli, Manuel Nuno Melo, Marcelo C. R. Melo, Dominik &#39;Rathann&#39; Mierzejewski, David Minh, Geongi Moon, Sampurna Mukherjee, Henry Mull, Morgan L. Nance, Fiona B. Naughton, Alex Nesterenko, Hai Nguyen, Sang Young Noh, Meghan Osato, Daniele Padula, Nabarun Pal, Mattia F. Palermo, Dimitrios Papageorgiou, Rafael R. Pappalardo, Vishal Parmar, Danny Parton, Shakul Pathak, Christian Pfaendner, Joshua L. Phillips, Marcelo D. Poleto, Hannah Pollak, Kashish Punjani, Michael Quevillon, Vedant Rathore, Tyler Reddy, Pedro Reis, Paul Rigor, Andrea Rizzi, Xiaoxu Ruan, Carlos Yanez S., Utkarsh Saxena, Moritz Schaeffler, Alexander Schlaich, Marcello Sega, Ricky Sexton, Sean L. Seyler, Faraaz Shah, Sulay Shah, Abhishek Shandilya, Shubham Sharma, Rishabh Shukla, Karthikeyan Singaravelan, Tamandeep Singh, Brigitta Sipőcz, Paul Smith, Andy Somogyi, Caio S. Souza, David van der Spoel, Shantanu Srivastava, Lukas Stelzl, Jan Stevens, Gorman Stock, Philipp Stärk, Johannes Stöckelmaier, Fenil Suchak, Ayush Suhane, Filip T. Szczypiński, Sukeerti T, Matthijs Tadema, Joao Miguel Correia Teixeira, Paarth Thadani, Matthew W. Thompson, Hao Tian, Matteo Tiberti, Zaheer Timol, Wiep van der Toorn, Mieczyslaw Torchala, Aditi Tripathi, Heet Vekariya, Mark Verma, Josh Vermaas, Isaac Virshup, Lily Wang, Leon Wehrhan, Nestor Wendt, Lawson Woods, Zhiyi Wu, Tengyu Xie, Zhuyi Xue, Mingyi Xue, Alexander Yang, Juan Eiros Zamora, Johannes Zeman, Yibo Zhang, Raymond Zhao, Yuxuan Zhuang, and Oliver Beckstein.</p>
  </div>

  

<div class="footer"><p>Please see
    our <a href="https://www.mdanalysis.org/pages/privacy/">Privacy Policy</a>
    to learn how <a href="https://www.mdanalysis.org">MDAnalysis</a> collects data.</p>
    <script data-goatcounter="https://mdanalysis.goatcounter.com/count"
        async src="//gc.zgo.at/count.js"></script>
</div>



</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
    var versions_json_url = 'https://docs.mdanalysis.org/versions.json'
</script>

<div class="rst-versions" data-toggle="rst-versions" role="note"
     aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      <span class="fa fa-book"></span>
        2.8.0-dev0
      <span class="fa fa-caret-down"></span>
    </span>

    <div class="rst-other-versions">
        <dl id="versionselector">
            <dt>Other Versions</dt>
        </dl>

    </div>
</div><script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>