<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>11.2.2. Core objects: Containers — MDAnalysis.core.groups &mdash; MDAnalysis 2.5.0 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/msmb.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
    <link rel="shortcut icon" href="../../_static/mdanalysis-logo.ico"/>
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script src="../../_static/js/versions.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="search" type="application/opensearchdescription+xml"
          title="Search within MDAnalysis 2.5.0 documentation"
          href="../../_static/opensearch.xml"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="11.3.1. Core Topology object — MDAnalysis.core.topology" href="topology.html" />
    <link rel="prev" title="11.2.1. Core object: Universe — MDAnalysis.core.universe" href="universe.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >

          
          
          <a href="../../index.html">
            
              <img src="../../_static/mdanalysis-logo-thin.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                2.5.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../overview.html">1. Overview over MDAnalysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../topology.html">2. The topology system</a></li>
<li class="toctree-l1"><a class="reference internal" href="../selections.html">3. Selection commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../analysis_modules.html">4. Analysis modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../topology_modules.html">5. Topology modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../coordinates_modules.html">6. Coordinates modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../converters.html">7. Converter modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectory_transformations.html">8. Trajectory transformations (“on-the-fly” transformations)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../selections_modules.html">9. Selection exporters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../auxiliary_modules.html">10. Auxiliary modules</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../core_modules.html">11. Core modules</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="init.html">11.1. Core functions of MDAnalysis</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../core_modules.html#important-objects-for-users">11.2. Important objects for users</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="universe.html">11.2.1. Core object: Universe — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.core.universe</span></code></a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">11.2.2. Core objects: Containers — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.core.groups</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#classes">11.2.2.1. Classes</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../core_modules.html#topology-system">11.3. Topology system</a></li>
<li class="toctree-l2"><a class="reference internal" href="../core_modules.html#selection-system">11.4. Selection system</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../visualization_modules.html">12. Visualization modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../lib_modules.html">13. Library functions — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.lib</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../version.html">14. Version information for MDAnalysis - <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.version</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../units.html">15. Constants and unit conversion — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.units</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../exceptions.html">16. Custom exceptions and warnings — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.exceptions</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">17. References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: white" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">MDAnalysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../core_modules.html"><span class="section-number">11. </span>Core modules</a></li>
      <li class="breadcrumb-item active"><span class="section-number">11.2.2. </span>Core objects: Containers — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.core.groups</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/documentation_pages/core/groups.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <span class="target" id="module-MDAnalysis.core.groups"></span><section id="core-objects-containers-mdanalysis-core-groups">
<h1><span class="section-number">11.2.2. </span>Core objects: Containers — <a class="reference internal" href="#module-MDAnalysis.core.groups" title="MDAnalysis.core.groups"><code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.core.groups</span></code></a><a class="headerlink" href="#core-objects-containers-mdanalysis-core-groups" title="Permalink to this heading"></a></h1>
<p>The <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> instance contains all the
particles in the system (which MDAnalysis calls <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>). Groups of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">atoms</span></code></a> are handled as <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> instances. The
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is probably the most important object in MDAnalysis because
virtually everything can be accessed through it. <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> instances
can be easily created (e.g., from an <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.select_atoms" title="MDAnalysis.core.groups.AtomGroup.select_atoms"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AtomGroup.select_atoms()</span></code></a> selection or
simply by slicing).</p>
<p>For convenience, chemically meaningful groups of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> such as a
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> or a <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> (typically a whole molecule or all of the
solvent) also exist as containers, as well as groups of these units
(<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>).</p>
<section id="classes">
<h2><span class="section-number">11.2.2.1. </span>Classes<a class="headerlink" href="#classes" title="Permalink to this heading"></a></h2>
<section id="collections">
<h3><span class="section-number">11.2.2.1.1. </span>Collections<a class="headerlink" href="#collections" title="Permalink to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.groups.</span></span><span class="sig-name descname"><span class="pre">AtomGroup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#AtomGroup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup" title="Permalink to this definition"></a></dt>
<dd><p>An ordered array of atoms.</p>
<p>Can be initiated from an iterable of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ag</span> <span class="o">=</span> <span class="n">AtomGroup</span><span class="p">([</span><span class="n">Atom1</span><span class="p">,</span> <span class="n">Atom2</span><span class="p">,</span> <span class="n">Atom3</span><span class="p">])</span>
</pre></div>
</div>
<p>Or from providing a list of indices and the
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> it should belong to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ag</span> <span class="o">=</span> <span class="n">AtomGroup</span><span class="p">([</span><span class="mi">72</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">25</span><span class="p">],</span> <span class="n">u</span><span class="p">)</span>
</pre></div>
</div>
<p>Alternatively, an <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is generated by indexing/slicing
another <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>, such as the group of all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in
the <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> at
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe.atoms" title="MDAnalysis.core.universe.Universe.atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MDAnalysis.core.universe.Universe.atoms</span></code></a>.</p>
<p>An <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> can be indexed and sliced like a list:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ag</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ag</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>will return the first and the last <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> in the group whereas the
slice:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ag</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">6</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
</pre></div>
</div>
<p>returns an <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> of every second element, corresponding to
indices 0, 2, and 4.</p>
<p>It also supports “advanced slicing” when the argument is a
<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> or a <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aslice</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span>
<span class="n">ag</span><span class="p">[</span><span class="n">aslice</span><span class="p">]</span>
</pre></div>
</div>
<p>will return a new <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> with those
indices in the old <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p>Finally, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroups</span></code></a> can be created from a selection.
See <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.select_atoms" title="MDAnalysis.core.groups.AtomGroup.select_atoms"><code class="xref py py-meth docutils literal notranslate"><span class="pre">select_atoms()</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroups</span></code></a> originating from a selection are sorted
and duplicate elements are removed. This is not true for
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroups</span></code></a> produced by slicing. Thus, slicing can be
used when the order of atoms is crucial (for instance, in order to
define angles or dihedrals).</p>
</div>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroups</span></code></a> can be compared and combined using group
operators. For instance, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroups</span></code></a> can be concatenated
using <cite>+</cite> or <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.concatenate" title="MDAnalysis.core.groups.AtomGroup.concatenate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">concatenate()</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ag_concat</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">+</span> <span class="n">ag2</span>  <span class="c1"># or ag_concat = ag1.concatenate(ag2)</span>
</pre></div>
</div>
<p>When groups are concatenated, the order of the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> is
conserved. If <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> appear several times in one of the
groups, the duplicates are kept in the resulting group. On the contrary to
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.concatenate" title="MDAnalysis.core.groups.AtomGroup.concatenate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">concatenate()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.union" title="MDAnalysis.core.groups.AtomGroup.union"><code class="xref py py-meth docutils literal notranslate"><span class="pre">union()</span></code></a> treats the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroups</span></code></a>
as sets so that duplicates are removed from the resulting group, and
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> are ordered. The <cite>|</cite> operator is synomymous to
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.union" title="MDAnalysis.core.groups.AtomGroup.union"><code class="xref py py-meth docutils literal notranslate"><span class="pre">union()</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ag_union</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">|</span> <span class="n">ag2</span>  <span class="c1"># or ag_union = ag1.union(ag2)</span>
</pre></div>
</div>
<p>The opposite operation to <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.concatenate" title="MDAnalysis.core.groups.AtomGroup.concatenate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">concatenate()</span></code></a> is <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.subtract" title="MDAnalysis.core.groups.AtomGroup.subtract"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subtract()</span></code></a>. This
method creates a new group with all the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> of the group
that are not in a given other group; the order of the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
is kept, and so are duplicates. <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.difference" title="MDAnalysis.core.groups.AtomGroup.difference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">difference()</span></code></a> is the set version of
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.subtract" title="MDAnalysis.core.groups.AtomGroup.subtract"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subtract()</span></code></a>. The resulting group is sorted and deduplicated.</p>
<p>All set methods are listed in the table below. These methods treat the
groups as sorted and deduplicated sets of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 44%" />
<col style="width: 17%" />
<col style="width: 39%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Operation</p></th>
<th class="head"><p>Equivalent</p></th>
<th class="head"><p>Result</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">s.isdisjoint(t)</span></code></p></td>
<td></td>
<td><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if <code class="docutils literal notranslate"><span class="pre">s</span></code> and
<code class="docutils literal notranslate"><span class="pre">t</span></code> do not share
elements</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">s.issubset(t)</span></code></p></td>
<td></td>
<td><p>test if all elements of
<code class="docutils literal notranslate"><span class="pre">s</span></code> are part of <code class="docutils literal notranslate"><span class="pre">t</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">s.is_strict_subset(t)</span></code></p></td>
<td></td>
<td><p>test if all elements of
<code class="docutils literal notranslate"><span class="pre">s</span></code> are part of <code class="docutils literal notranslate"><span class="pre">t</span></code>,
and <code class="docutils literal notranslate"><span class="pre">s</span> <span class="pre">!=</span> <span class="pre">t</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">s.issuperset(t)</span></code></p></td>
<td></td>
<td><p>test if all elements of
<code class="docutils literal notranslate"><span class="pre">t</span></code> are part of <code class="docutils literal notranslate"><span class="pre">s</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">s.is_strict_superset(t)</span></code></p></td>
<td></td>
<td><p>test if all elements of
<code class="docutils literal notranslate"><span class="pre">t</span></code> are part of <code class="docutils literal notranslate"><span class="pre">s</span></code>,
and <code class="docutils literal notranslate"><span class="pre">s</span> <span class="pre">!=</span> <span class="pre">t</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">s.union(t)</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">s</span> <span class="pre">|</span> <span class="pre">t</span></code></p></td>
<td><p>new Group with elements
from both <code class="docutils literal notranslate"><span class="pre">s</span></code> and <code class="docutils literal notranslate"><span class="pre">t</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">s.intersection(t)</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">s</span> <span class="pre">&amp;</span> <span class="pre">t</span></code></p></td>
<td><p>new Group with elements
common to <code class="docutils literal notranslate"><span class="pre">s</span></code> and <code class="docutils literal notranslate"><span class="pre">t</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">s.difference(t)</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">s</span> <span class="pre">-</span> <span class="pre">t</span></code></p></td>
<td><p>new Group with elements of
<code class="docutils literal notranslate"><span class="pre">s</span></code> that are not in <code class="docutils literal notranslate"><span class="pre">t</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">s.symmetric_difference(t)</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">s</span> <span class="pre">^</span> <span class="pre">t</span></code></p></td>
<td><p>new Group with elements
that are part of <code class="docutils literal notranslate"><span class="pre">s</span></code> or
<code class="docutils literal notranslate"><span class="pre">t</span></code> but not both</p></td>
</tr>
</tbody>
</table>
<p>The following methods keep the order of the atoms as well as duplicates.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 44%" />
<col style="width: 17%" />
<col style="width: 39%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Operation</p></th>
<th class="head"><p>Equivalent</p></th>
<th class="head"><p>Result</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">len(s)</span></code></p></td>
<td></td>
<td><p>number of elements (atoms,
residues or segment) in
the group</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">s</span> <span class="pre">==</span> <span class="pre">t</span></code></p></td>
<td></td>
<td><p>test if <code class="docutils literal notranslate"><span class="pre">s</span></code> and <code class="docutils literal notranslate"><span class="pre">t</span></code>
contain the same elements
in the same order</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">s.concatenate(t)</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">s</span> <span class="pre">+</span> <span class="pre">t</span></code></p></td>
<td><p>new Group with elements
from <code class="docutils literal notranslate"><span class="pre">s</span></code> and from <code class="docutils literal notranslate"><span class="pre">t</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">s.subtract(t)</span></code></p></td>
<td></td>
<td><p>new Group with elements
from <code class="docutils literal notranslate"><span class="pre">s</span></code> that are not
in <code class="docutils literal notranslate"><span class="pre">t</span></code></p></td>
</tr>
</tbody>
</table>
<p>The <cite>in</cite> operator allows to test if an <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> is in the
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> instances are always bound to a
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.universe.Universe</span></code></a>. They cannot exist in isolation.</p>
<p>During serialization, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> will be pickled with its bound
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.universe.Universe</span></code></a> which means after unpickling,
a new <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.universe.Universe</span></code></a> will be created and
be attached by the new <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>. If the Universe is serialized
with its <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>, they will still be bound together afterwards:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">g_pickled</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;g_pickled.universe is u: &quot;</span><span class="p">,</span> <span class="n">u</span> <span class="ow">is</span> <span class="n">g_pickled</span><span class="o">.</span><span class="n">universe</span><span class="p">)</span>
<span class="go">g_pickled.universe is u: False</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">g_pickled</span><span class="p">,</span> <span class="n">u_pickled</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">u</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;g_pickled.universe is u_pickled: &quot;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>      <span class="n">u_pickle</span> <span class="ow">is</span> <span class="n">g_pickled</span><span class="o">.</span><span class="n">universe</span><span class="p">)</span>
<span class="go">g_pickled.universe is u_pickled: True</span>
</pre></div>
</div>
<p>If multiple <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> are bound to the same
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.universe.Universe</span></code></a>, they will bound to the same one
after serialization:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">h</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">g_pickled</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">h_pickled</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">h</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;g_pickled.universe is h_pickled.universe : &quot;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>      <span class="n">g_pickled</span><span class="o">.</span><span class="n">universe</span> <span class="ow">is</span> <span class="n">h_pickled</span><span class="o">.</span><span class="n">universe</span><span class="p">)</span>
<span class="go">g_pickled.universe is h_pickled.universe: False</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">g_pickled</span><span class="p">,</span> <span class="n">h_pickled</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">h</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;g_pickled.universe is h_pickled.universe: &quot;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>      <span class="n">g_pickle</span><span class="o">.</span><span class="n">universe</span> <span class="ow">is</span> <span class="n">h_pickled</span><span class="o">.</span><span class="n">universe</span><span class="p">)</span>
<span class="go">g_pickled.universe is h_pickled.universe: True</span>
</pre></div>
</div>
<p>The aforementioned two cases are useful for implementation of parallel
analysis base classes. First, you always get an independent
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.universe.Universe</span></code></a>
in the new process; you don’t have to worry about detaching and reattaching
Universe with <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>. It also means the state of the
new pickled AtomGroup will not be changed with the old Universe,
So either the Universe has to pickled together with the AtomGroup
(e.g. as a tuple, or as attributes of the object to be pickled), or the
implicit new Universe (<cite>AtomGroup.Universe</cite>) needs to be used.
Second, When multiple AtomGroup need to be pickled, they will recognize if
they belong to the same Univese or not.
Also keep in mind that they need to be pickled together.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.universe.Universe</span></code></a></p>
</div>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.16.2: </span><em>Instant selectors</em> of <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> will be removed in the 1.0
release.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed instant selectors, use select_atoms(‘name …’) to select
atoms by name.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.0.0: </span><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> can always be pickled with or without its universe,
instead of failing when not finding its anchored universe.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Indexing an AtomGroup with <code class="docutils literal notranslate"><span class="pre">None</span></code> raises a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code>.</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.accumulate">
<span class="sig-name descname"><span class="pre">accumulate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attribute</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">function=&lt;function</span> <span class="pre">sum&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound='group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.accumulate" title="Permalink to this definition"></a></dt>
<dd><p>Accumulates the attribute associated with (compounds of) the group.</p>
<p>Accumulates the attribute of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
The accumulation per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule,
or fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. By default, the method sums up all attributes per compound,
but any function that takes an array and returns an acuumulation over a
given axis can be used. For multi-dimensional input arrays, the
accumulation is performed along the first axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> or </em><em>array_like</em>) – Attribute or array of values to accumulate.
If a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> (or compatible) is provided, its first
dimension must have the same length as the total number of atoms in
the group.</p></li>
<li><p><strong>function</strong> (<em>callable</em><em>, </em><em>optional</em>) – The function performing the accumulation. It must take the array of
attribute values to accumulate as its only positional argument and
accept an (optional) keyword argument <code class="docutils literal notranslate"><span class="pre">axis</span></code> allowing to specify
the axis along which the accumulation is performed.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the accumulation of all attributes associated with
atoms in the group will be returned as a single value. Otherwise,
the accumulation of the attributes per <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>,
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment will be returned as a 1d
array. Note that, in any case, <em>only</em> the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Acuumulation of the <cite>attribute</cite>.
If <cite>compound</cite> is set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the first dimension of the
<cite>attribute</cite> array will be contracted to a single value.
If <cite>compound</cite> is set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>, <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>, the length of the first
dimension will correspond to the number of compounds. In all cases,
the other dimensions of the returned array will be of the original
shape (without the first dimension).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If the length of a provided <cite>attribute</cite> array does not correspond to
    the number of atoms in the group.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>compound</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>,
    <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>.</p></li>
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecule'</span></code> but the topology doesn’t
    contain molecule information (molnums), or if <cite>compound</cite> is
    <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> but the topology doesn’t contain bonds.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>To find the total charge of a given <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;prop mass &gt; 4.0&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">accumulate</span><span class="p">(</span><span class="s1">&#39;charges&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>To find the total mass per residue of all CA <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">accumulate</span><span class="p">(</span><span class="s1">&#39;masses&#39;</span><span class="p">,</span> <span class="n">compound</span><span class="o">=</span><span class="s1">&#39;residues&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>To find the maximum atomic charge per fragment of a given
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">accumulate</span><span class="p">(</span><span class="s1">&#39;charges&#39;</span><span class="p">,</span> <span class="n">compound</span><span class="o">=</span><span class="s2">&quot;fragments&quot;</span><span class="p">,</span> <span class="n">function</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">)</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.align_principal_axis">
<span class="sig-name descname"><span class="pre">align_principal_axis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">axis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.align_principal_axis" title="Permalink to this definition"></a></dt>
<dd><p>Align principal axis with index <cite>axis</cite> with <cite>vector</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>{0</em><em>, </em><em>1</em><em>, </em><em>2}</em>) – Index of the principal axis (0, 1, or 2), as produced by
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.principal_axes" title="MDAnalysis.core.groups.AtomGroup.principal_axes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">principal_axes()</span></code></a>.</p></li>
<li><p><strong>vector</strong> (<em>array_like</em>) – Vector to align principal axis with.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>To align the long axis of a channel (the first principal axis, i.e.
<em>axis</em> = 0) with the z-axis:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">align_principal_axis</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;aligned.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.angle">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">angle</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.angle" title="Permalink to this definition"></a></dt>
<dd><p>This <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> represented as an
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.Angle" title="MDAnalysis.core.topologyobjects.Angle"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.topologyobjects.Angle</span></code></a> object</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is not length 3</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.11.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.asphericity">
<span class="sig-name descname"><span class="pre">asphericity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.asphericity" title="Permalink to this definition"></a></dt>
<dd><p>Asphericity.</p>
<p>See <a class="reference internal" href="../references.html#dima2004b" id="id1"><span>[Dima2004b]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <em>unwrap</em> and <em>compound</em> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.5.0: </span>Added calculation for any <cite>compound</cite> type</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.asunique">
<span class="sig-name descname"><span class="pre">asunique</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sorted</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#AtomGroup.asunique"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.asunique" title="Permalink to this definition"></a></dt>
<dd><p>Return a <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> containing unique
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> only, with optional sorting.</p>
<p>If the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is unique, this is the group itself.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sorted</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Whether or not the returned AtomGroup should be sorted
by index.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Unique <code class="docutils literal notranslate"><span class="pre">AtomGroup</span></code></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">MDAnalysis</span> <span class="k">as</span> <span class="nn">mda</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">asunique</span><span class="p">(</span><span class="nb">sorted</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="ow">is</span> <span class="n">ag</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([2, 1, 0], dtype=int64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">asunique</span><span class="p">(</span><span class="nb">sorted</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="ow">is</span> <span class="n">ag</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([0, 1, 2], dtype=int64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">asunique</span><span class="p">(</span><span class="nb">sorted</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([2, 1, 0], dtype=int64)</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.0.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.atoms" title="Permalink to this definition"></a></dt>
<dd><p>The <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> itself.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.copy" title="MDAnalysis.core.groups.AtomGroup.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a></dt><dd><p>return a true copy of the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a></p>
</dd>
</dl>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>In previous versions, this returned a copy, but now
the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> itself is returned. This should
not affect any code but only speed up calculations.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.bbox">
<span class="sig-name descname"><span class="pre">bbox</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.bbox" title="Permalink to this definition"></a></dt>
<dd><p>Return the bounding box of the selection.</p>
<p>The lengths A,B,C of the orthorhombic enclosing box are</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">L</span> <span class="o">=</span> <span class="n">AtomGroup</span><span class="o">.</span><span class="n">bbox</span><span class="p">()</span>
<span class="n">A</span><span class="p">,</span><span class="n">B</span><span class="p">,</span><span class="n">C</span> <span class="o">=</span> <span class="n">L</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">L</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to the primary unit cell
before calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>corners</strong> – 2x3 array giving corners of bounding box as
<code class="docutils literal notranslate"><span class="pre">[[xmin,</span> <span class="pre">ymin,</span> <span class="pre">zmin],</span> <span class="pre">[xmax,</span> <span class="pre">ymax,</span> <span class="pre">zmax]]</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.2.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.bfactors">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bfactors</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.bfactors" title="Permalink to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have tempfactors. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.bond">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bond</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.bond" title="Permalink to this definition"></a></dt>
<dd><p>This <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> represented as a
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.Bond" title="MDAnalysis.core.topologyobjects.Bond"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.topologyobjects.Bond</span></code></a> object</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is not length 2</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.11.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.bsphere">
<span class="sig-name descname"><span class="pre">bsphere</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.bsphere" title="Permalink to this definition"></a></dt>
<dd><p>Return the bounding sphere of the selection.</p>
<p>The sphere is calculated relative to the
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.center_of_geometry" title="MDAnalysis.core.groups.AtomGroup.center_of_geometry"><code class="xref py py-meth docutils literal notranslate"><span class="pre">center</span> <span class="pre">of</span> <span class="pre">geometry</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms to the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>R</strong> (<em>float</em>) – Radius of the bounding sphere.</p></li>
<li><p><strong>center</strong> (<em>numpy.ndarray</em>) – Coordinates of the sphere center as <code class="docutils literal notranslate"><span class="pre">[xcen,</span> <span class="pre">ycen,</span> <span class="pre">zcen]</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.3.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.center">
<span class="sig-name descname"><span class="pre">center</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">weights</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.center" title="Permalink to this definition"></a></dt>
<dd><p>Weighted center of (compounds of) the group</p>
<p>Computes the weighted center of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Weighted centers per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the weights of a compound sum up to zero, the
coordinates of that compound’s weighted center will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>weights</strong> (<em>array_like</em><em> or </em><em>None</em>) – Weights to be used. Setting <cite>weights=None</cite> is equivalent to passing
identical weights for all atoms of the group.</p></li>
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">'group'</span></code> the center of each
compound will be calculated without moving any
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to keep the compounds intact. Instead, the
resulting position vectors will be moved to the primary unit cell
after calculation. Default [<code class="docutils literal notranslate"><span class="pre">False</span></code>].</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – <dl class="simple">
<dt>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their</dt><dd><p>centers.</p>
</dd>
</dl>
</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the weighted center of all atoms in the group will
be returned as a single position vector. Else, the weighted centers
of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment
will be returned as an array of position vectors, i.e. a 2d array.
Note that, in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the weighted center(s) of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>, <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>, the output will be a 2d array
of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>compound</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>,
    <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If both ‘wrap’ and ‘unwrap’ set to true.</p></li>
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecule'</span></code> but the topology doesn’t
    contain molecule information (molnums) or if <cite>compound</cite> is
    <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> but the topology doesn’t contain bonds.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>To find the center of charge of a given <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;prop mass &gt; 4.0&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">center</span><span class="p">(</span><span class="n">sel</span><span class="o">.</span><span class="n">charges</span><span class="p">)</span>
</pre></div>
</div>
<p>To find the centers of mass per residue of all CA <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">center</span><span class="p">(</span><span class="n">sel</span><span class="o">.</span><span class="n">masses</span><span class="p">,</span> <span class="n">compound</span><span class="o">=</span><span class="s1">&#39;residues&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>
compounds</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.center_of_charge">
<span class="sig-name descname"><span class="pre">center_of_charge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.center_of_charge" title="Permalink to this definition"></a></dt>
<dd><p>Center of (absolute) charge of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \vert q_i \vert \boldsymbol r_i}
                     {\sum_i \vert q_i \vert}\]</div>
<p>where <span class="math notranslate nohighlight">\(q_i\)</span> is the charge and <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> the
position of atom <span class="math notranslate nohighlight">\(i\)</span> in the given
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup</span></code></a>.
Centers of charge per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the charges of a compound sum up to zero, the
center of mass coordinates of that compound will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of mass of all atoms in the group will
be returned as a single position vector. Otherwise, the centers of
mass of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the center(s) of charge of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where
<code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method can only be accessed if the underlying topology has
information about atomic charges.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.2.0.</span></p>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.center_of_geometry">
<span class="sig-name descname"><span class="pre">center_of_geometry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.center_of_geometry" title="Permalink to this definition"></a></dt>
<dd><p>Center of geometry of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \boldsymbol r_i}{\sum_i 1}\]</div>
<p>where <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <span class="math notranslate nohighlight">\(i\)</span>.
Centers of geometry per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> or per <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> can
be obtained by setting the <cite>compound</cite> parameter accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation. If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is
<code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the center of each compound will
be calculated without moving any <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to keep the
compounds intact. Instead, the resulting position vectors will be
moved to the primary unit cell after calculation. Default False.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of geometry of all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
in the group will be returned as a single position vector. Else,
the centers of geometry of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> will be returned as an array of position vectors,
i.e. a 2d array. Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the geometric center(s) of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the
number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>
compounds</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.center_of_mass">
<span class="sig-name descname"><span class="pre">center_of_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.center_of_mass" title="Permalink to this definition"></a></dt>
<dd><p>Center of mass of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i m_i \boldsymbol r_i}{\sum m_i}\]</div>
<p>where <span class="math notranslate nohighlight">\(m_i\)</span> is the mass and <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> the
position of atom <span class="math notranslate nohighlight">\(i\)</span> in the given
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup</span></code></a>.
Centers of mass per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the masses of a compound sum up to zero, the
center of mass coordinates of that compound will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of mass of all atoms in the group will
be returned as a single position vector. Otherwise, the centers of
mass of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e. a 2d
array.
Note that, in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the center(s) of mass of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code>
is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method can only be accessed if the underlying topology has
information about atomic masses.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> compounds;
added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.centroid">
<span class="sig-name descname"><span class="pre">centroid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.centroid" title="Permalink to this definition"></a></dt>
<dd><p>Center of geometry of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \boldsymbol r_i}{\sum_i 1}\]</div>
<p>where <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <span class="math notranslate nohighlight">\(i\)</span>.
Centers of geometry per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> or per <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> can
be obtained by setting the <cite>compound</cite> parameter accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation. If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is
<code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the center of each compound will
be calculated without moving any <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to keep the
compounds intact. Instead, the resulting position vectors will be
moved to the primary unit cell after calculation. Default False.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of geometry of all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
in the group will be returned as a single position vector. Else,
the centers of geometry of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> will be returned as an array of position vectors,
i.e. a 2d array. Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the geometric center(s) of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the
number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>
compounds</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.cmap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cmap</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.cmap" title="Permalink to this definition"></a></dt>
<dd><p>This <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> represented as an
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.CMap" title="MDAnalysis.core.topologyobjects.CMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.topologyobjects.CMap</span></code></a> object</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is not length 5</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.0.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.concatenate">
<span class="sig-name descname"><span class="pre">concatenate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.concatenate" title="Permalink to this definition"></a></dt>
<dd><p>Concatenate with another Group or Component of the same level.</p>
<p>Duplicate entries and original order is preserved. It is synomymous to
the <cite>+</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with elements of <cite>self</cite> and <cite>other</cite> concatenated</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>The order of the original contents (including duplicates)
are preserved when performing a concatenation.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name O&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name N&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">+</span> <span class="n">ag2</span>  <span class="c1"># or ag1.concatenate(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">names</span>
<span class="go">array([&#39;O&#39;, &#39;O&#39;, &#39;O&#39;], dtype=object)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]</span><span class="o">.</span><span class="n">names</span>
<span class="go">array([&#39;N&#39;, &#39;N&#39;, &#39;N&#39;], dtype=object)</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.0.</span></p>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.convert_to">
<span class="sig-name descname"><span class="pre">convert_to</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.convert_to" title="Permalink to this definition"></a></dt>
<dd><p>alias of <a class="reference internal" href="accessors.html#MDAnalysis.core.accessors.ConverterWrapper" title="MDAnalysis.core.accessors.ConverterWrapper"><code class="xref py py-class docutils literal notranslate"><span class="pre">ConverterWrapper</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.copy" title="Permalink to this definition"></a></dt>
<dd><p>Get another group identical to this one.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.19.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.difference">
<span class="sig-name descname"><span class="pre">difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.difference" title="Permalink to this definition"></a></dt>
<dd><p>Elements from this Group that do not appear in another</p>
<p>This method removes duplicate elements and sorts the result. As such,
it is different from <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.subtract" title="MDAnalysis.core.groups.AtomGroup.subtract"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subtract()</span></code></a>. <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.difference" title="MDAnalysis.core.groups.AtomGroup.difference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">difference()</span></code></a> is synomymous
to the <cite>-</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the elements of <cite>self</cite> that are not in  <cite>other</cite>, without
duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.subtract" title="MDAnalysis.core.groups.AtomGroup.subtract"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subtract</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.symmetric_difference" title="MDAnalysis.core.groups.AtomGroup.symmetric_difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">symmetric_difference</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.dihedral">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dihedral</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.dihedral" title="Permalink to this definition"></a></dt>
<dd><p>This <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> represented as a
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.Dihedral" title="MDAnalysis.core.topologyobjects.Dihedral"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dihedral</span></code></a> object</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is not length 4</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.11.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.dimensions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dimensions</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.dimensions" title="Permalink to this definition"></a></dt>
<dd><p>Obtain a copy of the dimensions of the currently loaded Timestep</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.dipole_moment">
<span class="sig-name descname"><span class="pre">dipole_moment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.dipole_moment" title="Permalink to this definition"></a></dt>
<dd><p>Dipole moment of the group or compounds in a group.</p>
<div class="math notranslate nohighlight">
\[\mu = |\boldsymbol{\mu}| = \sqrt{ \sum_{i=1}^{D} \mu^2 }\]</div>
<p>Where <span class="math notranslate nohighlight">\(D\)</span> is the number of dimensions, in this case 3.</p>
<p>Computes the dipole moment of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Dipole per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is a net charge, the magnitude of the dipole
moment is dependent on the <cite>center</cite> chosen.
See <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.dipole_vector" title="MDAnalysis.core.groups.AtomGroup.dipole_vector"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dipole_vector()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single dipole vector returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dipole moment(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.dipole_vector">
<span class="sig-name descname"><span class="pre">dipole_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mass'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.dipole_vector" title="Permalink to this definition"></a></dt>
<dd><p>Dipole vector of the group.</p>
<div class="math notranslate nohighlight">
\[\boldsymbol{\mu} = \sum_{i=1}^{N} q_{i} ( \mathbf{r}_{i} -
\mathbf{r}_{COM} )\]</div>
<p>Computes the dipole vector of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Dipole vector per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that the magnitude of the dipole moment is independent of the
<code class="docutils literal notranslate"><span class="pre">center</span></code> chosen unless the species has a net charge. In the case of
a charged group the dipole moment can be later adjusted  with:</p>
<div class="math notranslate nohighlight">
\[\boldsymbol{\mu}_{COC} = \boldsymbol{\mu}_{COM} +
q_{ag}\mathbf{r}_{COM} - q_{ag}\boldsymbol{r}_{COC}\]</div>
<p>Where <span class="math notranslate nohighlight">\(\mathbf{r}_{COM}\)</span> is the center of mass and
<span class="math notranslate nohighlight">\(\mathbf{r}_{COC}\)</span> is the center of charge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single dipole vector returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dipole vector(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,3)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.forces">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">forces</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.forces" title="Permalink to this definition"></a></dt>
<dd><p>Forces on each <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> in the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p>A <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> with
<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.shape.html#numpy.ndarray.shape" title="(in NumPy v1.25)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">shape</span></code></a><code class="docutils literal notranslate"><span class="pre">=(</span></code><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.n_atoms" title="MDAnalysis.core.groups.AtomGroup.n_atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">n_atoms</span></code></a><code class="docutils literal notranslate"><span class="pre">,</span> <span class="pre">3)</span></code>
and <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.dtype.html#numpy.ndarray.dtype" title="(in NumPy v1.25)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dtype</span></code></a><code class="docutils literal notranslate"><span class="pre">=numpy.float32</span></code>.</p>
<p>The forces can be changed by assigning an array of the appropriate
shape, i.e. either <code class="docutils literal notranslate"><span class="pre">(</span></code><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.n_atoms" title="MDAnalysis.core.groups.AtomGroup.n_atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">n_atoms</span></code></a><code class="docutils literal notranslate"><span class="pre">,</span> <span class="pre">3)</span></code> to assign
individual forces or <code class="docutils literal notranslate"><span class="pre">(3,)</span></code> to assign the <em>same</em> force to all
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> (e.g. <code class="docutils literal notranslate"><span class="pre">ag.forces</span> <span class="pre">=</span> <span class="pre">array([0,0,0])</span></code> will give all
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> a zero <a class="reference internal" href="#MDAnalysis.core.groups.Atom.force" title="MDAnalysis.core.groups.Atom.force"><code class="xref py py-attr docutils literal notranslate"><span class="pre">force</span></code></a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If the <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a> does not
    contain <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.forces" title="MDAnalysis.coordinates.timestep.Timestep.forces"><code class="xref py py-attr docutils literal notranslate"><span class="pre">forces</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.fragindices">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fragindices</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.fragindices" title="Permalink to this definition"></a></dt>
<dd><p>The
<a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.Bonds.fragindex" title="MDAnalysis.core.topologyattrs.Bonds.fragindex"><code class="xref py py-class docutils literal notranslate"><span class="pre">fragment</span> <span class="pre">indices</span></code></a>
of all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in this
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p>A <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> with
<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.shape.html#numpy.ndarray.shape" title="(in NumPy v1.25)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">shape</span></code></a><code class="docutils literal notranslate"><span class="pre">=(</span></code><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.n_atoms" title="MDAnalysis.core.groups.AtomGroup.n_atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">n_atoms</span></code></a><code class="docutils literal notranslate"><span class="pre">,)</span></code>
and <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.dtype.html#numpy.ndarray.dtype" title="(in NumPy v1.25)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dtype</span></code></a><code class="docutils literal notranslate"><span class="pre">=numpy.int64</span></code>.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have bonds. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.fragments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fragments</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.fragments" title="Permalink to this definition"></a></dt>
<dd><p>Read-only <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a> of
<a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.Bonds.fragment" title="MDAnalysis.core.topologyattrs.Bonds.fragment"><code class="xref py py-class docutils literal notranslate"><span class="pre">fragments</span></code></a>.</p>
<p>Contains all fragments that
any <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> in this
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is part of.</p>
<p>A fragment is a
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">group</span> <span class="pre">of</span> <span class="pre">atoms</span></code></a> which are
interconnected by <a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.Bonds" title="MDAnalysis.core.topologyattrs.Bonds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bonds</span></code></a>, i.e.,
there exists a path along one
or more <a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.Bonds" title="MDAnalysis.core.topologyattrs.Bonds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bonds</span></code></a> between any pair
of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
within a fragment. Thus, a fragment typically corresponds to a molecule.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>The contents of the fragments may extend beyond the contents of this
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p></li>
</ul>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have bonds. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.get_connections">
<span class="sig-name descname"><span class="pre">get_connections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">typename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outside</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.get_connections" title="Permalink to this definition"></a></dt>
<dd><p>Get bonded connections between atoms as a
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.TopologyGroup" title="MDAnalysis.core.topologyobjects.TopologyGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyGroup</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>typename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a>) – group name. One of {“bonds”, “angles”, “dihedrals”,
“impropers”, “ureybradleys”, “cmaps”}</p></li>
<li><p><strong>outside</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Whether to include connections involving atoms outside
this group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>TopologyGroup</em> – containing the bonded group of choice, i.e. bonds, angles,
dihedrals, impropers, ureybradleys or cmaps.</p></li>
<li><p><em>.. versionadded:: 1.1.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.groupby">
<span class="sig-name descname"><span class="pre">groupby</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">topattrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.groupby" title="Permalink to this definition"></a></dt>
<dd><p>Group together items in this group according to values of <em>topattr</em></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>topattrs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.11)"><em>list</em></a>) – One or more topology attributes to group components by.
Single arguments are passed as a string. Multiple arguments
are passed as a list.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Unique values of the multiple combinations of topology attributes
as keys, Groups as values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.11)">dict</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>To group atoms with the same mass together:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;masses&#39;</span><span class="p">)</span>
<span class="go">{12.010999999999999: &lt;AtomGroup with 462 atoms&gt;,</span>
<span class="go"> 14.007: &lt;AtomGroup with 116 atoms&gt;,</span>
<span class="go"> 15.999000000000001: &lt;AtomGroup with 134 atoms&gt;}</span>
</pre></div>
</div>
<p>To group atoms with the same residue name and mass together:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;resnames&#39;</span><span class="p">,</span> <span class="s1">&#39;masses&#39;</span><span class="p">])</span>
<span class="go">{(&#39;ALA&#39;, 1.008): &lt;AtomGroup with 95 atoms&gt;,</span>
<span class="go"> (&#39;ALA&#39;, 12.011): &lt;AtomGroup with 57 atoms&gt;,</span>
<span class="go"> (&#39;ALA&#39;, 14.007): &lt;AtomGroup with 19 atoms&gt;,</span>
<span class="go"> (&#39;ALA&#39;, 15.999): &lt;AtomGroup with 19 atoms&gt;},</span>
<span class="go"> (&#39;ARG&#39;, 1.008): &lt;AtomGroup with 169 atoms&gt;,</span>
<span class="go"> ...}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;resnames&#39;</span><span class="p">,</span> <span class="s1">&#39;masses&#39;</span><span class="p">])(</span><span class="s1">&#39;ALA&#39;</span><span class="p">,</span> <span class="mf">15.999</span><span class="p">)</span>
<span class="go"> &lt;AtomGroup with 19 atoms&gt;</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.18.0: </span>The function accepts multiple attributes</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.guess_bonds">
<span class="sig-name descname"><span class="pre">guess_bonds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vdwradii</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fudge_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.55</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#AtomGroup.guess_bonds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.guess_bonds" title="Permalink to this definition"></a></dt>
<dd><p>Guess bonds, angles, and dihedrals between the atoms in this
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> and add them to the underlying
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.universe" title="MDAnalysis.core.groups.AtomGroup.universe"><code class="xref py py-attr docutils literal notranslate"><span class="pre">universe</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vdwradii</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.11)"><em>dict</em></a><em>, </em><em>optional</em>) – Dict relating atom types: vdw radii</p></li>
<li><p><strong>fudge_factor</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a><em>, </em><em>optional</em>) – The factor by which atoms must overlap each other to be considered
a bond.  Larger values will increase the number of bonds found. [0.55]</p></li>
<li><p><strong>lower_bound</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a><em>, </em><em>optional</em>) – The minimum bond length. All bonds found shorter than this length
will be ignored. This is useful for parsing PDB with altloc records
where atoms with altloc A and B may be very close together and
there should be no chemical bond between them. [0.1]</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="../topology/guessers.html#MDAnalysis.topology.guessers.guess_bonds" title="MDAnalysis.topology.guessers.guess_bonds"><code class="xref py py-func docutils literal notranslate"><span class="pre">MDAnalysis.topology.guessers.guess_bonds()</span></code></a>, <a class="reference internal" href="../topology/guessers.html#MDAnalysis.topology.guessers.guess_angles" title="MDAnalysis.topology.guessers.guess_angles"><code class="xref py py-func docutils literal notranslate"><span class="pre">MDAnalysis.topology.guessers.guess_angles()</span></code></a>, <a class="reference internal" href="../topology/guessers.html#MDAnalysis.topology.guessers.guess_dihedrals" title="MDAnalysis.topology.guessers.guess_dihedrals"><code class="xref py py-func docutils literal notranslate"><span class="pre">MDAnalysis.topology.guessers.guess_dihedrals()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.10.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.2: </span>Now applies periodic boundary conditions when guessing bonds.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.5.0: </span>Corrected misleading docs, and now allows passing of <cite>fudge_factor</cite>
and <cite>lower_bound</cite> arguments.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.gyration_moments">
<span class="sig-name descname"><span class="pre">gyration_moments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.gyration_moments" title="Permalink to this definition"></a></dt>
<dd><p>Moments of the gyration tensor.</p>
<p>The moments are defined as the eigenvalues of the gyration
tensor.</p>
<div class="math notranslate nohighlight">
\[\mathsf{T} = \frac{1}{N} \sum_{i=1}^{N} (\mathbf{r}_\mathrm{i} -
    \mathbf{r}_\mathrm{COM})(\mathbf{r}_\mathrm{i} - \mathbf{r}_\mathrm{COM})\]</div>
<p>Where <span class="math notranslate nohighlight">\(\mathbf{r}_\mathrm{COM}\)</span> is the center of mass.</p>
<p>See <a class="reference internal" href="../references.html#dima2004a" id="id2"><span>[Dima2004a]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>principle_moments_of_gyration</strong> – Gyration vector(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(Å^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
vector of length 3. Otherwise, the output will be a 2D array of shape
<code class="docutils literal notranslate"><span class="pre">(n,3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.5.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.improper">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">improper</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.improper" title="Permalink to this definition"></a></dt>
<dd><p>This <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> represented as an
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.ImproperDihedral" title="MDAnalysis.core.topologyobjects.ImproperDihedral"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.topologyobjects.ImproperDihedral</span></code></a> object</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is not length 4</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.11.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.intersection">
<span class="sig-name descname"><span class="pre">intersection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.intersection" title="Permalink to this definition"></a></dt>
<dd><p>Group of elements which are in both this Group and another</p>
<p>This method removes duplicate elements and sorts the result. It is
synomymous to the <cite>&amp;</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the common elements of <cite>self</cite> and <cite>other</cite>, without
duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Intersections can be used when the select atoms string would
become too complicated.  For example to find the water atoms
which are within 4.0A of two segments:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">shell1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;resname SOL and around 4.0 segid 1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">shell2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;resname SOL and around 4.0 segid 2&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">common</span> <span class="o">=</span> <span class="n">shell1</span> <span class="o">&amp;</span> <span class="n">shell2</span>  <span class="c1"># or shell1.intersection(shell2)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.union" title="MDAnalysis.core.groups.AtomGroup.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">union</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.is_strict_subset">
<span class="sig-name descname"><span class="pre">is_strict_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.is_strict_subset" title="Permalink to this definition"></a></dt>
<dd><p>If this Group is a subset of another Group but not identical</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a strict subset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.is_strict_superset">
<span class="sig-name descname"><span class="pre">is_strict_superset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.is_strict_superset" title="Permalink to this definition"></a></dt>
<dd><p>If this Group is a superset of another Group but not identical</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a strict superset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.isdisjoint">
<span class="sig-name descname"><span class="pre">isdisjoint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.isdisjoint" title="Permalink to this definition"></a></dt>
<dd><p>If the Group has no elements in common with the other Group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if the two Groups do not have common elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.issubset">
<span class="sig-name descname"><span class="pre">issubset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.issubset" title="Permalink to this definition"></a></dt>
<dd><p>If all elements of this Group are part of another Group</p>
<p>Note that an empty group is a subset of any group of the same level.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a subset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.issuperset">
<span class="sig-name descname"><span class="pre">issuperset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.issuperset" title="Permalink to this definition"></a></dt>
<dd><p>If all elements of another Group are part of this Group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a subset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.isunique">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">isunique</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.isunique" title="Permalink to this definition"></a></dt>
<dd><p>Boolean indicating whether all components of the group are unique,
i.e., the group contains no duplicates.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span>
<span class="go">&lt;AtomGroup with 6 atoms&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">isunique</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">unique</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span>
<span class="go">&lt;AtomGroup with 3 atoms&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span><span class="o">.</span><span class="n">isunique</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.asunique" title="MDAnalysis.core.groups.AtomGroup.asunique"><code class="xref py py-obj docutils literal notranslate"><span class="pre">asunique</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.19.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.ix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.ix" title="Permalink to this definition"></a></dt>
<dd><p>Unique indices of the components in the Group.</p>
<ul class="simple">
<li><p>If this Group is an <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>, these are the
indices of the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> instances.</p></li>
<li><p>If it is a <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a>, these are the indices of
the <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> instances.</p></li>
<li><p>If it is a <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>, these are the indices of
the <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> instances.</p></li>
</ul>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.ix_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix_array</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.ix_array" title="Permalink to this definition"></a></dt>
<dd><p>Unique indices of the components in the Group.</p>
<p>For a Group, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.ix_array" title="MDAnalysis.core.groups.AtomGroup.ix_array"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ix_array</span></code></a> is the same as <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.ix" title="MDAnalysis.core.groups.AtomGroup.ix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ix</span></code></a>. This method
gives a consistent API between components and groups.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.ix" title="MDAnalysis.core.groups.AtomGroup.ix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ix</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.moment_of_inertia">
<span class="sig-name descname"><span class="pre">moment_of_inertia</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.moment_of_inertia" title="Permalink to this definition"></a></dt>
<dd><p>Moment of inertia tensor relative to center of mass.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers and tensor of inertia.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
<cite>compound</cite> determines the behavior of <cite>wrap</cite>.
Note that, in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>moment_of_inertia</strong> – Moment of inertia tensor as a 3 x 3 numpy array.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The moment of inertia tensor <span class="math notranslate nohighlight">\(\mathsf{I}\)</span> is calculated for a group of
<span class="math notranslate nohighlight">\(N\)</span> atoms with coordinates <span class="math notranslate nohighlight">\(\mathbf{r}_i,\ 1 \le i \le N\)</span>
relative to its center of mass from the relative coordinates</p>
<div class="math notranslate nohighlight">
\[\mathbf{r}'_i = \mathbf{r}_i - \frac{1}{\sum_{i=1}^{N} m_i} \sum_{i=1}^{N} m_i \mathbf{r}_i\]</div>
<p>as</p>
<div class="math notranslate nohighlight">
\[\mathsf{I} = \sum_{i=1}^{N} m_i \Big[(\mathbf{r}'_i\cdot\mathbf{r}'_i) \sum_{\alpha=1}^{3}
      \hat{\mathbf{e}}_\alpha \otimes \hat{\mathbf{e}}_\alpha - \mathbf{r}'_i \otimes \mathbf{r}'_i\Big]\]</div>
<p>where <span class="math notranslate nohighlight">\(\hat{\mathbf{e}}_\alpha\)</span> are Cartesian unit vectors, or in Cartesian coordinates</p>
<div class="math notranslate nohighlight">
\[I_{\alpha,\beta} = \sum_{k=1}^{N} m_k
      \Big(\big(\sum_{\gamma=1}^3 (x'^{(k)}_{\gamma})^2 \big)\delta_{\alpha,\beta}
      - x'^{(k)}_{\alpha} x'^{(k)}_{\beta} \Big).\]</div>
<p>where <span class="math notranslate nohighlight">\(x'^{(k)}_{\alpha}\)</span> are the Cartesian coordinates of the
relative coordinates <span class="math notranslate nohighlight">\(\mathbf{r}'_k\)</span>.</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.n_atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.n_atoms" title="Permalink to this definition"></a></dt>
<dd><p>Number of atoms in the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p>Equivalent to <code class="docutils literal notranslate"><span class="pre">len(self)</span></code>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.n_fragments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_fragments</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.n_fragments" title="Permalink to this definition"></a></dt>
<dd><p>The number of unique
<a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.Bonds.fragments" title="MDAnalysis.core.topologyattrs.Bonds.fragments"><code class="xref py py-class docutils literal notranslate"><span class="pre">fragments</span></code></a> the
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> of this
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> are part of.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have bonds. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.n_residues">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_residues</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.n_residues" title="Permalink to this definition"></a></dt>
<dd><p>Number of unique <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residues</span></code></a> present in the
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p>Equivalent to <code class="docutils literal notranslate"><span class="pre">len(self.residues)</span></code>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.n_segments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_segments</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.n_segments" title="Permalink to this definition"></a></dt>
<dd><p>Number of unique segments present in the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p>Equivalent to <code class="docutils literal notranslate"><span class="pre">len(self.segments)</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.pack_into_box">
<span class="sig-name descname"><span class="pre">pack_into_box</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.pack_into_box" title="Permalink to this definition"></a></dt>
<dd><p>Shift all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in this group to the primary unit
cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>box</strong> (<em>array_like</em>) – Box dimensions, can be either orthogonal or triclinic information.
Cell dimensions must be in an identical to format to those returned
by <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.dimensions" title="MDAnalysis.coordinates.timestep.Timestep.dimensions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MDAnalysis.coordinates.timestep.Timestep.dimensions</span></code></a>,
<code class="docutils literal notranslate"><span class="pre">[lx,</span> <span class="pre">ly,</span> <span class="pre">lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, uses these
timestep dimensions.</p></li>
<li><p><strong>inplace</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a>) – <code class="docutils literal notranslate"><span class="pre">True</span></code> to change coordinates in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>coords</strong> – Shifted atom coordinates.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>All atoms will be moved so that they lie between 0 and boxlength
<span class="math notranslate nohighlight">\(L_i\)</span> in all dimensions, i.e. the lower left corner of the
simulation box is taken to be at (0,0,0):</p>
<div class="math notranslate nohighlight">
\[x_i' = x_i - \left\lfloor\frac{x_i}{L_i}\right\rfloor\]</div>
<p>The default is to take unit cell information from the underlying
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a> instance. The optional
argument <cite>box</cite> can be used to provide alternative unit cell information
(in the MDAnalysis standard format
<code class="docutils literal notranslate"><span class="pre">[Lx,</span> <span class="pre">Ly,</span> <span class="pre">Lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>).</p>
<p>Works with either orthogonal or triclinic box types.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.pack_into_box" title="MDAnalysis.core.groups.AtomGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pack_into_box()</span></code></a> is identical to <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.wrap" title="MDAnalysis.core.groups.AtomGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">wrap()</span></code></a> with all default
keywords.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.pack_into_box" title="MDAnalysis.core.groups.AtomGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AtomGroup.pack_into_box()</span></code></a> is currently faster than
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.pack_into_box" title="MDAnalysis.core.groups.ResidueGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ResidueGroup.pack_into_box()</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.pack_into_box" title="MDAnalysis.core.groups.SegmentGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SegmentGroup.pack_into_box()</span></code></a>.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.8.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.positions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">positions</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.positions" title="Permalink to this definition"></a></dt>
<dd><p>Coordinates of the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p>A <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> with
<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.shape.html#numpy.ndarray.shape" title="(in NumPy v1.25)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">shape</span></code></a><code class="docutils literal notranslate"><span class="pre">=(</span></code><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.n_atoms" title="MDAnalysis.core.groups.AtomGroup.n_atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">n_atoms</span></code></a><code class="docutils literal notranslate"><span class="pre">,</span> <span class="pre">3)</span></code>
and <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.dtype.html#numpy.ndarray.dtype" title="(in NumPy v1.25)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dtype</span></code></a><code class="docutils literal notranslate"><span class="pre">=numpy.float32</span></code>.</p>
<p>The positions can be changed by assigning an array of the appropriate
shape, i.e., either <code class="docutils literal notranslate"><span class="pre">(</span></code><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.n_atoms" title="MDAnalysis.core.groups.AtomGroup.n_atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">n_atoms</span></code></a><code class="docutils literal notranslate"><span class="pre">,</span> <span class="pre">3)</span></code> to
assign individual coordinates, or <code class="docutils literal notranslate"><span class="pre">(3,)</span></code> to assign the <em>same</em>
coordinate to all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> (e.g.,
<code class="docutils literal notranslate"><span class="pre">ag.positions</span> <span class="pre">=</span> <span class="pre">array([0,0,0])</span></code> will move all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
to the origin).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Changing positions is not reflected in any files; reading any
frame from the
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe.trajectory" title="MDAnalysis.core.universe.Universe.trajectory"><code class="xref py py-attr docutils literal notranslate"><span class="pre">trajectory</span></code></a> will
replace the change with that from the file <em>except</em> if the
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe.trajectory" title="MDAnalysis.core.universe.Universe.trajectory"><code class="xref py py-attr docutils literal notranslate"><span class="pre">trajectory</span></code></a> is held
in memory, e.g., when the
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe.transfer_to_memory" title="MDAnalysis.core.universe.Universe.transfer_to_memory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">transfer_to_memory()</span></code></a>
method was used.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If the underlying <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a>
    does not contain
    <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.positions" title="MDAnalysis.coordinates.timestep.Timestep.positions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">positions</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.principal_axes">
<span class="sig-name descname"><span class="pre">principal_axes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.principal_axes" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the principal axes from the moment of inertia.</p>
<p>e1,e2,e3 = AtomGroup.principal_axes()</p>
<p>The eigenvectors are sorted by eigenvalue, i.e. the first one
corresponds to the highest eigenvalue and is thus the first principal
axes.</p>
<p>The eigenvectors form a right-handed coordinate system.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>axis_vectors</strong> – 3 x 3 array with <code class="docutils literal notranslate"><span class="pre">v[0]</span></code> as first, <code class="docutils literal notranslate"><span class="pre">v[1]</span></code> as second, and
<code class="docutils literal notranslate"><span class="pre">v[2]</span></code> as third eigenvector.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Always return principal axes in right-hand convention.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.quadrupole_moment">
<span class="sig-name descname"><span class="pre">quadrupole_moment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.quadrupole_moment" title="Permalink to this definition"></a></dt>
<dd><p>Quadrupole moment of the group according to <span id="id3">[<a class="reference internal" href="topologyattrs.html#id2" title="C. G. Gray, Keith E. Gubbins, and C. G. Joslin. Theory of molecular fluids. Number 9 in The International series of monographs on chemistry. Oxford University Press, Oxford ; New York, 1984. ISBN 978-0-19-855602-2.">Gray1984</a>]</span>.</p>
<div class="math notranslate nohighlight">
\[Q = \sqrt{\frac{2}{3}{\hat{\mathsf{Q}}}:{\hat{\mathsf{Q}}}}\]</div>
<p>where the quadrupole moment is calculated from the tensor double
contraction of the traceless quadropole tensor <span class="math notranslate nohighlight">\(\hat{\mathsf{Q}}\)</span></p>
<p>Computes the quadrupole moment of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Quadrupole per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is an unsymmetrical plane in the molecule or
group, the magnitude of the quadrupole moment is dependant on the
<code class="docutils literal notranslate"><span class="pre">center</span></code> chosen and cannot be translated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single quadrupole value returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 1d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Quadrupole moment(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.quadrupole_tensor">
<span class="sig-name descname"><span class="pre">quadrupole_tensor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mass'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.quadrupole_tensor" title="Permalink to this definition"></a></dt>
<dd><p>Traceless quadrupole tensor of the group or compounds.</p>
<p>This tensor is first computed as the outer product of vectors from
a reference point to some atom, and multiplied by the atomic charge.
The tensor of each atom is then summed to produce the quadrupole
tensor of the group:</p>
<div class="math notranslate nohighlight">
\[\mathsf{Q} = \sum_{i=1}^{N} q_{i} ( \mathbf{r}_{i} -
\mathbf{r}_{COM} ) \otimes ( \mathbf{r}_{i} - \mathbf{r}_{COM} )\]</div>
<p>The traceless quadrupole tensor, <span class="math notranslate nohighlight">\(\hat{\mathsf{Q}}\)</span>, is then
taken from:</p>
<div class="math notranslate nohighlight">
\[\hat{\mathsf{Q}} = \frac{3}{2} \mathsf{Q} - \frac{1}{2}
tr(\mathsf{Q})\]</div>
<p>Computes the quadrupole tensor of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Tensor per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is an unsymmetrical plane in the molecule or
group, the magnitude of the quadrupole tensor is dependent on the
<code class="docutils literal notranslate"><span class="pre">center</span></code> (e.g., <span class="math notranslate nohighlight">\(\mathbf{r}_{COM}\)</span>) chosen and cannot be translated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single quadrupole value returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 1d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Quadrupole tensor(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
tensor of shape <code class="docutils literal notranslate"><span class="pre">(3,3)</span></code>. Otherwise, the output will be a 1d array
of shape <code class="docutils literal notranslate"><span class="pre">(n,3,3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.radius_of_gyration">
<span class="sig-name descname"><span class="pre">radius_of_gyration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.radius_of_gyration" title="Permalink to this definition"></a></dt>
<dd><p>Radius of gyration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.residues">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">residues</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.residues" title="Permalink to this definition"></a></dt>
<dd><p>A sorted <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> of the unique
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residues</span></code></a> present in the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.rotate">
<span class="sig-name descname"><span class="pre">rotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">R</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">0,</span> <span class="pre">0)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.rotate" title="Permalink to this definition"></a></dt>
<dd><p>Apply a rotation matrix <cite>R</cite> to the selection’s coordinates.
<span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is a 3x3 orthogonal matrix that transforms a vector
<span class="math notranslate nohighlight">\(\mathbf{x} \rightarrow \mathbf{x}'\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathsf{R}\mathbf{x}\]</div>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> coordinates are rotated in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>R</strong> (<em>array_like</em>) – 3x3 rotation matrix</p></li>
<li><p><strong>point</strong> (<em>array_like</em><em>, </em><em>optional</em>) – Center of rotation</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>By default, rotates about the origin <code class="docutils literal notranslate"><span class="pre">point=(0,</span> <span class="pre">0,</span> <span class="pre">0)</span></code>. To rotate
a group <code class="docutils literal notranslate"><span class="pre">g</span></code> around its center of geometry, use
<code class="docutils literal notranslate"><span class="pre">g.rotate(R,</span> <span class="pre">point=g.center_of_geometry())</span></code>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.rotateby" title="MDAnalysis.core.groups.AtomGroup.rotateby"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotateby</span></code></a></dt><dd><p>rotate around given axis and angle</p>
</dd>
<dt><a class="reference internal" href="../lib/transformations.html#module-MDAnalysis.lib.transformations" title="MDAnalysis.lib.transformations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations</span></code></a></dt><dd><p>module of all coordinate transforms</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.rotateby">
<span class="sig-name descname"><span class="pre">rotateby</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">angle</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.rotateby" title="Permalink to this definition"></a></dt>
<dd><p>Apply a rotation to the selection’s coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>angle</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a>) – Rotation angle in degrees.</p></li>
<li><p><strong>axis</strong> (<em>array_like</em>) – Rotation axis vector.</p></li>
<li><p><strong>point</strong> (<em>array_like</em><em>, </em><em>optional</em>) – Center of rotation. If <code class="docutils literal notranslate"><span class="pre">None</span></code> then the center of geometry of this
group is used.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The transformation from current coordinates <span class="math notranslate nohighlight">\(\mathbf{x}\)</span>
to new coordinates <span class="math notranslate nohighlight">\(\mathbf{x}'\)</span> is</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathsf{R}\,(\mathbf{x}-\mathbf{p}) + \mathbf{p}\]</div>
<p>where <span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is the rotation by <cite>angle</cite> around the
<cite>axis</cite> going through <cite>point</cite> <span class="math notranslate nohighlight">\(\mathbf{p}\)</span>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations.rotation_matrix</span></code></p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">calculate</span></code></dt><dd><p>math:<cite>mathsf{R}</cite></p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.segments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">segments</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.segments" title="Permalink to this definition"></a></dt>
<dd><p>A sorted <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> of the unique segments present in the
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.select_atoms">
<span class="sig-name descname"><span class="pre">select_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sel</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">othersel</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">periodic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rtol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-08</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">updating</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sorted</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rdkit_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smarts_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">selgroups</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#AtomGroup.select_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.select_atoms" title="Permalink to this definition"></a></dt>
<dd><p>Select atoms from within this Group using a selection string.</p>
<p>Returns an <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> sorted according to their index in the
topology (this is to ensure that there are no duplicates, which can
happen with complicated selections).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sel</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a>) – string of the selection, eg “name Ca”, see below for possibilities.</p></li>
<li><p><strong>othersel</strong> (<em>iterable of str</em>) – further selections to perform.  The results of these selections
will be appended onto the results of the first.</p></li>
<li><p><strong>periodic</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – for geometric selections, whether to account for atoms in different
periodic images when searching</p></li>
<li><p><strong>atol</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a><em>, </em><em>optional</em>) – The absolute tolerance parameter for float comparisons.
Passed to :func:<code class="docutils literal notranslate"><span class="pre">numpy.isclose</span></code>.</p></li>
<li><p><strong>rtol</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a><em>, </em><em>optional</em>) – The relative tolerance parameter for float comparisons.
Passed to :func:<code class="docutils literal notranslate"><span class="pre">numpy.isclose</span></code>.</p></li>
<li><p><strong>updating</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – force the selection to be re evaluated each time the Timestep of the
trajectory is changed.  See section on <strong>Dynamic selections</strong> below.
[<code class="docutils literal notranslate"><span class="pre">True</span></code>]</p></li>
<li><p><strong>sorted</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to sort the output AtomGroup by index.</p></li>
<li><p><strong>rdkit_kwargs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.11)"><em>dict</em></a><em> (</em><em>optional</em><em>)</em>) – Arguments passed to the
<a class="reference internal" href="../converters/RDKit.html#MDAnalysis.converters.RDKit.RDKitConverter" title="MDAnalysis.converters.RDKit.RDKitConverter"><code class="xref py py-class docutils literal notranslate"><span class="pre">RDKitConverter</span></code></a> when using
selection based on SMARTS queries</p></li>
<li><p><strong>smarts_kwargs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.11)"><em>dict</em></a><em> (</em><em>optional</em><em>)</em>) – Arguments passed internally to RDKit’s <a class="reference external" href="https://www.rdkit.org/docs/source/rdkit.Chem.rdchem.html#rdkit.Chem.rdchem.Mol.GetSubstructMatches">GetSubstructMatches</a>.</p></li>
<li><p><strong>**selgroups</strong> (<em>keyword arguments of str: AtomGroup</em><em> (</em><em>optional</em><em>)</em>) – when using the “group” keyword in selections, groups are defined by
passing them as keyword arguments.  See section on <strong>preexisting
selections</strong> below.</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.11)"><strong>TypeError</strong></a> – If the arbitrary groups passed are not of type
    <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup</span></code></a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>All simple selection listed below support multiple arguments which are
implicitly combined with an or operator. For example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">universe</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;resname MET GLY&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>is equivalent to</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">universe</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;resname MET or resname GLY&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Will select all atoms with a residue name of either MET or GLY.</p>
<p>Subselections can be grouped with parentheses.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">universe</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s2">&quot;segid DMPC and not ( name H* O* )&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span>
<span class="go">&lt;AtomGroup with 3420 atoms&gt;</span>
</pre></div>
</div>
<p>Existing <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> objects can be passed as named arguments,
which will then be available to the selection parser.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">universe</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s2">&quot;around 10 group notHO&quot;</span><span class="p">,</span> <span class="n">notHO</span><span class="o">=</span><span class="n">sel</span><span class="p">)</span>
<span class="go">&lt;AtomGroup with 1250 atoms&gt;</span>
</pre></div>
</div>
<p>Selections can be set to update automatically on frame change, by
setting the <cite>updating</cite> keyword argument to <cite>True</cite>.  This will return
a <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a> which can represent the solvation shell
around another object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">universe</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s2">&quot;resname SOL and around 2.0 protein&quot;</span><span class="p">,</span> <span class="n">updating</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">&lt;Updating AtomGroup with 100 atoms&gt;</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>If exact ordering of atoms is required (for instance, for
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.angle" title="MDAnalysis.core.groups.AtomGroup.angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">angle()</span></code></a> or <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.dihedral" title="MDAnalysis.core.groups.AtomGroup.dihedral"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dihedral()</span></code></a> calculations)
then one supplies selections <em>separately</em> in the required order. Also,
when multiple <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> instances are concatenated with the
<code class="docutils literal notranslate"><span class="pre">+</span></code> operator, then the order of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> instances is preserved
and duplicates are <em>not</em> removed.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="../selections.html#selection-commands-label"><span class="std std-ref">Selection commands</span></a></p>
</div>
<p class="rubric">Selection syntax</p>
<p>The selection parser understands the following CASE SENSITIVE
<em>keywords</em>:</p>
<p><strong>Simple selections</strong></p>
<blockquote>
<div><dl>
<dt>protein, backbone, nucleic, nucleicbackbone</dt><dd><p>selects all atoms that belong to a standard set of residues;
a protein is identfied by a hard-coded set of residue names so
it  may not work for esoteric residues.</p>
</dd>
<dt>segid <em>seg-name</em></dt><dd><p>select by segid (as given in the topology), e.g. <code class="docutils literal notranslate"><span class="pre">segid</span> <span class="pre">4AKE</span></code>
or <code class="docutils literal notranslate"><span class="pre">segid</span> <span class="pre">DMPC</span></code></p>
</dd>
<dt>resid <em>residue-number-range</em></dt><dd><p>resid can take a single residue number or a range of numbers. A
range consists of two numbers separated by a colon (inclusive)
such as <code class="docutils literal notranslate"><span class="pre">resid</span> <span class="pre">1:5</span></code>. A residue number (“resid”) is taken
directly from the topology.
If icodes are present in the topology, then these will be
taken into account.  Ie ‘resid 163B’ will only select resid
163 with icode B while ‘resid 163’ will select only residue 163.
Range selections will also respect icodes, so ‘resid 162-163B’
will select all residues in 162 and those in 163 up to icode B.</p>
</dd>
<dt>resnum <em>resnum-number-range</em></dt><dd><p>resnum is the canonical residue number; typically it is set to
the residue id in the original PDB structure.</p>
</dd>
<dt>resname <em>residue-name</em></dt><dd><p>select by residue name, e.g. <code class="docutils literal notranslate"><span class="pre">resname</span> <span class="pre">LYS</span></code></p>
</dd>
<dt>name <em>atom-name</em></dt><dd><p>select by atom name (as given in the topology). Often, this is
force field dependent. Example: <code class="docutils literal notranslate"><span class="pre">name</span> <span class="pre">CA</span></code> (for C&amp;alpha; atoms)
or <code class="docutils literal notranslate"><span class="pre">name</span> <span class="pre">OW</span></code> (for SPC water oxygen)</p>
</dd>
<dt>type <em>atom-type</em></dt><dd><p>select by atom type; this is either a string or a number and
depends on the force field; it is read from the topology file
(e.g. the CHARMM PSF file contains numeric atom types). It has
non-sensical values when a PDB or GRO file is used as a topology</p>
</dd>
<dt>atom <em>seg-name</em>  <em>residue-number</em>  <em>atom-name</em></dt><dd><p>a selector for a single atom consisting of segid resid atomname,
e.g. <code class="docutils literal notranslate"><span class="pre">DMPC</span> <span class="pre">1</span> <span class="pre">C2</span></code> selects the C2 carbon of the first residue of
the DMPC segment</p>
</dd>
<dt>altloc <em>alternative-location</em></dt><dd><p>a selection for atoms where alternative locations are available,
which is often the case with high-resolution crystal structures
e.g. <cite>resid 4 and resname ALA and altloc B</cite> selects only the
atoms of ALA-4 that have an altloc B record.</p>
</dd>
<dt>moltype <em>molecule-type</em></dt><dd><p>select by molecule type, e.g. <code class="docutils literal notranslate"><span class="pre">moltype</span> <span class="pre">Protein_A</span></code>. At the
moment, only the TPR format defines the molecule type.</p>
</dd>
<dt>record_type <em>record_type</em></dt><dd><p>for selecting either ATOM or HETATM from PDB-like files.
e.g. <code class="docutils literal notranslate"><span class="pre">select_atoms('name</span> <span class="pre">CA</span> <span class="pre">and</span> <span class="pre">not</span> <span class="pre">record_type</span> <span class="pre">HETATM')</span></code></p>
</dd>
<dt>smarts <em>SMARTS-query</em></dt><dd><p>select atoms using Daylight’s SMARTS queries, e.g. <code class="docutils literal notranslate"><span class="pre">smarts</span>
<span class="pre">[#7;R]</span></code> to find nitrogen atoms in rings. Requires RDKit.
All matches are combined as a single unique match. The <cite>smarts</cite>
selection accepts two sets of key word arguments from
<cite>select_atoms()</cite>: the <code class="docutils literal notranslate"><span class="pre">rdkit_kwargs</span></code> are passed internally to
<cite>RDKitConverter.convert()</cite> and the <code class="docutils literal notranslate"><span class="pre">smarts_kwargs</span></code> are passed to
RDKit’s <a class="reference external" href="https://www.rdkit.org/docs/source/rdkit.Chem.rdchem.html#rdkit.Chem.rdchem.Mol.GetSubstructMatches">GetSubstructMatches</a>.
By default, the <cite>useChirality</cite> kwarg in <code class="docutils literal notranslate"><span class="pre">rdkit_kwargs</span></code> is set to true
and maxMatches in <code class="docutils literal notranslate"><span class="pre">smarts_kwargs</span></code> is
<code class="docutils literal notranslate"><span class="pre">max(1000,</span> <span class="pre">10</span> <span class="pre">*</span> <span class="pre">n_atoms)</span></code>, where <code class="docutils literal notranslate"><span class="pre">n_atoms</span></code> is either
<code class="docutils literal notranslate"><span class="pre">len(AtomGroup)</span></code> or <code class="docutils literal notranslate"><span class="pre">len(Universe.atoms)</span></code>, whichever is
applicable. Note that the number of matches can occasionally
exceed the default value of maxMatches, causing too few atoms
to be returned. If this occurs, a warning will be issued. The
problem can be fixed by increasing the value of maxMatches.
This behavior may be updated in the future.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">universe</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">smarts_kwargs</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;maxMatches&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">})</span>
<span class="go">&lt;AtomGroup with 100 atoms&gt;</span>
</pre></div>
</div>
</dd>
<dt>chiral <em>R | S</em></dt><dd><p>select a particular stereocenter. e.g. <code class="docutils literal notranslate"><span class="pre">name</span> <span class="pre">C</span> <span class="pre">and</span> <span class="pre">chirality</span>
<span class="pre">S</span></code> to select only S-chiral carbon atoms.  Only <code class="docutils literal notranslate"><span class="pre">R</span></code> and
<code class="docutils literal notranslate"><span class="pre">S</span></code> will be possible options but other values will not raise
an error.</p>
</dd>
<dt>formalcharge <em>formal-charge</em></dt><dd><p>select atoms based on their formal charge, e.g.
<code class="docutils literal notranslate"><span class="pre">name</span> <span class="pre">O</span> <span class="pre">and</span> <span class="pre">formalcharge</span> <span class="pre">-1</span></code> to select all oxygens with a
negative 1 formal charge.</p>
</dd>
</dl>
</div></blockquote>
<p><strong>Boolean</strong></p>
<blockquote>
<div><dl class="simple">
<dt>not</dt><dd><p>all atoms not in the selection, e.g. <code class="docutils literal notranslate"><span class="pre">not</span> <span class="pre">protein</span></code> selects
all atoms that aren’t part of a protein</p>
</dd>
<dt>and, or</dt><dd><p>combine two selections according to the rules of boolean
algebra, e.g. <code class="docutils literal notranslate"><span class="pre">protein</span> <span class="pre">and</span> <span class="pre">not</span> <span class="pre">resname</span> <span class="pre">ALA</span> <span class="pre">LYS</span></code>
selects all atoms that belong to a protein, but are not in a
lysine or alanine residue</p>
</dd>
</dl>
</div></blockquote>
<p><strong>Geometric</strong></p>
<blockquote>
<div><dl class="simple">
<dt>around <em>distance</em>  <em>selection</em></dt><dd><p>selects all atoms a certain cutoff away from another selection,
e.g. <code class="docutils literal notranslate"><span class="pre">around</span> <span class="pre">3.5</span> <span class="pre">protein</span></code> selects all atoms not belonging to
protein that are within 3.5 Angstroms from the protein</p>
</dd>
<dt>point <em>x</em> <em>y</em> <em>z</em>  <em>distance</em></dt><dd><p>selects all atoms within a cutoff of a point in space, make sure
coordinate is separated by spaces,
e.g. <code class="docutils literal notranslate"><span class="pre">point</span> <span class="pre">5.0</span> <span class="pre">5.0</span> <span class="pre">5.0</span>&#160; <span class="pre">3.5</span></code> selects all atoms within 3.5
Angstroms of the coordinate (5.0, 5.0, 5.0)</p>
</dd>
<dt>prop [abs] <em>property</em>  <em>operator</em>  <em>value</em></dt><dd><p>selects atoms based on position, using <em>property</em>  <strong>x</strong>, <strong>y</strong>,
or <strong>z</strong> coordinate. Supports the <strong>abs</strong> keyword (for absolute
value) and the following <em>operators</em>: <strong>&lt;, &gt;, &lt;=, &gt;=, ==, !=</strong>.
For example, <code class="docutils literal notranslate"><span class="pre">prop</span> <span class="pre">z</span> <span class="pre">&gt;=</span> <span class="pre">5.0</span></code> selects all atoms with z
coordinate greater than 5.0; <code class="docutils literal notranslate"><span class="pre">prop</span> <span class="pre">abs</span> <span class="pre">z</span> <span class="pre">&lt;=</span> <span class="pre">5.0</span></code> selects all
atoms within -5.0 &lt;= z &lt;= 5.0.</p>
</dd>
<dt>sphzone <em>radius</em> <em>selection</em></dt><dd><p>Selects all atoms that are within <em>radius</em> of the center of
geometry of <em>selection</em></p>
</dd>
<dt>sphlayer <em>inner radius</em> <em>outer radius</em> <em>selection</em></dt><dd><p>Similar to sphzone, but also excludes atoms that are within
<em>inner radius</em> of the selection COG</p>
</dd>
<dt>isolayer <em>inner radius</em> <em>outer radius</em> <em>selection</em></dt><dd><p>Similar to sphlayer, but will find layer around all reference
layer, creating an iso-surface.</p>
</dd>
<dt>cyzone <em>externalRadius</em> <em>zMax</em> <em>zMin</em> <em>selection</em></dt><dd><p>selects all atoms within a cylindric zone centered in the
center of geometry (COG) of a given selection,
e.g. <code class="docutils literal notranslate"><span class="pre">cyzone</span> <span class="pre">15</span> <span class="pre">4</span> <span class="pre">-8</span> <span class="pre">protein</span> <span class="pre">and</span> <span class="pre">resid</span> <span class="pre">42</span></code> selects the
center of geometry of protein and resid 42, and creates a
cylinder of external radius 15 centered on the COG. In z, the
cylinder extends from 4 above the COG to 8 below. Positive
values for <em>zMin</em>, or negative ones for <em>zMax</em>, are allowed.</p>
</dd>
<dt>cylayer <em>innerRadius</em> <em>externalRadius</em> <em>zMax</em> <em>zMin</em> <em>selection</em></dt><dd><p>selects all atoms within a cylindric layer centered in the
center of geometry (COG) of a given selection,
e.g. <code class="docutils literal notranslate"><span class="pre">cylayer</span> <span class="pre">5</span> <span class="pre">10</span> <span class="pre">10</span> <span class="pre">-8</span> <span class="pre">protein</span></code> selects the center of
geometry of protein, and creates a cylindrical layer of inner
radius 5, external radius 10 centered on the COG. In z, the
cylinder extends from 10 above the COG to 8 below. Positive
values for <em>zMin</em>, or negative ones for <em>zMax</em>, are allowed.</p>
</dd>
</dl>
</div></blockquote>
<p><strong>Connectivity</strong></p>
<blockquote>
<div><dl class="simple">
<dt>byres <em>selection</em></dt><dd><p>selects all atoms that are in the same segment and residue as
selection, e.g. specify the subselection after the byres keyword</p>
</dd>
<dt>bonded <em>selection</em></dt><dd><p>selects all atoms that are bonded to selection
eg: <code class="docutils literal notranslate"><span class="pre">select</span> <span class="pre">name</span> <span class="pre">H</span> <span class="pre">and</span> <span class="pre">bonded</span> <span class="pre">name</span> <span class="pre">O</span></code> selects only hydrogens
bonded to oxygens</p>
</dd>
</dl>
</div></blockquote>
<p><strong>Index</strong></p>
<blockquote>
<div><dl class="simple">
<dt>bynum <em>index-range</em></dt><dd><p>selects all atoms within a range of (1-based) inclusive indices,
e.g. <code class="docutils literal notranslate"><span class="pre">bynum</span> <span class="pre">1</span></code> selects the first atom in the universe;
<code class="docutils literal notranslate"><span class="pre">bynum</span> <span class="pre">5:10</span></code> selects atoms 5 through 10 inclusive. All atoms
in the <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> are
consecutively numbered, and the index runs from 1 up to the
total number of atoms.</p>
</dd>
<dt>index <em>index-range</em></dt><dd><p>selects all atoms within a range of (0-based) inclusive indices,
e.g. <code class="docutils literal notranslate"><span class="pre">index</span> <span class="pre">0</span></code> selects the first atom in the universe;
<code class="docutils literal notranslate"><span class="pre">index</span> <span class="pre">5:10</span></code> selects atoms 6 through 11 inclusive. All atoms
in the <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> are
consecutively numbered, and the index runs from 0 up to the
total number of atoms - 1.</p>
</dd>
</dl>
</div></blockquote>
<p><strong>Preexisting selections</strong></p>
<blockquote>
<div><dl class="simple">
<dt>group <cite>group-name</cite></dt><dd><p>selects the atoms in the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> passed to the
function as a keyword argument named <cite>group-name</cite>. Only the
atoms common to <cite>group-name</cite> and the instance
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.select_atoms" title="MDAnalysis.core.groups.AtomGroup.select_atoms"><code class="xref py py-meth docutils literal notranslate"><span class="pre">select_atoms()</span></code></a>
was called from will be considered, unless <code class="docutils literal notranslate"><span class="pre">group</span></code> is
preceded by the <code class="docutils literal notranslate"><span class="pre">global</span></code> keyword. <cite>group-name</cite> will be
included in the parsing just by comparison of atom indices.
This means that it is up to the user to make sure the
<cite>group-name</cite> group was defined in an appropriate
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a>.</p>
</dd>
<dt>global <em>selection</em></dt><dd><p>by default, when issuing
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.select_atoms" title="MDAnalysis.core.groups.AtomGroup.select_atoms"><code class="xref py py-meth docutils literal notranslate"><span class="pre">select_atoms()</span></code></a> from an
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>, selections and
subselections are returned intersected with the atoms of that
instance. Prefixing a selection term with <code class="docutils literal notranslate"><span class="pre">global</span></code> causes its
selection to be returned in its entirety.  As an example, the
<code class="docutils literal notranslate"><span class="pre">global</span></code> keyword allows for
<code class="docutils literal notranslate"><span class="pre">lipids.select_atoms(&quot;around</span> <span class="pre">10</span> <span class="pre">global</span> <span class="pre">protein&quot;)</span></code> — where
<code class="docutils literal notranslate"><span class="pre">lipids</span></code> is a group that does not contain any proteins. Were
<code class="docutils literal notranslate"><span class="pre">global</span></code> absent, the result would be an empty selection since
the <code class="docutils literal notranslate"><span class="pre">protein</span></code> subselection would itself be empty. When issuing
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.select_atoms" title="MDAnalysis.core.groups.AtomGroup.select_atoms"><code class="xref py py-meth docutils literal notranslate"><span class="pre">select_atoms()</span></code></a> from a
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a>, <code class="docutils literal notranslate"><span class="pre">global</span></code> is
ignored.</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt><strong>Dynamic selections</strong></dt><dd><p>If <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.select_atoms" title="MDAnalysis.core.groups.AtomGroup.select_atoms"><code class="xref py py-meth docutils literal notranslate"><span class="pre">select_atoms()</span></code></a> is
invoked with named argument <cite>updating</cite> set to <cite>True</cite>, an
<a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a> instance will be
returned, instead of a regular
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>. It behaves just like
the latter, with the difference that the selection expressions are
re-evaluated every time the trajectory frame changes (this happens
lazily, only when the
<a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a> is accessed so
that there is no redundant updating going on).
Issuing an updating selection from an already updating group will
cause later updates to also reflect the updating of the base group.
A non-updating selection or a slicing operation made on an
<a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a> will return a
static <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>, which will no
longer update across frames.</p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.7.4: </span>Added <em>resnum</em> selection.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8.1: </span>Added <em>group</em> and <em>fullgroup</em> selections.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.13.0: </span>Added <em>bonded</em> selection.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.16.0: </span>Resid selection now takes icodes into account
where present.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.16.0: </span>Updating selections now possible by setting
the <cite>updating</cite> argument.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.17.0: </span>Added <em>moltype</em> and <em>molnum</em> selections.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added strict type checking for passed groups.
Added periodic kwarg (default True)</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.2: </span>Empty sel string now returns an empty Atom group.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>The <code class="docutils literal notranslate"><span class="pre">fullgroup</span></code> selection has now been removed in favor of the
equivalent <code class="docutils literal notranslate"><span class="pre">global</span> <span class="pre">group</span></code> selection.
Removed flags affecting default behaviour for periodic selections;
periodic are now on by default (as with default flags)</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.0.0: </span>Added the <em>smarts</em> selection. Added <cite>atol</cite> and <cite>rtol</cite> keywords
to select float values. Added the <code class="docutils literal notranslate"><span class="pre">sort</span></code> keyword. Added
<cite>rdkit_kwargs</cite> to pass parameters to the RDKitConverter.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.2.0: </span>Added <cite>smarts_kwargs</cite> to pass parameters to the RDKit
GetSubstructMatch for <em>smarts</em> selection.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.shape_parameter">
<span class="sig-name descname"><span class="pre">shape_parameter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.shape_parameter" title="Permalink to this definition"></a></dt>
<dd><p>Shape parameter.</p>
<p>See <a class="reference internal" href="../references.html#dima2004a" id="id5"><span>[Dima2004a]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.
Superfluous kwargs were removed.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.5.0: </span>Added calculation for any <cite>compound</cite> type</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.sort">
<span class="sig-name descname"><span class="pre">sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ix'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keyfunc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#AtomGroup.sort"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.sort" title="Permalink to this definition"></a></dt>
<dd><p>Returns a sorted <code class="docutils literal notranslate"><span class="pre">AtomGroup</span></code> using a specified attribute as the key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em>, </em><em>optional</em>) – The name of the <code class="docutils literal notranslate"><span class="pre">AtomGroup</span></code> attribute to sort by (e.g. <code class="docutils literal notranslate"><span class="pre">ids</span></code>,
<code class="docutils literal notranslate"><span class="pre">ix</span></code>. default= <code class="docutils literal notranslate"><span class="pre">ix</span></code> ).</p></li>
<li><p><strong>keyfunc</strong> (<em>callable</em><em>, </em><em>optional</em>) – A function to convert multidimensional arrays to a single
dimension. This 1D array will be used as the sort key and
is required when sorting with an <code class="docutils literal notranslate"><span class="pre">AtomGroup</span></code> attribute
key which has multiple dimensions. Note: this argument
is ignored when the attribute is one dimensional.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Sorted <code class="docutils literal notranslate"><span class="pre">AtomGroup</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">MDAnalysis</span> <span class="k">as</span> <span class="nn">mda</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysisTests.datafiles</span> <span class="kn">import</span> <span class="n">PDB_small</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PDB_small</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([3 2 1 0])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([0 1 2 3])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">positions</span>
<span class="go">array([[-11.921,  26.307,  10.41 ],</span>
<span class="go">       [-11.447,  26.741,   9.595],</span>
<span class="go">       [-12.44 ,  27.042,  10.926],</span>
<span class="go">       [-12.632,  25.619,  10.046]], dtype=float32)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="s2">&quot;positions&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">positions</span>
<span class="go">array([[-12.632,  25.619,  10.046],</span>
<span class="go">       [-11.921,  26.307,  10.41 ],</span>
<span class="go">       [-11.447,  26.741,   9.595],</span>
<span class="go">       [-12.44 ,  27.042,  10.926]], dtype=float32)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This uses a stable sort as implemented by
<cite>numpy.argsort(kind=’stable’)</cite>.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.0.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.split">
<span class="sig-name descname"><span class="pre">split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#AtomGroup.split"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.split" title="Permalink to this definition"></a></dt>
<dd><p>Split <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> into a <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a> of
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroups</span></code></a> by <cite>level</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>level</strong> (<em>{'atom'</em><em>, </em><em>'residue'</em><em>, </em><em>'molecule'</em><em>, </em><em>'segment'}</em>) – </p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.17.0: </span>Added the ‘molecule’ level.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.subtract">
<span class="sig-name descname"><span class="pre">subtract</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.subtract" title="Permalink to this definition"></a></dt>
<dd><p>Group with elements from this Group that don’t appear in other</p>
<p>The original order of this group is kept, as well as any duplicate
elements. If an element of this Group is duplicated and appears in
the other Group or Component, then all the occurences of that element
are removed from the returned Group.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the elements of <cite>self</cite> that are not in  <cite>other</cite>,
conserves order and duplicates.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Unlike <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.difference" title="MDAnalysis.core.groups.AtomGroup.difference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">difference()</span></code></a> this method will not sort or remove
duplicates.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">-</span> <span class="n">ag2</span>  <span class="c1"># or ag1.subtract(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span><span class="o">.</span><span class="n">indices</span>
<span class="go">array([3, 3, 1, 1])</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.concatenate" title="MDAnalysis.core.groups.AtomGroup.concatenate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concatenate</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.difference" title="MDAnalysis.core.groups.AtomGroup.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.symmetric_difference">
<span class="sig-name descname"><span class="pre">symmetric_difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.symmetric_difference" title="Permalink to this definition"></a></dt>
<dd><p>Group of elements which are only in one of this Group or another</p>
<p>This method removes duplicate elements and the result is sorted. It is
synomym to the <cite>^</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the elements that are in <cite>self</cite> or in <cite>other</cite> but not in
both, without duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">^</span> <span class="n">ag2</span>  <span class="c1"># or ag1.symmetric_difference(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="o">.</span><span class="n">indices</span>  <span class="c1"># 0 and 1 are only in ag1, 4 and 6 are only in ag2</span>
<span class="go">[0, 1, 4, 6]</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.difference" title="MDAnalysis.core.groups.AtomGroup.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.total_charge">
<span class="sig-name descname"><span class="pre">total_charge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.total_charge" title="Permalink to this definition"></a></dt>
<dd><p>Total charge of (compounds of) the group.</p>
<p>Computes the total charge of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Total charges per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If ‘group’, the total charge of all atoms in the group will
be returned as a single value. Otherwise, the total charges per
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment
will be returned as a 1d array.
Note that, in any case, <em>only</em> the charges of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Total charge of (compounds of) the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.total_mass">
<span class="sig-name descname"><span class="pre">total_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.total_mass" title="Permalink to this definition"></a></dt>
<dd><p>Total mass of (compounds of) the group.</p>
<p>Computes the total mass of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Total masses per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the total mass of all atoms in the group will be
returned as a single value. Otherwise, the total masses per
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment will be
returned as a 1d array.
Note that, in any case, <em>only</em> the masses of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Total mass of (compounds of) the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.transform">
<span class="sig-name descname"><span class="pre">transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">M</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.transform" title="Permalink to this definition"></a></dt>
<dd><p>Apply homogenous transformation matrix <cite>M</cite> to the coordinates.</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> coordinates are rotated and translated in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>M</strong> (<em>array_like</em>) – 4x4 matrix with the rotation in <code class="docutils literal notranslate"><span class="pre">R</span> <span class="pre">=</span> <span class="pre">M[:3,</span> <span class="pre">:3]</span></code> and the
translation in <code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">M[:3,</span> <span class="pre">3]</span></code>.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../lib/transformations.html#module-MDAnalysis.lib.transformations" title="MDAnalysis.lib.transformations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations</span></code></a></dt><dd><p>module of all coordinate transforms</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The rotation <span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is about the origin and is applied
before the translation <span class="math notranslate nohighlight">\(\mathbf{t}\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathsf{R}\mathbf{x} + \mathbf{t}\]</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.translate" title="Permalink to this definition"></a></dt>
<dd><p>Apply translation vector <cite>t</cite> to the selection’s coordinates.</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> coordinates are translated in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>t</strong> (<em>array_like</em>) – vector to translate coordinates with</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../lib/transformations.html#module-MDAnalysis.lib.transformations" title="MDAnalysis.lib.transformations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations</span></code></a></dt><dd><p>module of all coordinate transforms</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The method applies a translation to the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>
from current coordinates <span class="math notranslate nohighlight">\(\mathbf{x}\)</span> to new coordinates
<span class="math notranslate nohighlight">\(\mathbf{x}'\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathbf{x} + \mathbf{t}\]</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.ts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ts</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.ts" title="Permalink to this definition"></a></dt>
<dd><p>Temporary Timestep that contains the selection coordinates.</p>
<p>A <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a> instance,
which can be passed to a trajectory writer.</p>
<p>If <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.ts" title="MDAnalysis.core.groups.AtomGroup.ts"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ts</span></code></a> is modified then these modifications
will be present until the frame number changes (which
typically happens when the underlying
<a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe.trajectory" title="MDAnalysis.core.universe.Universe.trajectory"><code class="xref py py-attr docutils literal notranslate"><span class="pre">trajectory</span></code></a> frame changes).</p>
<p>It is not possible to assign a new
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a> to the
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.ts" title="MDAnalysis.core.groups.AtomGroup.ts"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AtomGroup.ts</span></code></a> attribute; change attributes of the object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.union">
<span class="sig-name descname"><span class="pre">union</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.union" title="Permalink to this definition"></a></dt>
<dd><p>Group of elements either in this Group or another</p>
<p>On the contrary to concatenation, this method sort the elements and
removes duplicate ones. It is synomymous to the <cite>|</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the combined elements of <cite>self</cite> and <cite>other</cite>, without
duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>In contrast to <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.concatenate" title="MDAnalysis.core.groups.AtomGroup.concatenate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">concatenate()</span></code></a>, any duplicates are dropped
and the result is sorted.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name O&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name N&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">|</span> <span class="n">ag2</span>  <span class="c1"># or ag1.union(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">names</span>
<span class="go">array([&#39;N&#39;, &#39;O&#39;, &#39;N&#39;], dtype=object)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.concatenate" title="MDAnalysis.core.groups.AtomGroup.concatenate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concatenate</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.intersection" title="MDAnalysis.core.groups.AtomGroup.intersection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">intersection</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.unique">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">unique</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.unique" title="Permalink to this definition"></a></dt>
<dd><p>An <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> containing sorted and unique
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> only.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">MDAnalysis</span> <span class="k">as</span> <span class="nn">mda</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span>
<span class="go">&lt;AtomGroup with 6 atoms&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([2, 1, 2, 2, 1, 0], dtype=int64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">unique</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span>
<span class="go">&lt;AtomGroup with 3 atoms&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([0, 1, 2], dtype=int64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span><span class="o">.</span><span class="n">unique</span> <span class="ow">is</span> <span class="n">ag2</span>
<span class="go">False</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.asunique" title="MDAnalysis.core.groups.AtomGroup.asunique"><code class="xref py py-obj docutils literal notranslate"><span class="pre">asunique</span></code></a>, <code class="xref py py-obj docutils literal notranslate"></code>, <code class="xref py py-obj docutils literal notranslate"></code></p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"></code></dt><dd><p>This function now always returns a copy.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.universe">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">universe</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.universe" title="Permalink to this definition"></a></dt>
<dd><p>The underlying <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> the group
belongs to.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.unwrap">
<span class="sig-name descname"><span class="pre">unwrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fragments'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'com'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.unwrap" title="Permalink to this definition"></a></dt>
<dd><p>Move atoms of this group so that bonds within the
group’s compounds aren’t split across periodic boundaries.</p>
<p>This function is most useful when atoms have been packed into the
primary unit cell, causing breaks mid-molecule, with the molecule then
appearing on either side of the unit cell. This is problematic for
operations such as calculating the center of mass of the molecule.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>+-----------+       +-----------+
|           |       |           |
| 6       3 |       |         3 | 6
| !       ! |       |         ! | !
|-5-8   1-2-|  ==&gt;  |       1-2-|-5-8
| !       ! |       |         ! | !
| 7       4 |       |         4 | 7
|           |       |           |
+-----------+       +-----------+
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
Which type of compound to unwrap. Note that, in any case, all
atoms within each compound must be interconnected by bonds, i.e.,
compounds must correspond to (parts of) molecules.</p></li>
<li><p><strong>reference</strong> (<em>{'com'</em><em>, </em><em>'cog'</em><em>, </em><em>None}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'com'</span></code> (center of mass) or <code class="docutils literal notranslate"><span class="pre">'cog'</span></code> (center of geometry), the
unwrapped compounds will be shifted so that their individual
reference point lies within the primary unit cell.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no such shift is performed.</p></li>
<li><p><strong>inplace</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, coordinates are modified in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>coords</strong> – Unwrapped atom coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> but the underlying topology does
    not contain molecule information, or if <cite>reference</cite> is <code class="docutils literal notranslate"><span class="pre">'com'</span></code>
    but the topology does not contain masses.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>reference</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'com'</span></code>, <code class="docutils literal notranslate"><span class="pre">'cog'</span></code>, or <code class="docutils literal notranslate"><span class="pre">None</span></code>, or
    if <cite>reference</cite> is <code class="docutils literal notranslate"><span class="pre">'com'</span></code> and the total mass of any <cite>compound</cite> is
    zero.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Be aware of the fact that only atoms <em>belonging to the group</em> will
be unwrapped! If you want entire molecules to be unwrapped, make sure
that all atoms of these molecules are part of the group.
An AtomGroup containing all atoms of all fragments in the group <code class="docutils literal notranslate"><span class="pre">ag</span></code>
can be created with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">all_frag_atoms</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">ag</span><span class="o">.</span><span class="n">fragments</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="../lib/mdamath.html#MDAnalysis.lib.mdamath.make_whole" title="MDAnalysis.lib.mdamath.make_whole"><code class="xref py py-func docutils literal notranslate"><span class="pre">make_whole()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.wrap" title="MDAnalysis.core.groups.AtomGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">wrap()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.pack_into_box" title="MDAnalysis.core.groups.AtomGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pack_into_box()</span></code></a>, <code class="xref py py-func docutils literal notranslate"><span class="pre">apply_PBC()</span></code></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.ureybradley">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ureybradley</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.ureybradley" title="Permalink to this definition"></a></dt>
<dd><p>This <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> represented as an
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.UreyBradley" title="MDAnalysis.core.topologyobjects.UreyBradley"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.topologyobjects.UreyBradley</span></code></a> object</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> is not length 2</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.0.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.velocities">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">velocities</span></span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.velocities" title="Permalink to this definition"></a></dt>
<dd><p>Velocities of the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>.</p>
<p>A <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> with
<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.shape.html#numpy.ndarray.shape" title="(in NumPy v1.25)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">shape</span></code></a><code class="docutils literal notranslate"><span class="pre">=(</span></code><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.n_atoms" title="MDAnalysis.core.groups.AtomGroup.n_atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">n_atoms</span></code></a><code class="docutils literal notranslate"><span class="pre">,</span> <span class="pre">3)</span></code>
and <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.dtype.html#numpy.ndarray.dtype" title="(in NumPy v1.25)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dtype</span></code></a><code class="docutils literal notranslate"><span class="pre">=numpy.float32</span></code>.</p>
<p>The velocities can be changed by assigning an array of the appropriate
shape, i.e. either <code class="docutils literal notranslate"><span class="pre">(</span></code><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.n_atoms" title="MDAnalysis.core.groups.AtomGroup.n_atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">n_atoms</span></code></a><code class="docutils literal notranslate"><span class="pre">,</span> <span class="pre">3)</span></code> to assign
individual velocities or <code class="docutils literal notranslate"><span class="pre">(3,)</span></code> to assign the <em>same</em> velocity to all
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> (e.g. <code class="docutils literal notranslate"><span class="pre">ag.velocities</span> <span class="pre">=</span> <span class="pre">array([0,0,0])</span></code> will give
all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> zero <a class="reference internal" href="#MDAnalysis.core.groups.Atom.velocity" title="MDAnalysis.core.groups.Atom.velocity"><code class="xref py py-attr docutils literal notranslate"><span class="pre">velocity</span></code></a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If the underlying <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a>
    does not contain
    <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.velocities" title="MDAnalysis.coordinates.timestep.Timestep.velocities"><code class="xref py py-attr docutils literal notranslate"><span class="pre">velocities</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.wrap">
<span class="sig-name descname"><span class="pre">wrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'atoms'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'com'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.wrap" title="Permalink to this definition"></a></dt>
<dd><p>Shift the contents of this group back into the primary unit cell
according to periodic boundary conditions.</p>
<p>Specifying a <cite>compound</cite> will keep the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in each
compound together during the process. If <cite>compound</cite> is different from
<code class="docutils literal notranslate"><span class="pre">'atoms'</span></code>, each compound as a whole will be shifted so that its
<cite>center</cite> lies within the primary unit cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>compound</strong> (<em>{'atoms'</em><em>, </em><em>'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
Which type of compound to keep together during wrapping. Note that,
in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
<li><p><strong>center</strong> (<em>{'com'</em><em>, </em><em>'cog'}</em>) – How to define the center of a given group of atoms. If <cite>compound</cite> is
<code class="docutils literal notranslate"><span class="pre">'atoms'</span></code>, this parameter is meaningless and therefore ignored.</p></li>
<li><p><strong>box</strong> (<em>array_like</em><em>, </em><em>optional</em>) – The unitcell dimensions of the system, which can be orthogonal or
triclinic and must be provided in the same format as returned by
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.dimensions" title="MDAnalysis.coordinates.timestep.Timestep.dimensions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MDAnalysis.coordinates.timestep.Timestep.dimensions</span></code></a>:
<code class="docutils literal notranslate"><span class="pre">[lx,</span> <span class="pre">ly,</span> <span class="pre">lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, uses the
dimensions of the current time step.</p></li>
<li><p><strong>inplace</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, coordinates will be changed in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array of wrapped atom coordinates of dtype <cite>np.float32</cite> and shape
<code class="docutils literal notranslate"><span class="pre">(len(self.atoms.n_atoms),</span> <span class="pre">3)</span></code></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>compound</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'atoms'</span></code>, <code class="docutils literal notranslate"><span class="pre">'group'</span></code>,
    <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>, <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>.</p></li>
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecule'</span></code> but the topology doesn’t
    contain molecule information (molnums) or if <cite>compound</cite> is
    <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> but the topology doesn’t contain bonds or if
    <cite>center</cite> is <code class="docutils literal notranslate"><span class="pre">'com'</span></code> but the topology doesn’t contain masses.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>All atoms of the group will be moved so that the centers of its
compounds lie within the primary periodic image. For orthorhombic unit
cells, the primary periodic image is defined as the half-open interval
<span class="math notranslate nohighlight">\([0,L_i)\)</span> between <span class="math notranslate nohighlight">\(0\)</span> and boxlength <span class="math notranslate nohighlight">\(L_i\)</span> in all
dimensions <span class="math notranslate nohighlight">\(i\in\{x,y,z\}\)</span>, i.e., the origin of the of the
simulation box is taken to be at the origin <span class="math notranslate nohighlight">\((0,0,0)\)</span> of the
euclidian coordinate system. A compound center residing at position
<span class="math notranslate nohighlight">\(x_i\)</span> in dimension <span class="math notranslate nohighlight">\(i\)</span> will be shifted to <span class="math notranslate nohighlight">\(x_i'\)</span>
according to</p>
<div class="math notranslate nohighlight">
\[x_i' = x_i - \left\lfloor\frac{x_i}{L_i}\right\rfloor\,.\]</div>
<p>When specifying a <cite>compound</cite>, the translation is calculated based on
each compound. The same translation is applied to all atoms
within this compound, meaning it will not be broken by the shift.
This might however mean that not all atoms of a compound will be
inside the unit cell after wrapping, but rather will be the center of
the compound.
Be aware of the fact that only atoms <em>belonging to the group</em> will be
taken into account!</p>
<p><cite>center</cite> allows to define how the center of each group is computed.
This can be either <code class="docutils literal notranslate"><span class="pre">'com'</span></code> for center of mass, or <code class="docutils literal notranslate"><span class="pre">'cog'</span></code> for
center of geometry.</p>
<p><cite>box</cite> allows a unit cell to be given for the transformation. If not
specified, the <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.dimensions" title="MDAnalysis.coordinates.timestep.Timestep.dimensions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dimensions</span></code></a>
information from the current
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a> will be used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.wrap" title="MDAnalysis.core.groups.AtomGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AtomGroup.wrap()</span></code></a> is currently faster than
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.wrap" title="MDAnalysis.core.groups.ResidueGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ResidueGroup.wrap()</span></code></a> or <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.wrap" title="MDAnalysis.core.groups.SegmentGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SegmentGroup.wrap()</span></code></a>.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.pack_into_box" title="MDAnalysis.core.groups.AtomGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pack_into_box()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.unwrap" title="MDAnalysis.core.groups.AtomGroup.unwrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">unwrap()</span></code></a>, <a class="reference internal" href="../lib/distances.html#MDAnalysis.lib.distances.apply_PBC" title="MDAnalysis.lib.distances.apply_PBC"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MDAnalysis.lib.distances.apply_PBC()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.2.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>The method only acts on atoms <em>belonging to the group</em> and returns
the wrapped positions as a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a>.
Added optional argument <cite>inplace</cite>.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.AtomGroup.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filenamefmt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'{trjname}_{frame}'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#AtomGroup.write"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.AtomGroup.write" title="Permalink to this definition"></a></dt>
<dd><p>Write <cite>AtomGroup</cite> to a file.</p>
<p>The output can either be a coordinate file or a selection, depending on
the format.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;selection.ndx&#39;</span><span class="p">)</span>  <span class="c1"># Write a gromacs index file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;coordinates.pdb&#39;</span><span class="p">)</span>  <span class="c1"># Write the current frame as PDB</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Write the trajectory in XTC format</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;trajectory.xtc&#39;</span><span class="p">,</span> <span class="n">frames</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Write every other frame of the trajectory in PBD format</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;trajectory.pdb&#39;</span><span class="p">,</span> <span class="n">frames</span><span class="o">=</span><span class="n">u</span><span class="o">.</span><span class="n">trajectory</span><span class="p">[::</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em>, </em><em>optional</em>) – <code class="docutils literal notranslate"><span class="pre">None</span></code>: create TRJNAME_FRAME.FORMAT from filenamefmt [<code class="docutils literal notranslate"><span class="pre">None</span></code>]</p></li>
<li><p><strong>file_format</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em>, </em><em>optional</em>) – The name or extension of a coordinate, trajectory, or selection
file format such as PDB, CRD, GRO, VMD (tcl), PyMol (pml), Gromacs
(ndx) CHARMM (str) or Jmol (spt); case-insensitive [PDB]</p></li>
<li><p><strong>filenamefmt</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em>, </em><em>optional</em>) – format string for default filename; use substitution tokens
‘trjname’ and ‘frame’ [“%(trjname)s_%(frame)d”]</p></li>
<li><p><strong>bonds</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em>, </em><em>optional</em>) – how to handle bond information, especially relevant for PDBs.
<code class="docutils literal notranslate"><span class="pre">&quot;conect&quot;</span></code>: write only the CONECT records defined in the original
file. <code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code>: write out all bonds, both the original defined and
those guessed by MDAnalysis. <code class="docutils literal notranslate"><span class="pre">None</span></code>: do not write out bonds.
Default is <code class="docutils literal notranslate"><span class="pre">&quot;conect&quot;</span></code>.</p></li>
<li><p><strong>frames</strong> (<em>array-like</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#slice" title="(in Python v3.11)"><em>slice</em></a><em> or </em><a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.FrameIteratorBase" title="MDAnalysis.coordinates.base.FrameIteratorBase"><em>FrameIteratorBase</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em>, </em><em>optional</em>) – An ensemble of frames to write. The ensemble can be an list or
array of frame indices, a mask of booleans, an instance of
<a class="reference external" href="https://docs.python.org/3/library/functions.html#slice" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">slice</span></code></a>, or the value returned when a trajectory is indexed.
By default, <cite>frames</cite> is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> and only the current frame
is written. If <cite>frames</cite> is set to “all”, then all the frame from
trajectory are written.</p></li>
</ul>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.9.0: </span>Merged with write_selection. This method can
now write both selections out.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Can write multiframe trajectories with the ‘frames’ argument.</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.groups.</span></span><span class="sig-name descname"><span class="pre">ResidueGroup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#ResidueGroup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup" title="Permalink to this definition"></a></dt>
<dd><p>ResidueGroup base class.</p>
<p>This class is used by a <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> for
generating its Topology-specific <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> class. All the
<a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.TopologyAttr" title="MDAnalysis.core.topologyattrs.TopologyAttr"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyAttr</span></code></a> components are obtained
from <code class="xref py py-class docutils literal notranslate"><span class="pre">GroupBase</span></code>, so this class only includes ad-hoc methods
specific to <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroups</span></code></a>.</p>
<p>ResidueGroups can be compared and combined using group operators. See the
list of these operators on <code class="xref py py-class docutils literal notranslate"><span class="pre">GroupBase</span></code>.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.16.2: </span><em>Instant selectors</em> of Segments will be removed in the 1.0 release.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed instant selectors, use select_atoms instead</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Indexing an ResidueGroup with <code class="docutils literal notranslate"><span class="pre">None</span></code> raises a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code>.</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.accumulate">
<span class="sig-name descname"><span class="pre">accumulate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attribute</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">function=&lt;function</span> <span class="pre">sum&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound='group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.accumulate" title="Permalink to this definition"></a></dt>
<dd><p>Accumulates the attribute associated with (compounds of) the group.</p>
<p>Accumulates the attribute of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
The accumulation per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule,
or fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. By default, the method sums up all attributes per compound,
but any function that takes an array and returns an acuumulation over a
given axis can be used. For multi-dimensional input arrays, the
accumulation is performed along the first axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> or </em><em>array_like</em>) – Attribute or array of values to accumulate.
If a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> (or compatible) is provided, its first
dimension must have the same length as the total number of atoms in
the group.</p></li>
<li><p><strong>function</strong> (<em>callable</em><em>, </em><em>optional</em>) – The function performing the accumulation. It must take the array of
attribute values to accumulate as its only positional argument and
accept an (optional) keyword argument <code class="docutils literal notranslate"><span class="pre">axis</span></code> allowing to specify
the axis along which the accumulation is performed.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the accumulation of all attributes associated with
atoms in the group will be returned as a single value. Otherwise,
the accumulation of the attributes per <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>,
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment will be returned as a 1d
array. Note that, in any case, <em>only</em> the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Acuumulation of the <cite>attribute</cite>.
If <cite>compound</cite> is set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the first dimension of the
<cite>attribute</cite> array will be contracted to a single value.
If <cite>compound</cite> is set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>, <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>, the length of the first
dimension will correspond to the number of compounds. In all cases,
the other dimensions of the returned array will be of the original
shape (without the first dimension).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If the length of a provided <cite>attribute</cite> array does not correspond to
    the number of atoms in the group.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>compound</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>,
    <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>.</p></li>
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecule'</span></code> but the topology doesn’t
    contain molecule information (molnums), or if <cite>compound</cite> is
    <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> but the topology doesn’t contain bonds.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>To find the total charge of a given <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;prop mass &gt; 4.0&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">accumulate</span><span class="p">(</span><span class="s1">&#39;charges&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>To find the total mass per residue of all CA <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">accumulate</span><span class="p">(</span><span class="s1">&#39;masses&#39;</span><span class="p">,</span> <span class="n">compound</span><span class="o">=</span><span class="s1">&#39;residues&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>To find the maximum atomic charge per fragment of a given
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">accumulate</span><span class="p">(</span><span class="s1">&#39;charges&#39;</span><span class="p">,</span> <span class="n">compound</span><span class="o">=</span><span class="s2">&quot;fragments&quot;</span><span class="p">,</span> <span class="n">function</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">)</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.align_principal_axis">
<span class="sig-name descname"><span class="pre">align_principal_axis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">axis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.align_principal_axis" title="Permalink to this definition"></a></dt>
<dd><p>Align principal axis with index <cite>axis</cite> with <cite>vector</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>{0</em><em>, </em><em>1</em><em>, </em><em>2}</em>) – Index of the principal axis (0, 1, or 2), as produced by
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.principal_axes" title="MDAnalysis.core.groups.ResidueGroup.principal_axes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">principal_axes()</span></code></a>.</p></li>
<li><p><strong>vector</strong> (<em>array_like</em>) – Vector to align principal axis with.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>To align the long axis of a channel (the first principal axis, i.e.
<em>axis</em> = 0) with the z-axis:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">align_principal_axis</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;aligned.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.asphericity">
<span class="sig-name descname"><span class="pre">asphericity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.asphericity" title="Permalink to this definition"></a></dt>
<dd><p>Asphericity.</p>
<p>See <a class="reference internal" href="../references.html#dima2004b" id="id6"><span>[Dima2004b]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <em>unwrap</em> and <em>compound</em> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.5.0: </span>Added calculation for any <cite>compound</cite> type</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.asunique">
<span class="sig-name descname"><span class="pre">asunique</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sorted</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#ResidueGroup.asunique"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.asunique" title="Permalink to this definition"></a></dt>
<dd><p>Return a <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> containing unique
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residues</span></code></a> only, with optional sorting.</p>
<p>If the <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> is unique, this is the group itself.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sorted</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Whether or not the returned ResidueGroup should be sorted
by resindex.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Unique <code class="docutils literal notranslate"><span class="pre">ResidueGroup</span></code></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rg</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">residues</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg</span>
<span class="go">&lt;ResidueGroup with 6 residues&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([2, 1, 2, 2, 1, 0])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg2</span> <span class="o">=</span> <span class="n">rg</span><span class="o">.</span><span class="n">asunique</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg2</span>
<span class="go">&lt;ResidueGroup with 3 residues&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg2</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([0, 1, 2])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg2</span><span class="o">.</span><span class="n">asunique</span><span class="p">()</span> <span class="ow">is</span> <span class="n">rg2</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.0.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.atoms" title="Permalink to this definition"></a></dt>
<dd><p>An <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> present in this
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a>.</p>
<p>The <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> are ordered locally by <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> in the
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a>.  Duplicates are <em>not</em> removed.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.bbox">
<span class="sig-name descname"><span class="pre">bbox</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.bbox" title="Permalink to this definition"></a></dt>
<dd><p>Return the bounding box of the selection.</p>
<p>The lengths A,B,C of the orthorhombic enclosing box are</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">L</span> <span class="o">=</span> <span class="n">AtomGroup</span><span class="o">.</span><span class="n">bbox</span><span class="p">()</span>
<span class="n">A</span><span class="p">,</span><span class="n">B</span><span class="p">,</span><span class="n">C</span> <span class="o">=</span> <span class="n">L</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">L</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to the primary unit cell
before calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>corners</strong> – 2x3 array giving corners of bounding box as
<code class="docutils literal notranslate"><span class="pre">[[xmin,</span> <span class="pre">ymin,</span> <span class="pre">zmin],</span> <span class="pre">[xmax,</span> <span class="pre">ymax,</span> <span class="pre">zmax]]</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.2.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.bfactors">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bfactors</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.bfactors" title="Permalink to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have tempfactors. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.bsphere">
<span class="sig-name descname"><span class="pre">bsphere</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.bsphere" title="Permalink to this definition"></a></dt>
<dd><p>Return the bounding sphere of the selection.</p>
<p>The sphere is calculated relative to the
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.center_of_geometry" title="MDAnalysis.core.groups.ResidueGroup.center_of_geometry"><code class="xref py py-meth docutils literal notranslate"><span class="pre">center</span> <span class="pre">of</span> <span class="pre">geometry</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms to the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>R</strong> (<em>float</em>) – Radius of the bounding sphere.</p></li>
<li><p><strong>center</strong> (<em>numpy.ndarray</em>) – Coordinates of the sphere center as <code class="docutils literal notranslate"><span class="pre">[xcen,</span> <span class="pre">ycen,</span> <span class="pre">zcen]</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.3.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.center">
<span class="sig-name descname"><span class="pre">center</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">weights</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.center" title="Permalink to this definition"></a></dt>
<dd><p>Weighted center of (compounds of) the group</p>
<p>Computes the weighted center of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Weighted centers per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the weights of a compound sum up to zero, the
coordinates of that compound’s weighted center will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>weights</strong> (<em>array_like</em><em> or </em><em>None</em>) – Weights to be used. Setting <cite>weights=None</cite> is equivalent to passing
identical weights for all atoms of the group.</p></li>
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">'group'</span></code> the center of each
compound will be calculated without moving any
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to keep the compounds intact. Instead, the
resulting position vectors will be moved to the primary unit cell
after calculation. Default [<code class="docutils literal notranslate"><span class="pre">False</span></code>].</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – <dl class="simple">
<dt>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their</dt><dd><p>centers.</p>
</dd>
</dl>
</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the weighted center of all atoms in the group will
be returned as a single position vector. Else, the weighted centers
of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment
will be returned as an array of position vectors, i.e. a 2d array.
Note that, in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the weighted center(s) of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>, <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>, the output will be a 2d array
of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>compound</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>,
    <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If both ‘wrap’ and ‘unwrap’ set to true.</p></li>
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecule'</span></code> but the topology doesn’t
    contain molecule information (molnums) or if <cite>compound</cite> is
    <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> but the topology doesn’t contain bonds.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>To find the center of charge of a given <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;prop mass &gt; 4.0&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">center</span><span class="p">(</span><span class="n">sel</span><span class="o">.</span><span class="n">charges</span><span class="p">)</span>
</pre></div>
</div>
<p>To find the centers of mass per residue of all CA <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">center</span><span class="p">(</span><span class="n">sel</span><span class="o">.</span><span class="n">masses</span><span class="p">,</span> <span class="n">compound</span><span class="o">=</span><span class="s1">&#39;residues&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>
compounds</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.center_of_charge">
<span class="sig-name descname"><span class="pre">center_of_charge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.center_of_charge" title="Permalink to this definition"></a></dt>
<dd><p>Center of (absolute) charge of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \vert q_i \vert \boldsymbol r_i}
                     {\sum_i \vert q_i \vert}\]</div>
<p>where <span class="math notranslate nohighlight">\(q_i\)</span> is the charge and <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> the
position of atom <span class="math notranslate nohighlight">\(i\)</span> in the given
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup</span></code></a>.
Centers of charge per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the charges of a compound sum up to zero, the
center of mass coordinates of that compound will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of mass of all atoms in the group will
be returned as a single position vector. Otherwise, the centers of
mass of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the center(s) of charge of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where
<code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method can only be accessed if the underlying topology has
information about atomic charges.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.2.0.</span></p>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.center_of_geometry">
<span class="sig-name descname"><span class="pre">center_of_geometry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.center_of_geometry" title="Permalink to this definition"></a></dt>
<dd><p>Center of geometry of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \boldsymbol r_i}{\sum_i 1}\]</div>
<p>where <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <span class="math notranslate nohighlight">\(i\)</span>.
Centers of geometry per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> or per <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> can
be obtained by setting the <cite>compound</cite> parameter accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation. If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is
<code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the center of each compound will
be calculated without moving any <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to keep the
compounds intact. Instead, the resulting position vectors will be
moved to the primary unit cell after calculation. Default False.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of geometry of all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
in the group will be returned as a single position vector. Else,
the centers of geometry of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> will be returned as an array of position vectors,
i.e. a 2d array. Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the geometric center(s) of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the
number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>
compounds</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.center_of_mass">
<span class="sig-name descname"><span class="pre">center_of_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.center_of_mass" title="Permalink to this definition"></a></dt>
<dd><p>Center of mass of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i m_i \boldsymbol r_i}{\sum m_i}\]</div>
<p>where <span class="math notranslate nohighlight">\(m_i\)</span> is the mass and <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> the
position of atom <span class="math notranslate nohighlight">\(i\)</span> in the given
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup</span></code></a>.
Centers of mass per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the masses of a compound sum up to zero, the
center of mass coordinates of that compound will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of mass of all atoms in the group will
be returned as a single position vector. Otherwise, the centers of
mass of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e. a 2d
array.
Note that, in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the center(s) of mass of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code>
is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method can only be accessed if the underlying topology has
information about atomic masses.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> compounds;
added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.centroid">
<span class="sig-name descname"><span class="pre">centroid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.centroid" title="Permalink to this definition"></a></dt>
<dd><p>Center of geometry of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \boldsymbol r_i}{\sum_i 1}\]</div>
<p>where <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <span class="math notranslate nohighlight">\(i\)</span>.
Centers of geometry per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> or per <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> can
be obtained by setting the <cite>compound</cite> parameter accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation. If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is
<code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the center of each compound will
be calculated without moving any <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to keep the
compounds intact. Instead, the resulting position vectors will be
moved to the primary unit cell after calculation. Default False.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of geometry of all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
in the group will be returned as a single position vector. Else,
the centers of geometry of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> will be returned as an array of position vectors,
i.e. a 2d array. Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the geometric center(s) of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the
number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>
compounds</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.chi1_selections">
<span class="sig-name descname"><span class="pre">chi1_selections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cg_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CG</span> <span class="pre">CG1</span> <span class="pre">OG</span> <span class="pre">OG1</span> <span class="pre">SG'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.chi1_selections" title="Permalink to this definition"></a></dt>
<dd><p>Select list of AtomGroups corresponding to the chi1 sidechain dihedral
N-CA-CB-CG.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
<li><p><strong>cb_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the beta-carbon atom</p></li>
<li><p><strong>cg_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the gamma-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>List of AtomGroups</em> – 4-atom selections in the correct order. If no CB and/or CG is found
then the corresponding item in the list is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionadded:: 1.0.0</em></p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have names. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.concatenate">
<span class="sig-name descname"><span class="pre">concatenate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.concatenate" title="Permalink to this definition"></a></dt>
<dd><p>Concatenate with another Group or Component of the same level.</p>
<p>Duplicate entries and original order is preserved. It is synomymous to
the <cite>+</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with elements of <cite>self</cite> and <cite>other</cite> concatenated</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>The order of the original contents (including duplicates)
are preserved when performing a concatenation.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name O&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name N&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">+</span> <span class="n">ag2</span>  <span class="c1"># or ag1.concatenate(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">names</span>
<span class="go">array([&#39;O&#39;, &#39;O&#39;, &#39;O&#39;], dtype=object)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]</span><span class="o">.</span><span class="n">names</span>
<span class="go">array([&#39;N&#39;, &#39;N&#39;, &#39;N&#39;], dtype=object)</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.copy" title="Permalink to this definition"></a></dt>
<dd><p>Get another group identical to this one.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.19.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.difference">
<span class="sig-name descname"><span class="pre">difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.difference" title="Permalink to this definition"></a></dt>
<dd><p>Elements from this Group that do not appear in another</p>
<p>This method removes duplicate elements and sorts the result. As such,
it is different from <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.subtract" title="MDAnalysis.core.groups.ResidueGroup.subtract"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subtract()</span></code></a>. <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.difference" title="MDAnalysis.core.groups.ResidueGroup.difference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">difference()</span></code></a> is synomymous
to the <cite>-</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the elements of <cite>self</cite> that are not in  <cite>other</cite>, without
duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.subtract" title="MDAnalysis.core.groups.ResidueGroup.subtract"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subtract</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.symmetric_difference" title="MDAnalysis.core.groups.ResidueGroup.symmetric_difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">symmetric_difference</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.dimensions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dimensions</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.dimensions" title="Permalink to this definition"></a></dt>
<dd><p>Obtain a copy of the dimensions of the currently loaded Timestep</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.dipole_moment">
<span class="sig-name descname"><span class="pre">dipole_moment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.dipole_moment" title="Permalink to this definition"></a></dt>
<dd><p>Dipole moment of the group or compounds in a group.</p>
<div class="math notranslate nohighlight">
\[\mu = |\boldsymbol{\mu}| = \sqrt{ \sum_{i=1}^{D} \mu^2 }\]</div>
<p>Where <span class="math notranslate nohighlight">\(D\)</span> is the number of dimensions, in this case 3.</p>
<p>Computes the dipole moment of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Dipole per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is a net charge, the magnitude of the dipole
moment is dependent on the <cite>center</cite> chosen.
See <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.dipole_vector" title="MDAnalysis.core.groups.ResidueGroup.dipole_vector"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dipole_vector()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single dipole vector returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dipole moment(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.dipole_vector">
<span class="sig-name descname"><span class="pre">dipole_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mass'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.dipole_vector" title="Permalink to this definition"></a></dt>
<dd><p>Dipole vector of the group.</p>
<div class="math notranslate nohighlight">
\[\boldsymbol{\mu} = \sum_{i=1}^{N} q_{i} ( \mathbf{r}_{i} -
\mathbf{r}_{COM} )\]</div>
<p>Computes the dipole vector of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Dipole vector per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that the magnitude of the dipole moment is independent of the
<code class="docutils literal notranslate"><span class="pre">center</span></code> chosen unless the species has a net charge. In the case of
a charged group the dipole moment can be later adjusted  with:</p>
<div class="math notranslate nohighlight">
\[\boldsymbol{\mu}_{COC} = \boldsymbol{\mu}_{COM} +
q_{ag}\mathbf{r}_{COM} - q_{ag}\boldsymbol{r}_{COC}\]</div>
<p>Where <span class="math notranslate nohighlight">\(\mathbf{r}_{COM}\)</span> is the center of mass and
<span class="math notranslate nohighlight">\(\mathbf{r}_{COC}\)</span> is the center of charge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single dipole vector returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dipole vector(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,3)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.get_connections">
<span class="sig-name descname"><span class="pre">get_connections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">typename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outside</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.get_connections" title="Permalink to this definition"></a></dt>
<dd><p>Get bonded connections between atoms as a
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.TopologyGroup" title="MDAnalysis.core.topologyobjects.TopologyGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyGroup</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>typename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a>) – group name. One of {“bonds”, “angles”, “dihedrals”,
“impropers”, “ureybradleys”, “cmaps”}</p></li>
<li><p><strong>outside</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Whether to include connections involving atoms outside
this group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>TopologyGroup</em> – containing the bonded group of choice, i.e. bonds, angles,
dihedrals, impropers, ureybradleys or cmaps.</p></li>
<li><p><em>.. versionadded:: 1.1.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.groupby">
<span class="sig-name descname"><span class="pre">groupby</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">topattrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.groupby" title="Permalink to this definition"></a></dt>
<dd><p>Group together items in this group according to values of <em>topattr</em></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>topattrs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.11)"><em>list</em></a>) – One or more topology attributes to group components by.
Single arguments are passed as a string. Multiple arguments
are passed as a list.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Unique values of the multiple combinations of topology attributes
as keys, Groups as values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.11)">dict</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>To group atoms with the same mass together:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;masses&#39;</span><span class="p">)</span>
<span class="go">{12.010999999999999: &lt;AtomGroup with 462 atoms&gt;,</span>
<span class="go"> 14.007: &lt;AtomGroup with 116 atoms&gt;,</span>
<span class="go"> 15.999000000000001: &lt;AtomGroup with 134 atoms&gt;}</span>
</pre></div>
</div>
<p>To group atoms with the same residue name and mass together:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;resnames&#39;</span><span class="p">,</span> <span class="s1">&#39;masses&#39;</span><span class="p">])</span>
<span class="go">{(&#39;ALA&#39;, 1.008): &lt;AtomGroup with 95 atoms&gt;,</span>
<span class="go"> (&#39;ALA&#39;, 12.011): &lt;AtomGroup with 57 atoms&gt;,</span>
<span class="go"> (&#39;ALA&#39;, 14.007): &lt;AtomGroup with 19 atoms&gt;,</span>
<span class="go"> (&#39;ALA&#39;, 15.999): &lt;AtomGroup with 19 atoms&gt;},</span>
<span class="go"> (&#39;ARG&#39;, 1.008): &lt;AtomGroup with 169 atoms&gt;,</span>
<span class="go"> ...}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;resnames&#39;</span><span class="p">,</span> <span class="s1">&#39;masses&#39;</span><span class="p">])(</span><span class="s1">&#39;ALA&#39;</span><span class="p">,</span> <span class="mf">15.999</span><span class="p">)</span>
<span class="go"> &lt;AtomGroup with 19 atoms&gt;</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.18.0: </span>The function accepts multiple attributes</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.gyration_moments">
<span class="sig-name descname"><span class="pre">gyration_moments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.gyration_moments" title="Permalink to this definition"></a></dt>
<dd><p>Moments of the gyration tensor.</p>
<p>The moments are defined as the eigenvalues of the gyration
tensor.</p>
<div class="math notranslate nohighlight">
\[\mathsf{T} = \frac{1}{N} \sum_{i=1}^{N} (\mathbf{r}_\mathrm{i} -
    \mathbf{r}_\mathrm{COM})(\mathbf{r}_\mathrm{i} - \mathbf{r}_\mathrm{COM})\]</div>
<p>Where <span class="math notranslate nohighlight">\(\mathbf{r}_\mathrm{COM}\)</span> is the center of mass.</p>
<p>See <a class="reference internal" href="../references.html#dima2004a" id="id7"><span>[Dima2004a]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>principle_moments_of_gyration</strong> – Gyration vector(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(Å^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
vector of length 3. Otherwise, the output will be a 2D array of shape
<code class="docutils literal notranslate"><span class="pre">(n,3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.5.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.intersection">
<span class="sig-name descname"><span class="pre">intersection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.intersection" title="Permalink to this definition"></a></dt>
<dd><p>Group of elements which are in both this Group and another</p>
<p>This method removes duplicate elements and sorts the result. It is
synomymous to the <cite>&amp;</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the common elements of <cite>self</cite> and <cite>other</cite>, without
duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Intersections can be used when the select atoms string would
become too complicated.  For example to find the water atoms
which are within 4.0A of two segments:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">shell1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;resname SOL and around 4.0 segid 1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">shell2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;resname SOL and around 4.0 segid 2&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">common</span> <span class="o">=</span> <span class="n">shell1</span> <span class="o">&amp;</span> <span class="n">shell2</span>  <span class="c1"># or shell1.intersection(shell2)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.union" title="MDAnalysis.core.groups.ResidueGroup.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">union</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.is_strict_subset">
<span class="sig-name descname"><span class="pre">is_strict_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.is_strict_subset" title="Permalink to this definition"></a></dt>
<dd><p>If this Group is a subset of another Group but not identical</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a strict subset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.is_strict_superset">
<span class="sig-name descname"><span class="pre">is_strict_superset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.is_strict_superset" title="Permalink to this definition"></a></dt>
<dd><p>If this Group is a superset of another Group but not identical</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a strict superset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.isdisjoint">
<span class="sig-name descname"><span class="pre">isdisjoint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.isdisjoint" title="Permalink to this definition"></a></dt>
<dd><p>If the Group has no elements in common with the other Group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if the two Groups do not have common elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.issubset">
<span class="sig-name descname"><span class="pre">issubset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.issubset" title="Permalink to this definition"></a></dt>
<dd><p>If all elements of this Group are part of another Group</p>
<p>Note that an empty group is a subset of any group of the same level.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a subset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.issuperset">
<span class="sig-name descname"><span class="pre">issuperset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.issuperset" title="Permalink to this definition"></a></dt>
<dd><p>If all elements of another Group are part of this Group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a subset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.isunique">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">isunique</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.isunique" title="Permalink to this definition"></a></dt>
<dd><p>Boolean indicating whether all components of the group are unique,
i.e., the group contains no duplicates.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span>
<span class="go">&lt;AtomGroup with 6 atoms&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">isunique</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">unique</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span>
<span class="go">&lt;AtomGroup with 3 atoms&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span><span class="o">.</span><span class="n">isunique</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.asunique" title="MDAnalysis.core.groups.ResidueGroup.asunique"><code class="xref py py-obj docutils literal notranslate"><span class="pre">asunique</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.19.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.ix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.ix" title="Permalink to this definition"></a></dt>
<dd><p>Unique indices of the components in the Group.</p>
<ul class="simple">
<li><p>If this Group is an <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>, these are the
indices of the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> instances.</p></li>
<li><p>If it is a <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a>, these are the indices of
the <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> instances.</p></li>
<li><p>If it is a <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>, these are the indices of
the <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> instances.</p></li>
</ul>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.ix_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix_array</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.ix_array" title="Permalink to this definition"></a></dt>
<dd><p>Unique indices of the components in the Group.</p>
<p>For a Group, <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.ix_array" title="MDAnalysis.core.groups.ResidueGroup.ix_array"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ix_array</span></code></a> is the same as <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.ix" title="MDAnalysis.core.groups.ResidueGroup.ix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ix</span></code></a>. This method
gives a consistent API between components and groups.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.ix" title="MDAnalysis.core.groups.ResidueGroup.ix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ix</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.moment_of_inertia">
<span class="sig-name descname"><span class="pre">moment_of_inertia</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.moment_of_inertia" title="Permalink to this definition"></a></dt>
<dd><p>Moment of inertia tensor relative to center of mass.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers and tensor of inertia.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
<cite>compound</cite> determines the behavior of <cite>wrap</cite>.
Note that, in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>moment_of_inertia</strong> – Moment of inertia tensor as a 3 x 3 numpy array.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The moment of inertia tensor <span class="math notranslate nohighlight">\(\mathsf{I}\)</span> is calculated for a group of
<span class="math notranslate nohighlight">\(N\)</span> atoms with coordinates <span class="math notranslate nohighlight">\(\mathbf{r}_i,\ 1 \le i \le N\)</span>
relative to its center of mass from the relative coordinates</p>
<div class="math notranslate nohighlight">
\[\mathbf{r}'_i = \mathbf{r}_i - \frac{1}{\sum_{i=1}^{N} m_i} \sum_{i=1}^{N} m_i \mathbf{r}_i\]</div>
<p>as</p>
<div class="math notranslate nohighlight">
\[\mathsf{I} = \sum_{i=1}^{N} m_i \Big[(\mathbf{r}'_i\cdot\mathbf{r}'_i) \sum_{\alpha=1}^{3}
      \hat{\mathbf{e}}_\alpha \otimes \hat{\mathbf{e}}_\alpha - \mathbf{r}'_i \otimes \mathbf{r}'_i\Big]\]</div>
<p>where <span class="math notranslate nohighlight">\(\hat{\mathbf{e}}_\alpha\)</span> are Cartesian unit vectors, or in Cartesian coordinates</p>
<div class="math notranslate nohighlight">
\[I_{\alpha,\beta} = \sum_{k=1}^{N} m_k
      \Big(\big(\sum_{\gamma=1}^3 (x'^{(k)}_{\gamma})^2 \big)\delta_{\alpha,\beta}
      - x'^{(k)}_{\alpha} x'^{(k)}_{\beta} \Big).\]</div>
<p>where <span class="math notranslate nohighlight">\(x'^{(k)}_{\alpha}\)</span> are the Cartesian coordinates of the
relative coordinates <span class="math notranslate nohighlight">\(\mathbf{r}'_k\)</span>.</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.n_atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.n_atoms" title="Permalink to this definition"></a></dt>
<dd><p>Number of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> present in this <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a>,
including duplicate residues (and thus, duplicate atoms).</p>
<p>Equivalent to <code class="docutils literal notranslate"><span class="pre">len(self.atoms)</span></code>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.n_residues">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_residues</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.n_residues" title="Permalink to this definition"></a></dt>
<dd><p>Number of residues in the <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a>.</p>
<p>Equivalent to <code class="docutils literal notranslate"><span class="pre">len(self)</span></code>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.n_segments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_segments</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.n_segments" title="Permalink to this definition"></a></dt>
<dd><p>Number of unique segments present in the ResidueGroup.</p>
<p>Equivalent to <code class="docutils literal notranslate"><span class="pre">len(self.segments)</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.omega_selections">
<span class="sig-name descname"><span class="pre">omega_selections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.omega_selections" title="Permalink to this definition"></a></dt>
<dd><p>Select list of AtomGroups corresponding to the omega protein
backbone dihedral CA-C-N’-CA’.</p>
<p>omega describes the -C-N- peptide bond. Typically, it is trans (180
degrees) although cis-bonds (0 degrees) are also occasionally observed
(especially near Proline).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>List of AtomGroups</em> – 4-atom selections in the correct order. If no C’ found in the
previous residue (by resid) then the corresponding item in the
list is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionadded:: 1.0.0</em></p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have names. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.pack_into_box">
<span class="sig-name descname"><span class="pre">pack_into_box</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.pack_into_box" title="Permalink to this definition"></a></dt>
<dd><p>Shift all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in this group to the primary unit
cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>box</strong> (<em>array_like</em>) – Box dimensions, can be either orthogonal or triclinic information.
Cell dimensions must be in an identical to format to those returned
by <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.dimensions" title="MDAnalysis.coordinates.timestep.Timestep.dimensions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MDAnalysis.coordinates.timestep.Timestep.dimensions</span></code></a>,
<code class="docutils literal notranslate"><span class="pre">[lx,</span> <span class="pre">ly,</span> <span class="pre">lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, uses these
timestep dimensions.</p></li>
<li><p><strong>inplace</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a>) – <code class="docutils literal notranslate"><span class="pre">True</span></code> to change coordinates in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>coords</strong> – Shifted atom coordinates.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>All atoms will be moved so that they lie between 0 and boxlength
<span class="math notranslate nohighlight">\(L_i\)</span> in all dimensions, i.e. the lower left corner of the
simulation box is taken to be at (0,0,0):</p>
<div class="math notranslate nohighlight">
\[x_i' = x_i - \left\lfloor\frac{x_i}{L_i}\right\rfloor\]</div>
<p>The default is to take unit cell information from the underlying
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a> instance. The optional
argument <cite>box</cite> can be used to provide alternative unit cell information
(in the MDAnalysis standard format
<code class="docutils literal notranslate"><span class="pre">[Lx,</span> <span class="pre">Ly,</span> <span class="pre">Lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>).</p>
<p>Works with either orthogonal or triclinic box types.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.pack_into_box" title="MDAnalysis.core.groups.ResidueGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pack_into_box()</span></code></a> is identical to <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.wrap" title="MDAnalysis.core.groups.ResidueGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">wrap()</span></code></a> with all default
keywords.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.pack_into_box" title="MDAnalysis.core.groups.AtomGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AtomGroup.pack_into_box()</span></code></a> is currently faster than
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.pack_into_box" title="MDAnalysis.core.groups.ResidueGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ResidueGroup.pack_into_box()</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.pack_into_box" title="MDAnalysis.core.groups.SegmentGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SegmentGroup.pack_into_box()</span></code></a>.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.8.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.phi_selections">
<span class="sig-name descname"><span class="pre">phi_selections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.phi_selections" title="Permalink to this definition"></a></dt>
<dd><p>Select list of AtomGroups corresponding to the phi protein
backbone dihedral C’-N-CA-C.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>list of AtomGroups</em> – 4-atom selections in the correct order. If no C’ found in the
previous residue (by resid) then corresponding item in the list
is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionadded:: 1.0.0</em></p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have names. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.principal_axes">
<span class="sig-name descname"><span class="pre">principal_axes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.principal_axes" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the principal axes from the moment of inertia.</p>
<p>e1,e2,e3 = AtomGroup.principal_axes()</p>
<p>The eigenvectors are sorted by eigenvalue, i.e. the first one
corresponds to the highest eigenvalue and is thus the first principal
axes.</p>
<p>The eigenvectors form a right-handed coordinate system.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>axis_vectors</strong> – 3 x 3 array with <code class="docutils literal notranslate"><span class="pre">v[0]</span></code> as first, <code class="docutils literal notranslate"><span class="pre">v[1]</span></code> as second, and
<code class="docutils literal notranslate"><span class="pre">v[2]</span></code> as third eigenvector.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Always return principal axes in right-hand convention.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.psi_selections">
<span class="sig-name descname"><span class="pre">psi_selections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.psi_selections" title="Permalink to this definition"></a></dt>
<dd><p>Select list of AtomGroups corresponding to the psi protein
backbone dihedral N-CA-C-N’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>List of AtomGroups</em> – 4-atom selections in the correct order. If no N’ found in the
following residue (by resid) then the corresponding item in the
list is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionadded:: 1.0.0</em></p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have names. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.quadrupole_moment">
<span class="sig-name descname"><span class="pre">quadrupole_moment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.quadrupole_moment" title="Permalink to this definition"></a></dt>
<dd><p>Quadrupole moment of the group according to <span id="id8">[<a class="reference internal" href="topologyattrs.html#id2" title="C. G. Gray, Keith E. Gubbins, and C. G. Joslin. Theory of molecular fluids. Number 9 in The International series of monographs on chemistry. Oxford University Press, Oxford ; New York, 1984. ISBN 978-0-19-855602-2.">Gray1984</a>]</span>.</p>
<div class="math notranslate nohighlight">
\[Q = \sqrt{\frac{2}{3}{\hat{\mathsf{Q}}}:{\hat{\mathsf{Q}}}}\]</div>
<p>where the quadrupole moment is calculated from the tensor double
contraction of the traceless quadropole tensor <span class="math notranslate nohighlight">\(\hat{\mathsf{Q}}\)</span></p>
<p>Computes the quadrupole moment of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Quadrupole per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is an unsymmetrical plane in the molecule or
group, the magnitude of the quadrupole moment is dependant on the
<code class="docutils literal notranslate"><span class="pre">center</span></code> chosen and cannot be translated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single quadrupole value returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 1d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Quadrupole moment(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.quadrupole_tensor">
<span class="sig-name descname"><span class="pre">quadrupole_tensor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mass'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.quadrupole_tensor" title="Permalink to this definition"></a></dt>
<dd><p>Traceless quadrupole tensor of the group or compounds.</p>
<p>This tensor is first computed as the outer product of vectors from
a reference point to some atom, and multiplied by the atomic charge.
The tensor of each atom is then summed to produce the quadrupole
tensor of the group:</p>
<div class="math notranslate nohighlight">
\[\mathsf{Q} = \sum_{i=1}^{N} q_{i} ( \mathbf{r}_{i} -
\mathbf{r}_{COM} ) \otimes ( \mathbf{r}_{i} - \mathbf{r}_{COM} )\]</div>
<p>The traceless quadrupole tensor, <span class="math notranslate nohighlight">\(\hat{\mathsf{Q}}\)</span>, is then
taken from:</p>
<div class="math notranslate nohighlight">
\[\hat{\mathsf{Q}} = \frac{3}{2} \mathsf{Q} - \frac{1}{2}
tr(\mathsf{Q})\]</div>
<p>Computes the quadrupole tensor of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Tensor per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is an unsymmetrical plane in the molecule or
group, the magnitude of the quadrupole tensor is dependent on the
<code class="docutils literal notranslate"><span class="pre">center</span></code> (e.g., <span class="math notranslate nohighlight">\(\mathbf{r}_{COM}\)</span>) chosen and cannot be translated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single quadrupole value returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 1d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Quadrupole tensor(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
tensor of shape <code class="docutils literal notranslate"><span class="pre">(3,3)</span></code>. Otherwise, the output will be a 1d array
of shape <code class="docutils literal notranslate"><span class="pre">(n,3,3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.radius_of_gyration">
<span class="sig-name descname"><span class="pre">radius_of_gyration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.radius_of_gyration" title="Permalink to this definition"></a></dt>
<dd><p>Radius of gyration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.residues">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">residues</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.residues" title="Permalink to this definition"></a></dt>
<dd><p>The <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> itself.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.copy" title="MDAnalysis.core.groups.ResidueGroup.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a></dt><dd><p>return a true copy of the <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a></p>
</dd>
</dl>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>In previous versions, this returned a copy, but now
the <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> itself is returned. This should
not affect any code but only speed up calculations.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.rotate">
<span class="sig-name descname"><span class="pre">rotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">R</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">0,</span> <span class="pre">0)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.rotate" title="Permalink to this definition"></a></dt>
<dd><p>Apply a rotation matrix <cite>R</cite> to the selection’s coordinates.
<span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is a 3x3 orthogonal matrix that transforms a vector
<span class="math notranslate nohighlight">\(\mathbf{x} \rightarrow \mathbf{x}'\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathsf{R}\mathbf{x}\]</div>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> coordinates are rotated in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>R</strong> (<em>array_like</em>) – 3x3 rotation matrix</p></li>
<li><p><strong>point</strong> (<em>array_like</em><em>, </em><em>optional</em>) – Center of rotation</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>By default, rotates about the origin <code class="docutils literal notranslate"><span class="pre">point=(0,</span> <span class="pre">0,</span> <span class="pre">0)</span></code>. To rotate
a group <code class="docutils literal notranslate"><span class="pre">g</span></code> around its center of geometry, use
<code class="docutils literal notranslate"><span class="pre">g.rotate(R,</span> <span class="pre">point=g.center_of_geometry())</span></code>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.rotateby" title="MDAnalysis.core.groups.ResidueGroup.rotateby"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotateby</span></code></a></dt><dd><p>rotate around given axis and angle</p>
</dd>
<dt><a class="reference internal" href="../lib/transformations.html#module-MDAnalysis.lib.transformations" title="MDAnalysis.lib.transformations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations</span></code></a></dt><dd><p>module of all coordinate transforms</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.rotateby">
<span class="sig-name descname"><span class="pre">rotateby</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">angle</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.rotateby" title="Permalink to this definition"></a></dt>
<dd><p>Apply a rotation to the selection’s coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>angle</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a>) – Rotation angle in degrees.</p></li>
<li><p><strong>axis</strong> (<em>array_like</em>) – Rotation axis vector.</p></li>
<li><p><strong>point</strong> (<em>array_like</em><em>, </em><em>optional</em>) – Center of rotation. If <code class="docutils literal notranslate"><span class="pre">None</span></code> then the center of geometry of this
group is used.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The transformation from current coordinates <span class="math notranslate nohighlight">\(\mathbf{x}\)</span>
to new coordinates <span class="math notranslate nohighlight">\(\mathbf{x}'\)</span> is</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathsf{R}\,(\mathbf{x}-\mathbf{p}) + \mathbf{p}\]</div>
<p>where <span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is the rotation by <cite>angle</cite> around the
<cite>axis</cite> going through <cite>point</cite> <span class="math notranslate nohighlight">\(\mathbf{p}\)</span>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations.rotation_matrix</span></code></p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">calculate</span></code></dt><dd><p>math:<cite>mathsf{R}</cite></p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.segments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">segments</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.segments" title="Permalink to this definition"></a></dt>
<dd><p>Get sorted <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> of the unique segments present in
the <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.sequence">
<span class="sig-name descname"><span class="pre">sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.sequence" title="Permalink to this definition"></a></dt>
<dd><p>Returns the amino acid sequence.</p>
<p>The format of the sequence is selected with the keyword <em>format</em>:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 24%" />
<col style="width: 76%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p><em>format</em></p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>‘SeqRecord’</p></td>
<td><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code> (default)</p></td>
</tr>
<tr class="row-odd"><td><p>‘Seq’</p></td>
<td><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code></p></td>
</tr>
<tr class="row-even"><td><p>‘string’</p></td>
<td><p>string</p></td>
</tr>
</tbody>
</table>
<p>The sequence is returned by default (keyword <code class="docutils literal notranslate"><span class="pre">format</span> <span class="pre">=</span> <span class="pre">'SeqRecord'</span></code>)
as a <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code> instance, which can then be
further processed. In this case, all keyword arguments (such as the
<em>id</em> string or the <em>name</em> or the <em>description</em>) are directly passed to
<code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code>.</p>
<p>If the keyword <em>format</em> is set to <code class="docutils literal notranslate"><span class="pre">'Seq'</span></code>, all <em>kwargs</em> are ignored
and a <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code> instance is returned. The difference to the
record is that the record also contains metadata and can be directly
used as an input for other functions in <code class="xref py py-mod docutils literal notranslate"><span class="pre">Bio</span></code>.</p>
<p>If the keyword <em>format</em> is set to <code class="docutils literal notranslate"><span class="pre">'string'</span></code>, all <em>kwargs</em> are
ignored and a Python string is returned.</p>
<p class="rubric">Example: Write FASTA file</p>
<p>Use <code class="xref py py-func docutils literal notranslate"><span class="pre">Bio.SeqIO.write()</span></code>, which takes sequence records:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">Bio.SeqIO</span>

<span class="c1"># get the sequence record of a protein component of a Universe</span>
<span class="n">protein</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s2">&quot;protein&quot;</span><span class="p">)</span>
<span class="n">record</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="s2">&quot;myseq1&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;myprotein&quot;</span><span class="p">)</span>

<span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="s2">&quot;single.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>A FASTA file with multiple entries can be written with</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="n">record1</span><span class="p">,</span> <span class="n">record2</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span> <span class="s2">&quot;multi.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>format</strong> (<em>string</em><em>, </em><em>optional</em>) – <ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;string&quot;</span></code>: return sequence as a string of 1-letter codes</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;Seq&quot;</span></code>: return a <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code> instance</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;SeqRecord&quot;</span></code>: return a <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code>
instance</p></li>
</ul>
<blockquote>
<div><p>Default is <code class="docutils literal notranslate"><span class="pre">&quot;SeqRecord&quot;</span></code></p>
</div></blockquote>
</p></li>
<li><p><strong>id</strong> (<em>optional</em>) – Sequence ID for SeqRecord (should be different for different
sequences)</p></li>
<li><p><strong>name</strong> (<em>optional</em>) – Name of the protein.</p></li>
<li><p><strong>description</strong> (<em>optional</em>) – Short description of the sequence.</p></li>
<li><p><strong>kwargs</strong> (<em>optional</em>) – Any other keyword arguments that are understood by
class:<cite>Bio.SeqRecord.SeqRecord</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – </p></li>
<li><p><strong>1-letter IUPAC protein amino acid code; make sure to only</strong> – </p></li>
<li><p><strong>select protein residues.</strong> – </p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.11)"><strong>TypeError</strong></a> – </p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have resnames. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.shape_parameter">
<span class="sig-name descname"><span class="pre">shape_parameter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.shape_parameter" title="Permalink to this definition"></a></dt>
<dd><p>Shape parameter.</p>
<p>See <a class="reference internal" href="../references.html#dima2004a" id="id9"><span>[Dima2004a]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.
Superfluous kwargs were removed.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.5.0: </span>Added calculation for any <cite>compound</cite> type</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.subtract">
<span class="sig-name descname"><span class="pre">subtract</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.subtract" title="Permalink to this definition"></a></dt>
<dd><p>Group with elements from this Group that don’t appear in other</p>
<p>The original order of this group is kept, as well as any duplicate
elements. If an element of this Group is duplicated and appears in
the other Group or Component, then all the occurences of that element
are removed from the returned Group.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the elements of <cite>self</cite> that are not in  <cite>other</cite>,
conserves order and duplicates.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Unlike <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.difference" title="MDAnalysis.core.groups.ResidueGroup.difference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">difference()</span></code></a> this method will not sort or remove
duplicates.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">-</span> <span class="n">ag2</span>  <span class="c1"># or ag1.subtract(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span><span class="o">.</span><span class="n">indices</span>
<span class="go">array([3, 3, 1, 1])</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.concatenate" title="MDAnalysis.core.groups.ResidueGroup.concatenate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concatenate</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.difference" title="MDAnalysis.core.groups.ResidueGroup.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.symmetric_difference">
<span class="sig-name descname"><span class="pre">symmetric_difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.symmetric_difference" title="Permalink to this definition"></a></dt>
<dd><p>Group of elements which are only in one of this Group or another</p>
<p>This method removes duplicate elements and the result is sorted. It is
synomym to the <cite>^</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the elements that are in <cite>self</cite> or in <cite>other</cite> but not in
both, without duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">^</span> <span class="n">ag2</span>  <span class="c1"># or ag1.symmetric_difference(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="o">.</span><span class="n">indices</span>  <span class="c1"># 0 and 1 are only in ag1, 4 and 6 are only in ag2</span>
<span class="go">[0, 1, 4, 6]</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.difference" title="MDAnalysis.core.groups.ResidueGroup.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.total_charge">
<span class="sig-name descname"><span class="pre">total_charge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.total_charge" title="Permalink to this definition"></a></dt>
<dd><p>Total charge of (compounds of) the group.</p>
<p>Computes the total charge of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Total charges per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If ‘group’, the total charge of all atoms in the group will
be returned as a single value. Otherwise, the total charges per
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment
will be returned as a 1d array.
Note that, in any case, <em>only</em> the charges of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Total charge of (compounds of) the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.total_mass">
<span class="sig-name descname"><span class="pre">total_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.total_mass" title="Permalink to this definition"></a></dt>
<dd><p>Total mass of (compounds of) the group.</p>
<p>Computes the total mass of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Total masses per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the total mass of all atoms in the group will be
returned as a single value. Otherwise, the total masses per
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment will be
returned as a 1d array.
Note that, in any case, <em>only</em> the masses of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Total mass of (compounds of) the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.transform">
<span class="sig-name descname"><span class="pre">transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">M</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.transform" title="Permalink to this definition"></a></dt>
<dd><p>Apply homogenous transformation matrix <cite>M</cite> to the coordinates.</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> coordinates are rotated and translated in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>M</strong> (<em>array_like</em>) – 4x4 matrix with the rotation in <code class="docutils literal notranslate"><span class="pre">R</span> <span class="pre">=</span> <span class="pre">M[:3,</span> <span class="pre">:3]</span></code> and the
translation in <code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">M[:3,</span> <span class="pre">3]</span></code>.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../lib/transformations.html#module-MDAnalysis.lib.transformations" title="MDAnalysis.lib.transformations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations</span></code></a></dt><dd><p>module of all coordinate transforms</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The rotation <span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is about the origin and is applied
before the translation <span class="math notranslate nohighlight">\(\mathbf{t}\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathsf{R}\mathbf{x} + \mathbf{t}\]</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.translate" title="Permalink to this definition"></a></dt>
<dd><p>Apply translation vector <cite>t</cite> to the selection’s coordinates.</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> coordinates are translated in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>t</strong> (<em>array_like</em>) – vector to translate coordinates with</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../lib/transformations.html#module-MDAnalysis.lib.transformations" title="MDAnalysis.lib.transformations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations</span></code></a></dt><dd><p>module of all coordinate transforms</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The method applies a translation to the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>
from current coordinates <span class="math notranslate nohighlight">\(\mathbf{x}\)</span> to new coordinates
<span class="math notranslate nohighlight">\(\mathbf{x}'\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathbf{x} + \mathbf{t}\]</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.union">
<span class="sig-name descname"><span class="pre">union</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.union" title="Permalink to this definition"></a></dt>
<dd><p>Group of elements either in this Group or another</p>
<p>On the contrary to concatenation, this method sort the elements and
removes duplicate ones. It is synomymous to the <cite>|</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the combined elements of <cite>self</cite> and <cite>other</cite>, without
duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>In contrast to <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.concatenate" title="MDAnalysis.core.groups.ResidueGroup.concatenate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">concatenate()</span></code></a>, any duplicates are dropped
and the result is sorted.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name O&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name N&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">|</span> <span class="n">ag2</span>  <span class="c1"># or ag1.union(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">names</span>
<span class="go">array([&#39;N&#39;, &#39;O&#39;, &#39;N&#39;], dtype=object)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.concatenate" title="MDAnalysis.core.groups.ResidueGroup.concatenate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concatenate</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.intersection" title="MDAnalysis.core.groups.ResidueGroup.intersection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">intersection</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.unique">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">unique</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.unique" title="Permalink to this definition"></a></dt>
<dd><p>Return a <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> containing sorted and unique
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residues</span></code></a> only.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rg</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">residues</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg</span>
<span class="go">&lt;ResidueGroup with 6 residues&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([2, 1, 2, 2, 1, 0])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg2</span> <span class="o">=</span> <span class="n">rg</span><span class="o">.</span><span class="n">unique</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg2</span>
<span class="go">&lt;ResidueGroup with 3 residues&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg2</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([0, 1, 2])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rg2</span><span class="o">.</span><span class="n">unique</span> <span class="ow">is</span> <span class="n">rg2</span>
<span class="go">False</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>If the <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> is already
unique, <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.unique" title="MDAnalysis.core.groups.ResidueGroup.unique"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ResidueGroup.unique</span></code></a> now returns the group itself
instead of a copy.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.0.0: </span>This function now always returns a copy.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.universe">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">universe</span></span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.universe" title="Permalink to this definition"></a></dt>
<dd><p>The underlying <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> the group
belongs to.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.unwrap">
<span class="sig-name descname"><span class="pre">unwrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fragments'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'com'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.unwrap" title="Permalink to this definition"></a></dt>
<dd><p>Move atoms of this group so that bonds within the
group’s compounds aren’t split across periodic boundaries.</p>
<p>This function is most useful when atoms have been packed into the
primary unit cell, causing breaks mid-molecule, with the molecule then
appearing on either side of the unit cell. This is problematic for
operations such as calculating the center of mass of the molecule.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>+-----------+       +-----------+
|           |       |           |
| 6       3 |       |         3 | 6
| !       ! |       |         ! | !
|-5-8   1-2-|  ==&gt;  |       1-2-|-5-8
| !       ! |       |         ! | !
| 7       4 |       |         4 | 7
|           |       |           |
+-----------+       +-----------+
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
Which type of compound to unwrap. Note that, in any case, all
atoms within each compound must be interconnected by bonds, i.e.,
compounds must correspond to (parts of) molecules.</p></li>
<li><p><strong>reference</strong> (<em>{'com'</em><em>, </em><em>'cog'</em><em>, </em><em>None}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'com'</span></code> (center of mass) or <code class="docutils literal notranslate"><span class="pre">'cog'</span></code> (center of geometry), the
unwrapped compounds will be shifted so that their individual
reference point lies within the primary unit cell.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no such shift is performed.</p></li>
<li><p><strong>inplace</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, coordinates are modified in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>coords</strong> – Unwrapped atom coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> but the underlying topology does
    not contain molecule information, or if <cite>reference</cite> is <code class="docutils literal notranslate"><span class="pre">'com'</span></code>
    but the topology does not contain masses.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>reference</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'com'</span></code>, <code class="docutils literal notranslate"><span class="pre">'cog'</span></code>, or <code class="docutils literal notranslate"><span class="pre">None</span></code>, or
    if <cite>reference</cite> is <code class="docutils literal notranslate"><span class="pre">'com'</span></code> and the total mass of any <cite>compound</cite> is
    zero.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Be aware of the fact that only atoms <em>belonging to the group</em> will
be unwrapped! If you want entire molecules to be unwrapped, make sure
that all atoms of these molecules are part of the group.
An AtomGroup containing all atoms of all fragments in the group <code class="docutils literal notranslate"><span class="pre">ag</span></code>
can be created with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">all_frag_atoms</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">ag</span><span class="o">.</span><span class="n">fragments</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="../lib/mdamath.html#MDAnalysis.lib.mdamath.make_whole" title="MDAnalysis.lib.mdamath.make_whole"><code class="xref py py-func docutils literal notranslate"><span class="pre">make_whole()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.wrap" title="MDAnalysis.core.groups.ResidueGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">wrap()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.pack_into_box" title="MDAnalysis.core.groups.ResidueGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pack_into_box()</span></code></a>, <code class="xref py py-func docutils literal notranslate"><span class="pre">apply_PBC()</span></code></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.ResidueGroup.wrap">
<span class="sig-name descname"><span class="pre">wrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'atoms'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'com'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.ResidueGroup.wrap" title="Permalink to this definition"></a></dt>
<dd><p>Shift the contents of this group back into the primary unit cell
according to periodic boundary conditions.</p>
<p>Specifying a <cite>compound</cite> will keep the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in each
compound together during the process. If <cite>compound</cite> is different from
<code class="docutils literal notranslate"><span class="pre">'atoms'</span></code>, each compound as a whole will be shifted so that its
<cite>center</cite> lies within the primary unit cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>compound</strong> (<em>{'atoms'</em><em>, </em><em>'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
Which type of compound to keep together during wrapping. Note that,
in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
<li><p><strong>center</strong> (<em>{'com'</em><em>, </em><em>'cog'}</em>) – How to define the center of a given group of atoms. If <cite>compound</cite> is
<code class="docutils literal notranslate"><span class="pre">'atoms'</span></code>, this parameter is meaningless and therefore ignored.</p></li>
<li><p><strong>box</strong> (<em>array_like</em><em>, </em><em>optional</em>) – The unitcell dimensions of the system, which can be orthogonal or
triclinic and must be provided in the same format as returned by
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.dimensions" title="MDAnalysis.coordinates.timestep.Timestep.dimensions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MDAnalysis.coordinates.timestep.Timestep.dimensions</span></code></a>:
<code class="docutils literal notranslate"><span class="pre">[lx,</span> <span class="pre">ly,</span> <span class="pre">lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, uses the
dimensions of the current time step.</p></li>
<li><p><strong>inplace</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, coordinates will be changed in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array of wrapped atom coordinates of dtype <cite>np.float32</cite> and shape
<code class="docutils literal notranslate"><span class="pre">(len(self.atoms.n_atoms),</span> <span class="pre">3)</span></code></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>compound</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'atoms'</span></code>, <code class="docutils literal notranslate"><span class="pre">'group'</span></code>,
    <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>, <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>.</p></li>
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecule'</span></code> but the topology doesn’t
    contain molecule information (molnums) or if <cite>compound</cite> is
    <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> but the topology doesn’t contain bonds or if
    <cite>center</cite> is <code class="docutils literal notranslate"><span class="pre">'com'</span></code> but the topology doesn’t contain masses.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>All atoms of the group will be moved so that the centers of its
compounds lie within the primary periodic image. For orthorhombic unit
cells, the primary periodic image is defined as the half-open interval
<span class="math notranslate nohighlight">\([0,L_i)\)</span> between <span class="math notranslate nohighlight">\(0\)</span> and boxlength <span class="math notranslate nohighlight">\(L_i\)</span> in all
dimensions <span class="math notranslate nohighlight">\(i\in\{x,y,z\}\)</span>, i.e., the origin of the of the
simulation box is taken to be at the origin <span class="math notranslate nohighlight">\((0,0,0)\)</span> of the
euclidian coordinate system. A compound center residing at position
<span class="math notranslate nohighlight">\(x_i\)</span> in dimension <span class="math notranslate nohighlight">\(i\)</span> will be shifted to <span class="math notranslate nohighlight">\(x_i'\)</span>
according to</p>
<div class="math notranslate nohighlight">
\[x_i' = x_i - \left\lfloor\frac{x_i}{L_i}\right\rfloor\,.\]</div>
<p>When specifying a <cite>compound</cite>, the translation is calculated based on
each compound. The same translation is applied to all atoms
within this compound, meaning it will not be broken by the shift.
This might however mean that not all atoms of a compound will be
inside the unit cell after wrapping, but rather will be the center of
the compound.
Be aware of the fact that only atoms <em>belonging to the group</em> will be
taken into account!</p>
<p><cite>center</cite> allows to define how the center of each group is computed.
This can be either <code class="docutils literal notranslate"><span class="pre">'com'</span></code> for center of mass, or <code class="docutils literal notranslate"><span class="pre">'cog'</span></code> for
center of geometry.</p>
<p><cite>box</cite> allows a unit cell to be given for the transformation. If not
specified, the <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.dimensions" title="MDAnalysis.coordinates.timestep.Timestep.dimensions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dimensions</span></code></a>
information from the current
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a> will be used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.wrap" title="MDAnalysis.core.groups.AtomGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AtomGroup.wrap()</span></code></a> is currently faster than
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.wrap" title="MDAnalysis.core.groups.ResidueGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ResidueGroup.wrap()</span></code></a> or <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.wrap" title="MDAnalysis.core.groups.SegmentGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SegmentGroup.wrap()</span></code></a>.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.pack_into_box" title="MDAnalysis.core.groups.ResidueGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pack_into_box()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.unwrap" title="MDAnalysis.core.groups.ResidueGroup.unwrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">unwrap()</span></code></a>, <a class="reference internal" href="../lib/distances.html#MDAnalysis.lib.distances.apply_PBC" title="MDAnalysis.lib.distances.apply_PBC"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MDAnalysis.lib.distances.apply_PBC()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.2.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>The method only acts on atoms <em>belonging to the group</em> and returns
the wrapped positions as a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a>.
Added optional argument <cite>inplace</cite>.</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.groups.</span></span><span class="sig-name descname"><span class="pre">SegmentGroup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#SegmentGroup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup" title="Permalink to this definition"></a></dt>
<dd><p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> base class.</p>
<p>This class is used by a <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> for
generating its Topology-specific <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> class. All the
<a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.TopologyAttr" title="MDAnalysis.core.topologyattrs.TopologyAttr"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyAttr</span></code></a> components are obtained
from <code class="xref py py-class docutils literal notranslate"><span class="pre">GroupBase</span></code>, so this class only includes ad-hoc methods specific
to <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroups</span></code></a>.</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroups</span></code></a> can be compared and combined using
group operators. See the list of these operators on <code class="xref py py-class docutils literal notranslate"><span class="pre">GroupBase</span></code>.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.16.2: </span><em>Instant selectors</em> of Segments will be removed in the 1.0 release.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed instant selectors, use select_atoms instead</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Indexing an SegmentGroup with <code class="docutils literal notranslate"><span class="pre">None</span></code> raises a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code>.</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.accumulate">
<span class="sig-name descname"><span class="pre">accumulate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attribute</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">function=&lt;function</span> <span class="pre">sum&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound='group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.accumulate" title="Permalink to this definition"></a></dt>
<dd><p>Accumulates the attribute associated with (compounds of) the group.</p>
<p>Accumulates the attribute of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
The accumulation per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule,
or fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. By default, the method sums up all attributes per compound,
but any function that takes an array and returns an acuumulation over a
given axis can be used. For multi-dimensional input arrays, the
accumulation is performed along the first axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> or </em><em>array_like</em>) – Attribute or array of values to accumulate.
If a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> (or compatible) is provided, its first
dimension must have the same length as the total number of atoms in
the group.</p></li>
<li><p><strong>function</strong> (<em>callable</em><em>, </em><em>optional</em>) – The function performing the accumulation. It must take the array of
attribute values to accumulate as its only positional argument and
accept an (optional) keyword argument <code class="docutils literal notranslate"><span class="pre">axis</span></code> allowing to specify
the axis along which the accumulation is performed.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the accumulation of all attributes associated with
atoms in the group will be returned as a single value. Otherwise,
the accumulation of the attributes per <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>,
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment will be returned as a 1d
array. Note that, in any case, <em>only</em> the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Acuumulation of the <cite>attribute</cite>.
If <cite>compound</cite> is set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the first dimension of the
<cite>attribute</cite> array will be contracted to a single value.
If <cite>compound</cite> is set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>, <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>, the length of the first
dimension will correspond to the number of compounds. In all cases,
the other dimensions of the returned array will be of the original
shape (without the first dimension).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If the length of a provided <cite>attribute</cite> array does not correspond to
    the number of atoms in the group.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>compound</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>,
    <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>.</p></li>
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecule'</span></code> but the topology doesn’t
    contain molecule information (molnums), or if <cite>compound</cite> is
    <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> but the topology doesn’t contain bonds.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>To find the total charge of a given <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;prop mass &gt; 4.0&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">accumulate</span><span class="p">(</span><span class="s1">&#39;charges&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>To find the total mass per residue of all CA <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">accumulate</span><span class="p">(</span><span class="s1">&#39;masses&#39;</span><span class="p">,</span> <span class="n">compound</span><span class="o">=</span><span class="s1">&#39;residues&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>To find the maximum atomic charge per fragment of a given
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">accumulate</span><span class="p">(</span><span class="s1">&#39;charges&#39;</span><span class="p">,</span> <span class="n">compound</span><span class="o">=</span><span class="s2">&quot;fragments&quot;</span><span class="p">,</span> <span class="n">function</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">)</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.align_principal_axis">
<span class="sig-name descname"><span class="pre">align_principal_axis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">axis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.align_principal_axis" title="Permalink to this definition"></a></dt>
<dd><p>Align principal axis with index <cite>axis</cite> with <cite>vector</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>{0</em><em>, </em><em>1</em><em>, </em><em>2}</em>) – Index of the principal axis (0, 1, or 2), as produced by
<a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.principal_axes" title="MDAnalysis.core.groups.SegmentGroup.principal_axes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">principal_axes()</span></code></a>.</p></li>
<li><p><strong>vector</strong> (<em>array_like</em>) – Vector to align principal axis with.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>To align the long axis of a channel (the first principal axis, i.e.
<em>axis</em> = 0) with the z-axis:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">align_principal_axis</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;aligned.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.asphericity">
<span class="sig-name descname"><span class="pre">asphericity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.asphericity" title="Permalink to this definition"></a></dt>
<dd><p>Asphericity.</p>
<p>See <a class="reference internal" href="../references.html#dima2004b" id="id10"><span>[Dima2004b]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <em>unwrap</em> and <em>compound</em> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.5.0: </span>Added calculation for any <cite>compound</cite> type</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.asunique">
<span class="sig-name descname"><span class="pre">asunique</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sorted</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#SegmentGroup.asunique"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.asunique" title="Permalink to this definition"></a></dt>
<dd><p>Return a <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> containing unique
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segments</span></code></a> only, with optional sorting.</p>
<p>If the <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> is unique, this is the group itself.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sorted</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Whether or not the returned SegmentGroup should be sorted
by segindex.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Unique <code class="docutils literal notranslate"><span class="pre">SegmentGroup</span></code></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sg</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">segments</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg</span>
<span class="go">&lt;SegmentGroup with 6 segments&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([2, 1, 2, 2, 1, 0])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg2</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">asunique</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg2</span>
<span class="go">&lt;SegmentGroup with 3 segments&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg2</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([0, 1, 2])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg2</span><span class="o">.</span><span class="n">asunique</span><span class="p">()</span> <span class="ow">is</span> <span class="n">sg2</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.0.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.atoms" title="Permalink to this definition"></a></dt>
<dd><p>An <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> present in this
<a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>.</p>
<p>The <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> are ordered locally by <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, which
are further ordered by <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> in the <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>.
Duplicates are <em>not</em> removed.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.bbox">
<span class="sig-name descname"><span class="pre">bbox</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.bbox" title="Permalink to this definition"></a></dt>
<dd><p>Return the bounding box of the selection.</p>
<p>The lengths A,B,C of the orthorhombic enclosing box are</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">L</span> <span class="o">=</span> <span class="n">AtomGroup</span><span class="o">.</span><span class="n">bbox</span><span class="p">()</span>
<span class="n">A</span><span class="p">,</span><span class="n">B</span><span class="p">,</span><span class="n">C</span> <span class="o">=</span> <span class="n">L</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">L</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to the primary unit cell
before calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>corners</strong> – 2x3 array giving corners of bounding box as
<code class="docutils literal notranslate"><span class="pre">[[xmin,</span> <span class="pre">ymin,</span> <span class="pre">zmin],</span> <span class="pre">[xmax,</span> <span class="pre">ymax,</span> <span class="pre">zmax]]</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.2.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.bfactors">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bfactors</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.bfactors" title="Permalink to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have tempfactors. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.bsphere">
<span class="sig-name descname"><span class="pre">bsphere</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.bsphere" title="Permalink to this definition"></a></dt>
<dd><p>Return the bounding sphere of the selection.</p>
<p>The sphere is calculated relative to the
<a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.center_of_geometry" title="MDAnalysis.core.groups.SegmentGroup.center_of_geometry"><code class="xref py py-meth docutils literal notranslate"><span class="pre">center</span> <span class="pre">of</span> <span class="pre">geometry</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms to the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>R</strong> (<em>float</em>) – Radius of the bounding sphere.</p></li>
<li><p><strong>center</strong> (<em>numpy.ndarray</em>) – Coordinates of the sphere center as <code class="docutils literal notranslate"><span class="pre">[xcen,</span> <span class="pre">ycen,</span> <span class="pre">zcen]</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.3.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <em>pbc</em> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.center">
<span class="sig-name descname"><span class="pre">center</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">weights</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.center" title="Permalink to this definition"></a></dt>
<dd><p>Weighted center of (compounds of) the group</p>
<p>Computes the weighted center of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Weighted centers per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the weights of a compound sum up to zero, the
coordinates of that compound’s weighted center will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>weights</strong> (<em>array_like</em><em> or </em><em>None</em>) – Weights to be used. Setting <cite>weights=None</cite> is equivalent to passing
identical weights for all atoms of the group.</p></li>
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">'group'</span></code> the center of each
compound will be calculated without moving any
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to keep the compounds intact. Instead, the
resulting position vectors will be moved to the primary unit cell
after calculation. Default [<code class="docutils literal notranslate"><span class="pre">False</span></code>].</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – <dl class="simple">
<dt>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their</dt><dd><p>centers.</p>
</dd>
</dl>
</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the weighted center of all atoms in the group will
be returned as a single position vector. Else, the weighted centers
of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment
will be returned as an array of position vectors, i.e. a 2d array.
Note that, in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the weighted center(s) of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>, <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>, the output will be a 2d array
of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>compound</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>,
    <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If both ‘wrap’ and ‘unwrap’ set to true.</p></li>
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecule'</span></code> but the topology doesn’t
    contain molecule information (molnums) or if <cite>compound</cite> is
    <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> but the topology doesn’t contain bonds.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>To find the center of charge of a given <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;prop mass &gt; 4.0&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">center</span><span class="p">(</span><span class="n">sel</span><span class="o">.</span><span class="n">charges</span><span class="p">)</span>
</pre></div>
</div>
<p>To find the centers of mass per residue of all CA <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sel</span><span class="o">.</span><span class="n">center</span><span class="p">(</span><span class="n">sel</span><span class="o">.</span><span class="n">masses</span><span class="p">,</span> <span class="n">compound</span><span class="o">=</span><span class="s1">&#39;residues&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>
compounds</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.center_of_charge">
<span class="sig-name descname"><span class="pre">center_of_charge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.center_of_charge" title="Permalink to this definition"></a></dt>
<dd><p>Center of (absolute) charge of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \vert q_i \vert \boldsymbol r_i}
                     {\sum_i \vert q_i \vert}\]</div>
<p>where <span class="math notranslate nohighlight">\(q_i\)</span> is the charge and <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> the
position of atom <span class="math notranslate nohighlight">\(i\)</span> in the given
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup</span></code></a>.
Centers of charge per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the charges of a compound sum up to zero, the
center of mass coordinates of that compound will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of mass of all atoms in the group will
be returned as a single position vector. Otherwise, the centers of
mass of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the center(s) of charge of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where
<code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method can only be accessed if the underlying topology has
information about atomic charges.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.2.0.</span></p>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.center_of_geometry">
<span class="sig-name descname"><span class="pre">center_of_geometry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.center_of_geometry" title="Permalink to this definition"></a></dt>
<dd><p>Center of geometry of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \boldsymbol r_i}{\sum_i 1}\]</div>
<p>where <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <span class="math notranslate nohighlight">\(i\)</span>.
Centers of geometry per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> or per <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> can
be obtained by setting the <cite>compound</cite> parameter accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation. If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is
<code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the center of each compound will
be calculated without moving any <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to keep the
compounds intact. Instead, the resulting position vectors will be
moved to the primary unit cell after calculation. Default False.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of geometry of all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
in the group will be returned as a single position vector. Else,
the centers of geometry of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> will be returned as an array of position vectors,
i.e. a 2d array. Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the geometric center(s) of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the
number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>
compounds</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.center_of_mass">
<span class="sig-name descname"><span class="pre">center_of_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.center_of_mass" title="Permalink to this definition"></a></dt>
<dd><p>Center of mass of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i m_i \boldsymbol r_i}{\sum m_i}\]</div>
<p>where <span class="math notranslate nohighlight">\(m_i\)</span> is the mass and <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> the
position of atom <span class="math notranslate nohighlight">\(i\)</span> in the given
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup</span></code></a>.
Centers of mass per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly. If the masses of a compound sum up to zero, the
center of mass coordinates of that compound will be <code class="docutils literal notranslate"><span class="pre">nan</span></code> (not a
number).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of mass of all atoms in the group will
be returned as a single position vector. Otherwise, the centers of
mass of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e. a 2d
array.
Note that, in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the center(s) of mass of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code>
is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method can only be accessed if the underlying topology has
information about atomic masses.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> compounds;
added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.centroid">
<span class="sig-name descname"><span class="pre">centroid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.centroid" title="Permalink to this definition"></a></dt>
<dd><p>Center of geometry of (compounds of) the group</p>
<div class="math notranslate nohighlight">
\[\boldsymbol R = \frac{\sum_i \boldsymbol r_i}{\sum_i 1}\]</div>
<p>where <span class="math notranslate nohighlight">\(\boldsymbol r_i\)</span> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <span class="math notranslate nohighlight">\(i\)</span>.
Centers of geometry per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> or per <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> can
be obtained by setting the <cite>compound</cite> parameter accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation. If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is
<code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the center of each compound will
be calculated without moving any <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> to keep the
compounds intact. Instead, the resulting position vectors will be
moved to the primary unit cell after calculation. Default False.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the center of geometry of all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
in the group will be returned as a single position vector. Else,
the centers of geometry of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> will be returned as an array of position vectors,
i.e. a 2d array. Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>center</strong> – Position vector(s) of the geometric center(s) of the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
position vector.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'segments'</span></code> or <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, the
output will be a 2d array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the
number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> and <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>
compounds</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed flags affecting default behaviour</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.concatenate">
<span class="sig-name descname"><span class="pre">concatenate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.concatenate" title="Permalink to this definition"></a></dt>
<dd><p>Concatenate with another Group or Component of the same level.</p>
<p>Duplicate entries and original order is preserved. It is synomymous to
the <cite>+</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with elements of <cite>self</cite> and <cite>other</cite> concatenated</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>The order of the original contents (including duplicates)
are preserved when performing a concatenation.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name O&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name N&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">+</span> <span class="n">ag2</span>  <span class="c1"># or ag1.concatenate(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">names</span>
<span class="go">array([&#39;O&#39;, &#39;O&#39;, &#39;O&#39;], dtype=object)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]</span><span class="o">.</span><span class="n">names</span>
<span class="go">array([&#39;N&#39;, &#39;N&#39;, &#39;N&#39;], dtype=object)</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.copy" title="Permalink to this definition"></a></dt>
<dd><p>Get another group identical to this one.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.19.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.difference">
<span class="sig-name descname"><span class="pre">difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.difference" title="Permalink to this definition"></a></dt>
<dd><p>Elements from this Group that do not appear in another</p>
<p>This method removes duplicate elements and sorts the result. As such,
it is different from <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.subtract" title="MDAnalysis.core.groups.SegmentGroup.subtract"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subtract()</span></code></a>. <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.difference" title="MDAnalysis.core.groups.SegmentGroup.difference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">difference()</span></code></a> is synomymous
to the <cite>-</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the elements of <cite>self</cite> that are not in  <cite>other</cite>, without
duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.subtract" title="MDAnalysis.core.groups.SegmentGroup.subtract"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subtract</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.symmetric_difference" title="MDAnalysis.core.groups.SegmentGroup.symmetric_difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">symmetric_difference</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.dimensions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dimensions</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.dimensions" title="Permalink to this definition"></a></dt>
<dd><p>Obtain a copy of the dimensions of the currently loaded Timestep</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.dipole_moment">
<span class="sig-name descname"><span class="pre">dipole_moment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.dipole_moment" title="Permalink to this definition"></a></dt>
<dd><p>Dipole moment of the group or compounds in a group.</p>
<div class="math notranslate nohighlight">
\[\mu = |\boldsymbol{\mu}| = \sqrt{ \sum_{i=1}^{D} \mu^2 }\]</div>
<p>Where <span class="math notranslate nohighlight">\(D\)</span> is the number of dimensions, in this case 3.</p>
<p>Computes the dipole moment of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Dipole per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is a net charge, the magnitude of the dipole
moment is dependent on the <cite>center</cite> chosen.
See <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.dipole_vector" title="MDAnalysis.core.groups.SegmentGroup.dipole_vector"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dipole_vector()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single dipole vector returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dipole moment(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.dipole_vector">
<span class="sig-name descname"><span class="pre">dipole_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mass'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.dipole_vector" title="Permalink to this definition"></a></dt>
<dd><p>Dipole vector of the group.</p>
<div class="math notranslate nohighlight">
\[\boldsymbol{\mu} = \sum_{i=1}^{N} q_{i} ( \mathbf{r}_{i} -
\mathbf{r}_{COM} )\]</div>
<p>Computes the dipole vector of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Dipole vector per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that the magnitude of the dipole moment is independent of the
<code class="docutils literal notranslate"><span class="pre">center</span></code> chosen unless the species has a net charge. In the case of
a charged group the dipole moment can be later adjusted  with:</p>
<div class="math notranslate nohighlight">
\[\boldsymbol{\mu}_{COC} = \boldsymbol{\mu}_{COM} +
q_{ag}\mathbf{r}_{COM} - q_{ag}\boldsymbol{r}_{COC}\]</div>
<p>Where <span class="math notranslate nohighlight">\(\mathbf{r}_{COM}\)</span> is the center of mass and
<span class="math notranslate nohighlight">\(\mathbf{r}_{COC}\)</span> is the center of charge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single dipole vector returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 2d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dipole vector(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,3)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.get_connections">
<span class="sig-name descname"><span class="pre">get_connections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">typename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outside</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.get_connections" title="Permalink to this definition"></a></dt>
<dd><p>Get bonded connections between atoms as a
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.TopologyGroup" title="MDAnalysis.core.topologyobjects.TopologyGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyGroup</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>typename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a>) – group name. One of {“bonds”, “angles”, “dihedrals”,
“impropers”, “ureybradleys”, “cmaps”}</p></li>
<li><p><strong>outside</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Whether to include connections involving atoms outside
this group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>TopologyGroup</em> – containing the bonded group of choice, i.e. bonds, angles,
dihedrals, impropers, ureybradleys or cmaps.</p></li>
<li><p><em>.. versionadded:: 1.1.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.groupby">
<span class="sig-name descname"><span class="pre">groupby</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">topattrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.groupby" title="Permalink to this definition"></a></dt>
<dd><p>Group together items in this group according to values of <em>topattr</em></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>topattrs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.11)"><em>list</em></a>) – One or more topology attributes to group components by.
Single arguments are passed as a string. Multiple arguments
are passed as a list.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Unique values of the multiple combinations of topology attributes
as keys, Groups as values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.11)">dict</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>To group atoms with the same mass together:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;masses&#39;</span><span class="p">)</span>
<span class="go">{12.010999999999999: &lt;AtomGroup with 462 atoms&gt;,</span>
<span class="go"> 14.007: &lt;AtomGroup with 116 atoms&gt;,</span>
<span class="go"> 15.999000000000001: &lt;AtomGroup with 134 atoms&gt;}</span>
</pre></div>
</div>
<p>To group atoms with the same residue name and mass together:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;resnames&#39;</span><span class="p">,</span> <span class="s1">&#39;masses&#39;</span><span class="p">])</span>
<span class="go">{(&#39;ALA&#39;, 1.008): &lt;AtomGroup with 95 atoms&gt;,</span>
<span class="go"> (&#39;ALA&#39;, 12.011): &lt;AtomGroup with 57 atoms&gt;,</span>
<span class="go"> (&#39;ALA&#39;, 14.007): &lt;AtomGroup with 19 atoms&gt;,</span>
<span class="go"> (&#39;ALA&#39;, 15.999): &lt;AtomGroup with 19 atoms&gt;},</span>
<span class="go"> (&#39;ARG&#39;, 1.008): &lt;AtomGroup with 169 atoms&gt;,</span>
<span class="go"> ...}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;resnames&#39;</span><span class="p">,</span> <span class="s1">&#39;masses&#39;</span><span class="p">])(</span><span class="s1">&#39;ALA&#39;</span><span class="p">,</span> <span class="mf">15.999</span><span class="p">)</span>
<span class="go"> &lt;AtomGroup with 19 atoms&gt;</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.18.0: </span>The function accepts multiple attributes</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.gyration_moments">
<span class="sig-name descname"><span class="pre">gyration_moments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.gyration_moments" title="Permalink to this definition"></a></dt>
<dd><p>Moments of the gyration tensor.</p>
<p>The moments are defined as the eigenvalues of the gyration
tensor.</p>
<div class="math notranslate nohighlight">
\[\mathsf{T} = \frac{1}{N} \sum_{i=1}^{N} (\mathbf{r}_\mathrm{i} -
    \mathbf{r}_\mathrm{COM})(\mathbf{r}_\mathrm{i} - \mathbf{r}_\mathrm{COM})\]</div>
<p>Where <span class="math notranslate nohighlight">\(\mathbf{r}_\mathrm{COM}\)</span> is the center of mass.</p>
<p>See <a class="reference internal" href="../references.html#dima2004a" id="id11"><span>[Dima2004a]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>principle_moments_of_gyration</strong> – Gyration vector(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(Å^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
vector of length 3. Otherwise, the output will be a 2D array of shape
<code class="docutils literal notranslate"><span class="pre">(n,3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.5.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.intersection">
<span class="sig-name descname"><span class="pre">intersection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.intersection" title="Permalink to this definition"></a></dt>
<dd><p>Group of elements which are in both this Group and another</p>
<p>This method removes duplicate elements and sorts the result. It is
synomymous to the <cite>&amp;</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the common elements of <cite>self</cite> and <cite>other</cite>, without
duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Intersections can be used when the select atoms string would
become too complicated.  For example to find the water atoms
which are within 4.0A of two segments:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">shell1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;resname SOL and around 4.0 segid 1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">shell2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;resname SOL and around 4.0 segid 2&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">common</span> <span class="o">=</span> <span class="n">shell1</span> <span class="o">&amp;</span> <span class="n">shell2</span>  <span class="c1"># or shell1.intersection(shell2)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.union" title="MDAnalysis.core.groups.SegmentGroup.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">union</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.is_strict_subset">
<span class="sig-name descname"><span class="pre">is_strict_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.is_strict_subset" title="Permalink to this definition"></a></dt>
<dd><p>If this Group is a subset of another Group but not identical</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a strict subset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.is_strict_superset">
<span class="sig-name descname"><span class="pre">is_strict_superset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.is_strict_superset" title="Permalink to this definition"></a></dt>
<dd><p>If this Group is a superset of another Group but not identical</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a strict superset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.isdisjoint">
<span class="sig-name descname"><span class="pre">isdisjoint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.isdisjoint" title="Permalink to this definition"></a></dt>
<dd><p>If the Group has no elements in common with the other Group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if the two Groups do not have common elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.issubset">
<span class="sig-name descname"><span class="pre">issubset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.issubset" title="Permalink to this definition"></a></dt>
<dd><p>If all elements of this Group are part of another Group</p>
<p>Note that an empty group is a subset of any group of the same level.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a subset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.issuperset">
<span class="sig-name descname"><span class="pre">issuperset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.issuperset" title="Permalink to this definition"></a></dt>
<dd><p>If all elements of another Group are part of this Group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this Group is a subset of the other one</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.isunique">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">isunique</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.isunique" title="Permalink to this definition"></a></dt>
<dd><p>Boolean indicating whether all components of the group are unique,
i.e., the group contains no duplicates.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span>
<span class="go">&lt;AtomGroup with 6 atoms&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag</span><span class="o">.</span><span class="n">isunique</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">ag</span><span class="o">.</span><span class="n">unique</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span>
<span class="go">&lt;AtomGroup with 3 atoms&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span><span class="o">.</span><span class="n">isunique</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.asunique" title="MDAnalysis.core.groups.SegmentGroup.asunique"><code class="xref py py-obj docutils literal notranslate"><span class="pre">asunique</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.19.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.ix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.ix" title="Permalink to this definition"></a></dt>
<dd><p>Unique indices of the components in the Group.</p>
<ul class="simple">
<li><p>If this Group is an <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>, these are the
indices of the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> instances.</p></li>
<li><p>If it is a <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a>, these are the indices of
the <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> instances.</p></li>
<li><p>If it is a <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>, these are the indices of
the <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> instances.</p></li>
</ul>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.ix_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix_array</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.ix_array" title="Permalink to this definition"></a></dt>
<dd><p>Unique indices of the components in the Group.</p>
<p>For a Group, <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.ix_array" title="MDAnalysis.core.groups.SegmentGroup.ix_array"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ix_array</span></code></a> is the same as <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.ix" title="MDAnalysis.core.groups.SegmentGroup.ix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ix</span></code></a>. This method
gives a consistent API between components and groups.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.ix" title="MDAnalysis.core.groups.SegmentGroup.ix"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ix</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.moment_of_inertia">
<span class="sig-name descname"><span class="pre">moment_of_inertia</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.moment_of_inertia" title="Permalink to this definition"></a></dt>
<dd><p>Moment of inertia tensor relative to center of mass.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact. Instead, the resulting
center-of-mass position vectors will be moved to the primary unit
cell after calculation.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers and tensor of inertia.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
<cite>compound</cite> determines the behavior of <cite>wrap</cite>.
Note that, in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>moment_of_inertia</strong> – Moment of inertia tensor as a 3 x 3 numpy array.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The moment of inertia tensor <span class="math notranslate nohighlight">\(\mathsf{I}\)</span> is calculated for a group of
<span class="math notranslate nohighlight">\(N\)</span> atoms with coordinates <span class="math notranslate nohighlight">\(\mathbf{r}_i,\ 1 \le i \le N\)</span>
relative to its center of mass from the relative coordinates</p>
<div class="math notranslate nohighlight">
\[\mathbf{r}'_i = \mathbf{r}_i - \frac{1}{\sum_{i=1}^{N} m_i} \sum_{i=1}^{N} m_i \mathbf{r}_i\]</div>
<p>as</p>
<div class="math notranslate nohighlight">
\[\mathsf{I} = \sum_{i=1}^{N} m_i \Big[(\mathbf{r}'_i\cdot\mathbf{r}'_i) \sum_{\alpha=1}^{3}
      \hat{\mathbf{e}}_\alpha \otimes \hat{\mathbf{e}}_\alpha - \mathbf{r}'_i \otimes \mathbf{r}'_i\Big]\]</div>
<p>where <span class="math notranslate nohighlight">\(\hat{\mathbf{e}}_\alpha\)</span> are Cartesian unit vectors, or in Cartesian coordinates</p>
<div class="math notranslate nohighlight">
\[I_{\alpha,\beta} = \sum_{k=1}^{N} m_k
      \Big(\big(\sum_{\gamma=1}^3 (x'^{(k)}_{\gamma})^2 \big)\delta_{\alpha,\beta}
      - x'^{(k)}_{\alpha} x'^{(k)}_{\beta} \Big).\]</div>
<p>where <span class="math notranslate nohighlight">\(x'^{(k)}_{\alpha}\)</span> are the Cartesian coordinates of the
relative coordinates <span class="math notranslate nohighlight">\(\mathbf{r}'_k\)</span>.</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>unwrap</cite> parameter</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.n_atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.n_atoms" title="Permalink to this definition"></a></dt>
<dd><p>Number of atoms present in the <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>, including
duplicate segments (and thus, duplicate atoms).</p>
<p>Equivalent to <code class="docutils literal notranslate"><span class="pre">len(self.atoms)</span></code>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.n_residues">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_residues</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.n_residues" title="Permalink to this definition"></a></dt>
<dd><p>Number of residues present in this <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>, including
duplicate segments (and thus, residues).</p>
<p>Equivalent to <code class="docutils literal notranslate"><span class="pre">len(self.residues)</span></code>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.n_segments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n_segments</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.n_segments" title="Permalink to this definition"></a></dt>
<dd><p>Number of segments in the <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>.</p>
<p>Equivalent to <code class="docutils literal notranslate"><span class="pre">len(self)</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.pack_into_box">
<span class="sig-name descname"><span class="pre">pack_into_box</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.pack_into_box" title="Permalink to this definition"></a></dt>
<dd><p>Shift all <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in this group to the primary unit
cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>box</strong> (<em>array_like</em>) – Box dimensions, can be either orthogonal or triclinic information.
Cell dimensions must be in an identical to format to those returned
by <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.dimensions" title="MDAnalysis.coordinates.timestep.Timestep.dimensions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MDAnalysis.coordinates.timestep.Timestep.dimensions</span></code></a>,
<code class="docutils literal notranslate"><span class="pre">[lx,</span> <span class="pre">ly,</span> <span class="pre">lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, uses these
timestep dimensions.</p></li>
<li><p><strong>inplace</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a>) – <code class="docutils literal notranslate"><span class="pre">True</span></code> to change coordinates in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>coords</strong> – Shifted atom coordinates.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>All atoms will be moved so that they lie between 0 and boxlength
<span class="math notranslate nohighlight">\(L_i\)</span> in all dimensions, i.e. the lower left corner of the
simulation box is taken to be at (0,0,0):</p>
<div class="math notranslate nohighlight">
\[x_i' = x_i - \left\lfloor\frac{x_i}{L_i}\right\rfloor\]</div>
<p>The default is to take unit cell information from the underlying
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a> instance. The optional
argument <cite>box</cite> can be used to provide alternative unit cell information
(in the MDAnalysis standard format
<code class="docutils literal notranslate"><span class="pre">[Lx,</span> <span class="pre">Ly,</span> <span class="pre">Lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>).</p>
<p>Works with either orthogonal or triclinic box types.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.pack_into_box" title="MDAnalysis.core.groups.SegmentGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pack_into_box()</span></code></a> is identical to <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.wrap" title="MDAnalysis.core.groups.SegmentGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">wrap()</span></code></a> with all default
keywords.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.pack_into_box" title="MDAnalysis.core.groups.AtomGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AtomGroup.pack_into_box()</span></code></a> is currently faster than
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.pack_into_box" title="MDAnalysis.core.groups.ResidueGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ResidueGroup.pack_into_box()</span></code></a> or
<a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.pack_into_box" title="MDAnalysis.core.groups.SegmentGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SegmentGroup.pack_into_box()</span></code></a>.</p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.8.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.principal_axes">
<span class="sig-name descname"><span class="pre">principal_axes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.principal_axes" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the principal axes from the moment of inertia.</p>
<p>e1,e2,e3 = AtomGroup.principal_axes()</p>
<p>The eigenvectors are sorted by eigenvalue, i.e. the first one
corresponds to the highest eigenvalue and is thus the first principal
axes.</p>
<p>The eigenvectors form a right-handed coordinate system.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>axis_vectors</strong> – 3 x 3 array with <code class="docutils literal notranslate"><span class="pre">v[0]</span></code> as first, <code class="docutils literal notranslate"><span class="pre">v[1]</span></code> as second, and
<code class="docutils literal notranslate"><span class="pre">v[2]</span></code> as third eigenvector.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Always return principal axes in right-hand convention.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.quadrupole_moment">
<span class="sig-name descname"><span class="pre">quadrupole_moment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.quadrupole_moment" title="Permalink to this definition"></a></dt>
<dd><p>Quadrupole moment of the group according to <span id="id12">[<a class="reference internal" href="topologyattrs.html#id2" title="C. G. Gray, Keith E. Gubbins, and C. G. Joslin. Theory of molecular fluids. Number 9 in The International series of monographs on chemistry. Oxford University Press, Oxford ; New York, 1984. ISBN 978-0-19-855602-2.">Gray1984</a>]</span>.</p>
<div class="math notranslate nohighlight">
\[Q = \sqrt{\frac{2}{3}{\hat{\mathsf{Q}}}:{\hat{\mathsf{Q}}}}\]</div>
<p>where the quadrupole moment is calculated from the tensor double
contraction of the traceless quadropole tensor <span class="math notranslate nohighlight">\(\hat{\mathsf{Q}}\)</span></p>
<p>Computes the quadrupole moment of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Quadrupole per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is an unsymmetrical plane in the molecule or
group, the magnitude of the quadrupole moment is dependant on the
<code class="docutils literal notranslate"><span class="pre">center</span></code> chosen and cannot be translated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single quadrupole value returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 1d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Quadrupole moment(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.quadrupole_tensor">
<span class="sig-name descname"><span class="pre">quadrupole_tensor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mass'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.quadrupole_tensor" title="Permalink to this definition"></a></dt>
<dd><p>Traceless quadrupole tensor of the group or compounds.</p>
<p>This tensor is first computed as the outer product of vectors from
a reference point to some atom, and multiplied by the atomic charge.
The tensor of each atom is then summed to produce the quadrupole
tensor of the group:</p>
<div class="math notranslate nohighlight">
\[\mathsf{Q} = \sum_{i=1}^{N} q_{i} ( \mathbf{r}_{i} -
\mathbf{r}_{COM} ) \otimes ( \mathbf{r}_{i} - \mathbf{r}_{COM} )\]</div>
<p>The traceless quadrupole tensor, <span class="math notranslate nohighlight">\(\hat{\mathsf{Q}}\)</span>, is then
taken from:</p>
<div class="math notranslate nohighlight">
\[\hat{\mathsf{Q}} = \frac{3}{2} \mathsf{Q} - \frac{1}{2}
tr(\mathsf{Q})\]</div>
<p>Computes the quadrupole tensor of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Tensor per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<p>Note that when there is an unsymmetrical plane in the molecule or
group, the magnitude of the quadrupole tensor is dependent on the
<code class="docutils literal notranslate"><span class="pre">center</span></code> (e.g., <span class="math notranslate nohighlight">\(\mathbf{r}_{COM}\)</span>) chosen and cannot be translated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, move all atoms to the
primary unit cell before calculation.
If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <cite>compound</cite> is not <code class="docutils literal notranslate"><span class="pre">group</span></code>, the
centers of mass of each compound will be calculated without moving
any atoms to keep the compounds intact.</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their
centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, a single quadrupole value returns. Otherwise, an
array of each <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or
fragment will be returned as an array of position vectors, i.e.
a 1d array.
Note that, in any case, <em>only</em> the positions of
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> <em>belonging to the group</em> will be taken into
account.</p></li>
<li><p><strong>center</strong> (<em>{'mass'</em><em>, </em><em>'charge'}</em><em>, </em><em>optional</em>) – Choose whether the dipole vector is calculated at the center of
“mass” or the center of “charge”, default is “mass”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Quadrupole tensor(s) of (compounds of) the group in <span class="math notranslate nohighlight">\(eÅ^2\)</span>.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
tensor of shape <code class="docutils literal notranslate"><span class="pre">(3,3)</span></code>. Otherwise, the output will be a 1d array
of shape <code class="docutils literal notranslate"><span class="pre">(n,3,3)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 2.4.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.radius_of_gyration">
<span class="sig-name descname"><span class="pre">radius_of_gyration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.radius_of_gyration" title="Permalink to this definition"></a></dt>
<dd><p>Radius of gyration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.residues">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">residues</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.residues" title="Permalink to this definition"></a></dt>
<dd><p>A <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> of <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residues</span></code></a> present in this
<a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>.</p>
<p>The <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residues</span></code></a> are ordered locally by
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> in the <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a>. Duplicates are <em>not</em>
removed.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.rotate">
<span class="sig-name descname"><span class="pre">rotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">R</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">0,</span> <span class="pre">0)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.rotate" title="Permalink to this definition"></a></dt>
<dd><p>Apply a rotation matrix <cite>R</cite> to the selection’s coordinates.
<span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is a 3x3 orthogonal matrix that transforms a vector
<span class="math notranslate nohighlight">\(\mathbf{x} \rightarrow \mathbf{x}'\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathsf{R}\mathbf{x}\]</div>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> coordinates are rotated in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>R</strong> (<em>array_like</em>) – 3x3 rotation matrix</p></li>
<li><p><strong>point</strong> (<em>array_like</em><em>, </em><em>optional</em>) – Center of rotation</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>By default, rotates about the origin <code class="docutils literal notranslate"><span class="pre">point=(0,</span> <span class="pre">0,</span> <span class="pre">0)</span></code>. To rotate
a group <code class="docutils literal notranslate"><span class="pre">g</span></code> around its center of geometry, use
<code class="docutils literal notranslate"><span class="pre">g.rotate(R,</span> <span class="pre">point=g.center_of_geometry())</span></code>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.rotateby" title="MDAnalysis.core.groups.SegmentGroup.rotateby"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotateby</span></code></a></dt><dd><p>rotate around given axis and angle</p>
</dd>
<dt><a class="reference internal" href="../lib/transformations.html#module-MDAnalysis.lib.transformations" title="MDAnalysis.lib.transformations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations</span></code></a></dt><dd><p>module of all coordinate transforms</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.rotateby">
<span class="sig-name descname"><span class="pre">rotateby</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">angle</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.rotateby" title="Permalink to this definition"></a></dt>
<dd><p>Apply a rotation to the selection’s coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>angle</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a>) – Rotation angle in degrees.</p></li>
<li><p><strong>axis</strong> (<em>array_like</em>) – Rotation axis vector.</p></li>
<li><p><strong>point</strong> (<em>array_like</em><em>, </em><em>optional</em>) – Center of rotation. If <code class="docutils literal notranslate"><span class="pre">None</span></code> then the center of geometry of this
group is used.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The transformation from current coordinates <span class="math notranslate nohighlight">\(\mathbf{x}\)</span>
to new coordinates <span class="math notranslate nohighlight">\(\mathbf{x}'\)</span> is</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathsf{R}\,(\mathbf{x}-\mathbf{p}) + \mathbf{p}\]</div>
<p>where <span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is the rotation by <cite>angle</cite> around the
<cite>axis</cite> going through <cite>point</cite> <span class="math notranslate nohighlight">\(\mathbf{p}\)</span>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations.rotation_matrix</span></code></p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">calculate</span></code></dt><dd><p>math:<cite>mathsf{R}</cite></p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.segments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">segments</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.segments" title="Permalink to this definition"></a></dt>
<dd><p>The <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> itself.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.copy" title="MDAnalysis.core.groups.SegmentGroup.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a></dt><dd><p>return a true copy of the <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a></p>
</dd>
</dl>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>In previous versions, this returned a copy, but now
the <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> itself is returned. This should
not affect any code but only speed up calculations.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.shape_parameter">
<span class="sig-name descname"><span class="pre">shape_parameter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unwrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.shape_parameter" title="Permalink to this definition"></a></dt>
<dd><p>Shape parameter.</p>
<p>See <a class="reference internal" href="../references.html#dima2004a" id="id13"><span>[Dima2004a]</span></a> for background information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, move all atoms within the primary unit cell before
calculation. [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p></li>
<li><p><strong>unwrap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, compounds will be unwrapped before computing their centers.</p></li>
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>, </em><em>optional</em>) – Which type of component to keep together during unwrapping.</p></li>
</ul>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.7.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added <cite>pbc</cite> keyword</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.1.0: </span>Renamed <cite>pbc</cite> kwarg to <cite>wrap</cite>. <cite>pbc</cite> is still accepted but
is deprecated and will be removed in version 3.0.
Superfluous kwargs were removed.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.5.0: </span>Added calculation for any <cite>compound</cite> type</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.subtract">
<span class="sig-name descname"><span class="pre">subtract</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.subtract" title="Permalink to this definition"></a></dt>
<dd><p>Group with elements from this Group that don’t appear in other</p>
<p>The original order of this group is kept, as well as any duplicate
elements. If an element of this Group is duplicated and appears in
the other Group or Component, then all the occurences of that element
are removed from the returned Group.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the elements of <cite>self</cite> that are not in  <cite>other</cite>,
conserves order and duplicates.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Unlike <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.difference" title="MDAnalysis.core.groups.SegmentGroup.difference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">difference()</span></code></a> this method will not sort or remove
duplicates.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">-</span> <span class="n">ag2</span>  <span class="c1"># or ag1.subtract(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span><span class="o">.</span><span class="n">indices</span>
<span class="go">array([3, 3, 1, 1])</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.concatenate" title="MDAnalysis.core.groups.SegmentGroup.concatenate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concatenate</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.difference" title="MDAnalysis.core.groups.SegmentGroup.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.symmetric_difference">
<span class="sig-name descname"><span class="pre">symmetric_difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.symmetric_difference" title="Permalink to this definition"></a></dt>
<dd><p>Group of elements which are only in one of this Group or another</p>
<p>This method removes duplicate elements and the result is sorted. It is
synomym to the <cite>^</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the elements that are in <cite>self</cite> or in <cite>other</cite> but not in
both, without duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">^</span> <span class="n">ag2</span>  <span class="c1"># or ag1.symmetric_difference(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="o">.</span><span class="n">indices</span>  <span class="c1"># 0 and 1 are only in ag1, 4 and 6 are only in ag2</span>
<span class="go">[0, 1, 4, 6]</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.difference" title="MDAnalysis.core.groups.SegmentGroup.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.total_charge">
<span class="sig-name descname"><span class="pre">total_charge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.total_charge" title="Permalink to this definition"></a></dt>
<dd><p>Total charge of (compounds of) the group.</p>
<p>Computes the total charge of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Total charges per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If ‘group’, the total charge of all atoms in the group will
be returned as a single value. Otherwise, the total charges per
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment
will be returned as a 1d array.
Note that, in any case, <em>only</em> the charges of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Total charge of (compounds of) the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have charges. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.total_mass">
<span class="sig-name descname"><span class="pre">total_mass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.total_mass" title="Permalink to this definition"></a></dt>
<dd><p>Total mass of (compounds of) the group.</p>
<p>Computes the total mass of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the group.
Total masses per <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, molecule, or
fragment can be obtained by setting the <cite>compound</cite> parameter
accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em><em>'fragments'}</em><em>,</em>) – optional
If <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the total mass of all atoms in the group will be
returned as a single value. Otherwise, the total masses per
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, molecule, or fragment will be
returned as a 1d array.
Note that, in any case, <em>only</em> the masses of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Total mass of (compounds of) the group.
If <cite>compound</cite> was set to <code class="docutils literal notranslate"><span class="pre">'group'</span></code>, the output will be a single
value. Otherwise, the output will be a 1d array of shape <code class="docutils literal notranslate"><span class="pre">(n,)</span></code>
where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of compounds.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)">float</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>Added <cite>compound</cite> parameter</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have masses. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.transform">
<span class="sig-name descname"><span class="pre">transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">M</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.transform" title="Permalink to this definition"></a></dt>
<dd><p>Apply homogenous transformation matrix <cite>M</cite> to the coordinates.</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> coordinates are rotated and translated in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>M</strong> (<em>array_like</em>) – 4x4 matrix with the rotation in <code class="docutils literal notranslate"><span class="pre">R</span> <span class="pre">=</span> <span class="pre">M[:3,</span> <span class="pre">:3]</span></code> and the
translation in <code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">M[:3,</span> <span class="pre">3]</span></code>.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../lib/transformations.html#module-MDAnalysis.lib.transformations" title="MDAnalysis.lib.transformations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations</span></code></a></dt><dd><p>module of all coordinate transforms</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The rotation <span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is about the origin and is applied
before the translation <span class="math notranslate nohighlight">\(\mathbf{t}\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathsf{R}\mathbf{x} + \mathbf{t}\]</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.translate" title="Permalink to this definition"></a></dt>
<dd><p>Apply translation vector <cite>t</cite> to the selection’s coordinates.</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> coordinates are translated in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>t</strong> (<em>array_like</em>) – vector to translate coordinates with</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>self</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../lib/transformations.html#module-MDAnalysis.lib.transformations" title="MDAnalysis.lib.transformations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MDAnalysis.lib.transformations</span></code></a></dt><dd><p>module of all coordinate transforms</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The method applies a translation to the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>
from current coordinates <span class="math notranslate nohighlight">\(\mathbf{x}\)</span> to new coordinates
<span class="math notranslate nohighlight">\(\mathbf{x}'\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}' = \mathbf{x} + \mathbf{t}\]</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.union">
<span class="sig-name descname"><span class="pre">union</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.union" title="Permalink to this definition"></a></dt>
<dd><p>Group of elements either in this Group or another</p>
<p>On the contrary to concatenation, this method sort the elements and
removes duplicate ones. It is synomymous to the <cite>|</cite> operator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>Group</em><em> or </em><em>Component</em>) – Group or Component with <cite>other.level</cite> same as <cite>self.level</cite></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Group with the combined elements of <cite>self</cite> and <cite>other</cite>, without
duplicate elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Group</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>In contrast to <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.concatenate" title="MDAnalysis.core.groups.SegmentGroup.concatenate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">concatenate()</span></code></a>, any duplicates are dropped
and the result is sorted.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ag1</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name O&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag2</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name N&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span> <span class="o">=</span> <span class="n">ag1</span> <span class="o">|</span> <span class="n">ag2</span>  <span class="c1"># or ag1.union(ag2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ag3</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">names</span>
<span class="go">array([&#39;N&#39;, &#39;O&#39;, &#39;N&#39;], dtype=object)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.concatenate" title="MDAnalysis.core.groups.SegmentGroup.concatenate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concatenate</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.intersection" title="MDAnalysis.core.groups.SegmentGroup.intersection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">intersection</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.unique">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">unique</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.unique" title="Permalink to this definition"></a></dt>
<dd><p>Return a <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> containing sorted and unique
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segments</span></code></a> only.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sg</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">segments</span><span class="p">[[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg</span>
<span class="go">&lt;SegmentGroup with 6 segments&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([2, 1, 2, 2, 1, 0])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg2</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">unique</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg2</span>
<span class="go">&lt;SegmentGroup with 3 segments&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg2</span><span class="o">.</span><span class="n">ix</span>
<span class="go">array([0, 1, 2])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sg2</span><span class="o">.</span><span class="n">unique</span> <span class="ow">is</span> <span class="n">sg2</span>
<span class="go">False</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.0.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.19.0: </span>If the <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup" title="MDAnalysis.core.groups.SegmentGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentGroup</span></code></a> is already
unique, <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.unique" title="MDAnalysis.core.groups.SegmentGroup.unique"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SegmentGroup.unique</span></code></a> now returns the group itself
instead of a copy.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 2.0.0: </span>This function now always returns a copy.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.universe">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">universe</span></span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.universe" title="Permalink to this definition"></a></dt>
<dd><p>The underlying <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> the group
belongs to.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.unwrap">
<span class="sig-name descname"><span class="pre">unwrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fragments'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'com'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.unwrap" title="Permalink to this definition"></a></dt>
<dd><p>Move atoms of this group so that bonds within the
group’s compounds aren’t split across periodic boundaries.</p>
<p>This function is most useful when atoms have been packed into the
primary unit cell, causing breaks mid-molecule, with the molecule then
appearing on either side of the unit cell. This is problematic for
operations such as calculating the center of mass of the molecule.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>+-----------+       +-----------+
|           |       |           |
| 6       3 |       |         3 | 6
| !       ! |       |         ! | !
|-5-8   1-2-|  ==&gt;  |       1-2-|-5-8
| !       ! |       |         ! | !
| 7       4 |       |         4 | 7
|           |       |           |
+-----------+       +-----------+
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>compound</strong> (<em>{'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
Which type of compound to unwrap. Note that, in any case, all
atoms within each compound must be interconnected by bonds, i.e.,
compounds must correspond to (parts of) molecules.</p></li>
<li><p><strong>reference</strong> (<em>{'com'</em><em>, </em><em>'cog'</em><em>, </em><em>None}</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">'com'</span></code> (center of mass) or <code class="docutils literal notranslate"><span class="pre">'cog'</span></code> (center of geometry), the
unwrapped compounds will be shifted so that their individual
reference point lies within the primary unit cell.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no such shift is performed.</p></li>
<li><p><strong>inplace</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, coordinates are modified in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>coords</strong> – Unwrapped atom coordinate array of shape <code class="docutils literal notranslate"><span class="pre">(n,</span> <span class="pre">3)</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code> but the underlying topology does
    not contain molecule information, or if <cite>reference</cite> is <code class="docutils literal notranslate"><span class="pre">'com'</span></code>
    but the topology does not contain masses.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>reference</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'com'</span></code>, <code class="docutils literal notranslate"><span class="pre">'cog'</span></code>, or <code class="docutils literal notranslate"><span class="pre">None</span></code>, or
    if <cite>reference</cite> is <code class="docutils literal notranslate"><span class="pre">'com'</span></code> and the total mass of any <cite>compound</cite> is
    zero.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Be aware of the fact that only atoms <em>belonging to the group</em> will
be unwrapped! If you want entire molecules to be unwrapped, make sure
that all atoms of these molecules are part of the group.
An AtomGroup containing all atoms of all fragments in the group <code class="docutils literal notranslate"><span class="pre">ag</span></code>
can be created with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">all_frag_atoms</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">ag</span><span class="o">.</span><span class="n">fragments</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="../lib/mdamath.html#MDAnalysis.lib.mdamath.make_whole" title="MDAnalysis.lib.mdamath.make_whole"><code class="xref py py-func docutils literal notranslate"><span class="pre">make_whole()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.wrap" title="MDAnalysis.core.groups.SegmentGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">wrap()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.pack_into_box" title="MDAnalysis.core.groups.SegmentGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pack_into_box()</span></code></a>, <code class="xref py py-func docutils literal notranslate"><span class="pre">apply_PBC()</span></code></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.SegmentGroup.wrap">
<span class="sig-name descname"><span class="pre">wrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'atoms'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'com'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.SegmentGroup.wrap" title="Permalink to this definition"></a></dt>
<dd><p>Shift the contents of this group back into the primary unit cell
according to periodic boundary conditions.</p>
<p>Specifying a <cite>compound</cite> will keep the <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in each
compound together during the process. If <cite>compound</cite> is different from
<code class="docutils literal notranslate"><span class="pre">'atoms'</span></code>, each compound as a whole will be shifted so that its
<cite>center</cite> lies within the primary unit cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>compound</strong> (<em>{'atoms'</em><em>, </em><em>'group'</em><em>, </em><em>'segments'</em><em>, </em><em>'residues'</em><em>, </em><em>'molecules'</em><em>, </em>) – ‘fragments’}, optional
Which type of compound to keep together during wrapping. Note that,
in any case, <em>only</em> the positions of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
<em>belonging to the group</em> will be taken into account.</p></li>
<li><p><strong>center</strong> (<em>{'com'</em><em>, </em><em>'cog'}</em>) – How to define the center of a given group of atoms. If <cite>compound</cite> is
<code class="docutils literal notranslate"><span class="pre">'atoms'</span></code>, this parameter is meaningless and therefore ignored.</p></li>
<li><p><strong>box</strong> (<em>array_like</em><em>, </em><em>optional</em>) – The unitcell dimensions of the system, which can be orthogonal or
triclinic and must be provided in the same format as returned by
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.dimensions" title="MDAnalysis.coordinates.timestep.Timestep.dimensions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MDAnalysis.coordinates.timestep.Timestep.dimensions</span></code></a>:
<code class="docutils literal notranslate"><span class="pre">[lx,</span> <span class="pre">ly,</span> <span class="pre">lz,</span> <span class="pre">alpha,</span> <span class="pre">beta,</span> <span class="pre">gamma]</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, uses the
dimensions of the current time step.</p></li>
<li><p><strong>inplace</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, coordinates will be changed in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array of wrapped atom coordinates of dtype <cite>np.float32</cite> and shape
<code class="docutils literal notranslate"><span class="pre">(len(self.atoms.n_atoms),</span> <span class="pre">3)</span></code></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.11)"><strong>ValueError</strong></a> – If <cite>compound</cite> is not one of <code class="docutils literal notranslate"><span class="pre">'atoms'</span></code>, <code class="docutils literal notranslate"><span class="pre">'group'</span></code>,
    <code class="docutils literal notranslate"><span class="pre">'segments'</span></code>, <code class="docutils literal notranslate"><span class="pre">'residues'</span></code>, <code class="docutils literal notranslate"><span class="pre">'molecules'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code>.</p></li>
<li><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If <cite>compound</cite> is <code class="docutils literal notranslate"><span class="pre">'molecule'</span></code> but the topology doesn’t
    contain molecule information (molnums) or if <cite>compound</cite> is
    <code class="docutils literal notranslate"><span class="pre">'fragments'</span></code> but the topology doesn’t contain bonds or if
    <cite>center</cite> is <code class="docutils literal notranslate"><span class="pre">'com'</span></code> but the topology doesn’t contain masses.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>All atoms of the group will be moved so that the centers of its
compounds lie within the primary periodic image. For orthorhombic unit
cells, the primary periodic image is defined as the half-open interval
<span class="math notranslate nohighlight">\([0,L_i)\)</span> between <span class="math notranslate nohighlight">\(0\)</span> and boxlength <span class="math notranslate nohighlight">\(L_i\)</span> in all
dimensions <span class="math notranslate nohighlight">\(i\in\{x,y,z\}\)</span>, i.e., the origin of the of the
simulation box is taken to be at the origin <span class="math notranslate nohighlight">\((0,0,0)\)</span> of the
euclidian coordinate system. A compound center residing at position
<span class="math notranslate nohighlight">\(x_i\)</span> in dimension <span class="math notranslate nohighlight">\(i\)</span> will be shifted to <span class="math notranslate nohighlight">\(x_i'\)</span>
according to</p>
<div class="math notranslate nohighlight">
\[x_i' = x_i - \left\lfloor\frac{x_i}{L_i}\right\rfloor\,.\]</div>
<p>When specifying a <cite>compound</cite>, the translation is calculated based on
each compound. The same translation is applied to all atoms
within this compound, meaning it will not be broken by the shift.
This might however mean that not all atoms of a compound will be
inside the unit cell after wrapping, but rather will be the center of
the compound.
Be aware of the fact that only atoms <em>belonging to the group</em> will be
taken into account!</p>
<p><cite>center</cite> allows to define how the center of each group is computed.
This can be either <code class="docutils literal notranslate"><span class="pre">'com'</span></code> for center of mass, or <code class="docutils literal notranslate"><span class="pre">'cog'</span></code> for
center of geometry.</p>
<p><cite>box</cite> allows a unit cell to be given for the transformation. If not
specified, the <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.dimensions" title="MDAnalysis.coordinates.timestep.Timestep.dimensions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">dimensions</span></code></a>
information from the current
<a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a> will be used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.wrap" title="MDAnalysis.core.groups.AtomGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AtomGroup.wrap()</span></code></a> is currently faster than
<a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup.wrap" title="MDAnalysis.core.groups.ResidueGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ResidueGroup.wrap()</span></code></a> or <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.wrap" title="MDAnalysis.core.groups.SegmentGroup.wrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SegmentGroup.wrap()</span></code></a>.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.pack_into_box" title="MDAnalysis.core.groups.SegmentGroup.pack_into_box"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pack_into_box()</span></code></a>, <a class="reference internal" href="#MDAnalysis.core.groups.SegmentGroup.unwrap" title="MDAnalysis.core.groups.SegmentGroup.unwrap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">unwrap()</span></code></a>, <a class="reference internal" href="../lib/distances.html#MDAnalysis.lib.distances.apply_PBC" title="MDAnalysis.lib.distances.apply_PBC"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MDAnalysis.lib.distances.apply_PBC()</span></code></a></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.2.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.20.0: </span>The method only acts on atoms <em>belonging to the group</em> and returns
the wrapped positions as a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a>.
Added optional argument <cite>inplace</cite>.</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.UpdatingAtomGroup">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.groups.</span></span><span class="sig-name descname"><span class="pre">UpdatingAtomGroup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#UpdatingAtomGroup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="Permalink to this definition"></a></dt>
<dd><p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> subclass that dynamically updates its selected atoms.</p>
<p>Accessing any attribute/method of an <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a> instance
triggers a check for the last frame the group was updated. If the last
frame matches the current trajectory frame, the attribute is returned
normally; otherwise the group is updated (the stored selections are
re-applied), and only then is the attribute returned.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.16.0.</span></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>base_group</strong> (<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>) – group on which <em>selections</em> are to be applied.</p></li>
<li><p><strong>selections</strong> (a tuple of <code class="xref py py-class docutils literal notranslate"><span class="pre">Selection</span></code>) – instances selections ready to be applied to <em>base_group</em>.</p></li>
</ul>
</dd>
</dl>
<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.UpdatingAtomGroup.atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.UpdatingAtomGroup.atoms" title="Permalink to this definition"></a></dt>
<dd><p>Get a <em>static</em> <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> identical to the group of currently
selected <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> in the <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a>.</p>
<p>By returning a <em>static</em> <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> it becomes possible to
compare the contents of the group <em>between</em> trajectory frames. See the
Example below.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup.atoms" title="MDAnalysis.core.groups.UpdatingAtomGroup.atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">atoms</span></code></a> attribute of an <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a> behaves
differently from <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.atoms" title="MDAnalysis.core.groups.AtomGroup.atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">AtomGroup.atoms</span></code></a>: the latter returns the
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> itself whereas the former returns a
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> and not an <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a> (for this, use
<a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup.copy" title="MDAnalysis.core.groups.UpdatingAtomGroup.copy"><code class="xref py py-meth docutils literal notranslate"><span class="pre">UpdatingAtomGroup.copy()</span></code></a>).</p>
</div>
<p class="rubric">Example</p>
<p>The static <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup.atoms" title="MDAnalysis.core.groups.UpdatingAtomGroup.atoms"><code class="xref py py-attr docutils literal notranslate"><span class="pre">atoms</span></code></a> allows comparison of groups of atoms between
frames. For example, track water molecules that move in and out of a
solvation shell of a protein:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">u</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">TPR</span><span class="p">,</span> <span class="n">XTC</span><span class="p">)</span>
<span class="n">water_shell</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s2">&quot;name OW and around 3.5 protein&quot;</span><span class="p">,</span> <span class="n">updating</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">water_shell_prev</span> <span class="o">=</span> <span class="n">water_shell</span><span class="o">.</span><span class="n">atoms</span>

<span class="k">for</span> <span class="n">ts</span> <span class="ow">in</span> <span class="n">u</span><span class="o">.</span><span class="n">trajectory</span><span class="p">:</span>
    <span class="n">exchanged</span> <span class="o">=</span> <span class="n">water_shell</span> <span class="o">-</span> <span class="n">water_shell_prev</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">time</span><span class="p">,</span> <span class="s2">&quot;waters in shell =&quot;</span><span class="p">,</span> <span class="n">water_shell</span><span class="o">.</span><span class="n">n_residues</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">time</span><span class="p">,</span> <span class="s2">&quot;waters that exchanged = &quot;</span><span class="p">,</span> <span class="n">exchanged</span><span class="o">.</span><span class="n">n_residues</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">time</span><span class="p">,</span> <span class="s2">&quot;waters that remained bound = &quot;</span><span class="p">,</span>
          <span class="n">water_shell</span><span class="o">.</span><span class="n">n_residues</span> <span class="o">-</span> <span class="n">exchanged</span><span class="o">.</span><span class="n">n_residues</span><span class="p">)</span>

    <span class="n">water_shell_prev</span> <span class="o">=</span> <span class="n">water_shell</span><span class="o">.</span><span class="n">atoms</span>
</pre></div>
</div>
<p>By remembering the atoms of the current time step in
<cite>water_shell_prev</cite>, it becomes possible to use the <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup.subtract" title="MDAnalysis.core.groups.AtomGroup.subtract"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subtraction</span>
<span class="pre">of</span> <span class="pre">AtomGroups</span></code></a> to find the water molecules that
changed.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup.copy" title="MDAnalysis.core.groups.UpdatingAtomGroup.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a></dt><dd><p>return a true copy of the <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a></p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.UpdatingAtomGroup.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#UpdatingAtomGroup.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.UpdatingAtomGroup.copy" title="Permalink to this definition"></a></dt>
<dd><p>Get another <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup" title="MDAnalysis.core.groups.UpdatingAtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">UpdatingAtomGroup</span></code></a> identical to this one.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.19.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.UpdatingAtomGroup.is_uptodate">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_uptodate</span></span><a class="headerlink" href="#MDAnalysis.core.groups.UpdatingAtomGroup.is_uptodate" title="Permalink to this definition"></a></dt>
<dd><p>Checks whether the selection needs updating based on frame number only.</p>
<p>Modifications to the coordinate data that render selections stale are
not caught, and in those cases <a class="reference internal" href="#MDAnalysis.core.groups.UpdatingAtomGroup.is_uptodate" title="MDAnalysis.core.groups.UpdatingAtomGroup.is_uptodate"><code class="xref py py-attr docutils literal notranslate"><span class="pre">is_uptodate</span></code></a> may return an
erroneous value.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if the group’s selection is up-to-date, <code class="docutils literal notranslate"><span class="pre">False</span></code>
otherwise.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.UpdatingAtomGroup.update_selection">
<span class="sig-name descname"><span class="pre">update_selection</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#UpdatingAtomGroup.update_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.UpdatingAtomGroup.update_selection" title="Permalink to this definition"></a></dt>
<dd><p>Forces the reevaluation and application of the group’s selection(s).</p>
<p>This method is triggered automatically when accessing attributes, if
the last update occurred under a different trajectory frame.</p>
</dd></dl>

</dd></dl>

</section>
<section id="chemical-units">
<h3><span class="section-number">11.2.2.1.2. </span>Chemical units<a class="headerlink" href="#chemical-units" title="Permalink to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.groups.</span></span><span class="sig-name descname"><span class="pre">Atom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#Atom"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.Atom" title="Permalink to this definition"></a></dt>
<dd><p><a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> base class.</p>
<p>This class is used by a <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> for
generating its Topology-specific <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> class. All the
<a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.TopologyAttr" title="MDAnalysis.core.topologyattrs.TopologyAttr"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyAttr</span></code></a> components are obtained
from <code class="xref py py-class docutils literal notranslate"><span class="pre">ComponentBase</span></code>, so this class only includes ad-hoc methods
specific to <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>.</p>
<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.bfactor">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bfactor</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.bfactor" title="Permalink to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have tempfactors. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.bonded_atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bonded_atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.bonded_atoms" title="Permalink to this definition"></a></dt>
<dd><p>An <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> of all
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> bonded to this
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have bonds. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.force">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">force</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.force" title="Permalink to this definition"></a></dt>
<dd><p>Force on the atom.</p>
<p>The force can be changed by assigning an array of shape <code class="docutils literal notranslate"><span class="pre">(3,)</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>changing the force is not reflected in any files; reading any
frame from the trajectory will replace the change with that
from the file</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If the underlying <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a>
    does not contain
    <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.forces" title="MDAnalysis.coordinates.timestep.Timestep.forces"><code class="xref py py-attr docutils literal notranslate"><span class="pre">forces</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.fragindex">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fragindex</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.fragindex" title="Permalink to this definition"></a></dt>
<dd><p>The index (ID) of the
<a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.Bonds.fragment" title="MDAnalysis.core.topologyattrs.Bonds.fragment"><code class="xref py py-class docutils literal notranslate"><span class="pre">fragment</span></code></a> this
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> is part of.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.20.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have bonds. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.fragment">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fragment</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.fragment" title="Permalink to this definition"></a></dt>
<dd><p>An <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> representing the
fragment this <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a> is part of.</p>
<p>A fragment is a
<a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">group</span> <span class="pre">of</span> <span class="pre">atoms</span></code></a> which are
interconnected by <a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.Bonds" title="MDAnalysis.core.topologyattrs.Bonds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bonds</span></code></a>, i.e.,
there exists a path along one
or more <a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.Bonds" title="MDAnalysis.core.topologyattrs.Bonds"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bonds</span></code></a> between any pair
of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a>
within a fragment. Thus, a fragment typically corresponds to a molecule.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have bonds. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.get_connections">
<span class="sig-name descname"><span class="pre">get_connections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">typename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outside</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.get_connections" title="Permalink to this definition"></a></dt>
<dd><p>Get bonded connections between atoms as a
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.TopologyGroup" title="MDAnalysis.core.topologyobjects.TopologyGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyGroup</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>typename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a>) – group name. One of {“bonds”, “angles”, “dihedrals”,
“impropers”, “ureybradleys”, “cmaps”}</p></li>
<li><p><strong>outside</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Whether to include connections involving atoms outside
this group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>TopologyGroup</em> – containing the bonded group of choice, i.e. bonds, angles,
dihedrals, impropers, ureybradleys or cmaps.</p></li>
<li><p><em>.. versionadded:: 1.1.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.ix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.ix" title="Permalink to this definition"></a></dt>
<dd><p>Unique index of this component.</p>
<p>If this component is an <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>, this is the index of the
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>.
If it is a <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, this is the index of the <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>.
If it is a <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, this is the index of the <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.ix_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix_array</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.ix_array" title="Permalink to this definition"></a></dt>
<dd><p>Unique index of this component as an array.</p>
<p>This method gives a consistent API between components and groups.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Atom.ix" title="MDAnalysis.core.groups.Atom.ix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ix</span></code></a></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.position">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">position</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.position" title="Permalink to this definition"></a></dt>
<dd><p>Coordinates of the atom.</p>
<p>The position can be changed by assigning an array of length (3,).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>changing the position is not reflected in any files; reading
any frame from the trajectory will replace the change with
that from the file</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If the underlying <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a>
    does not contain
    <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.positions" title="MDAnalysis.coordinates.timestep.Timestep.positions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">positions</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Atom.velocity">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">velocity</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Atom.velocity" title="Permalink to this definition"></a></dt>
<dd><p>Velocity of the atom.</p>
<p>The velocity can be changed by assigning an array of shape <code class="docutils literal notranslate"><span class="pre">(3,)</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>changing the velocity is not reflected in any files; reading
any frame from the trajectory will replace the change with
that from the file</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><strong>NoDataError</strong></a> – If the underlying <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep" title="MDAnalysis.coordinates.timestep.Timestep"><code class="xref py py-class docutils literal notranslate"><span class="pre">Timestep</span></code></a>
    does not contain
    <a class="reference internal" href="../coordinates/timestep.html#MDAnalysis.coordinates.timestep.Timestep.velocities" title="MDAnalysis.coordinates.timestep.Timestep.velocities"><code class="xref py py-attr docutils literal notranslate"><span class="pre">velocities</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.groups.</span></span><span class="sig-name descname"><span class="pre">Residue</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#Residue"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.Residue" title="Permalink to this definition"></a></dt>
<dd><p><a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> base class.</p>
<p>This class is used by a <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> for
generating its Topology-specific <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> class. All the
<a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.TopologyAttr" title="MDAnalysis.core.topologyattrs.TopologyAttr"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyAttr</span></code></a> components are obtained
from <code class="xref py py-class docutils literal notranslate"><span class="pre">ComponentBase</span></code>, so this class only includes ad-hoc methods
specific to <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residues</span></code></a>.</p>
<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.atoms" title="Permalink to this definition"></a></dt>
<dd><p>An <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> present in this
<a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.bfactors">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bfactors</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.bfactors" title="Permalink to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have tempfactors. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.chi1_selection">
<span class="sig-name descname"><span class="pre">chi1_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cg_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CG</span> <span class="pre">CG1</span> <span class="pre">OG</span> <span class="pre">OG1</span> <span class="pre">SG'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.chi1_selection" title="Permalink to this definition"></a></dt>
<dd><p>Select AtomGroup corresponding to the chi1 sidechain dihedral <code class="docutils literal notranslate"><span class="pre">N-CA-CB-*G.</span></code>
The gamma atom is taken to be the heavy atom in the gamma position. If more than one
heavy atom is present (e.g. CG1 and CG2), the one with the lower number is used (CG1).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This numbering of chi1 atoms here in following with the IUPAC 1970 rules.
However, it should be noted that analyses which use dihedral angles may have
different definitions. For example, the
<a class="reference internal" href="../analysis/dihedrals.html#MDAnalysis.analysis.dihedrals.Janin" title="MDAnalysis.analysis.dihedrals.Janin"><code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.analysis.dihedrals.Janin</span></code></a> class does not incorporate
amino acids where the gamma atom is not carbon, into its chi1 selections.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
<li><p><strong>cb_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the beta-carbon atom</p></li>
<li><p><strong>cg_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the gamma-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>4-atom selection in the correct order. If no CB and/or CG is found
then this method returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup">AtomGroup</a></p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.5.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Added arguments for flexible atom names and refactored code for
faster atom matching with boolean arrays.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have names. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.get_connections">
<span class="sig-name descname"><span class="pre">get_connections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">typename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outside</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.get_connections" title="Permalink to this definition"></a></dt>
<dd><p>Get bonded connections between atoms as a
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.TopologyGroup" title="MDAnalysis.core.topologyobjects.TopologyGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyGroup</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>typename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a>) – group name. One of {“bonds”, “angles”, “dihedrals”,
“impropers”, “ureybradleys”, “cmaps”}</p></li>
<li><p><strong>outside</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Whether to include connections involving atoms outside
this group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>TopologyGroup</em> – containing the bonded group of choice, i.e. bonds, angles,
dihedrals, impropers, ureybradleys or cmaps.</p></li>
<li><p><em>.. versionadded:: 1.1.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.ix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.ix" title="Permalink to this definition"></a></dt>
<dd><p>Unique index of this component.</p>
<p>If this component is an <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>, this is the index of the
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>.
If it is a <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, this is the index of the <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>.
If it is a <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, this is the index of the <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.ix_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix_array</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.ix_array" title="Permalink to this definition"></a></dt>
<dd><p>Unique index of this component as an array.</p>
<p>This method gives a consistent API between components and groups.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Residue.ix" title="MDAnalysis.core.groups.Residue.ix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ix</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.omega_selection">
<span class="sig-name descname"><span class="pre">omega_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.omega_selection" title="Permalink to this definition"></a></dt>
<dd><p>Select AtomGroup corresponding to the omega protein backbone dihedral
CA-C-N’-CA’.</p>
<p>omega describes the -C-N- peptide bond. Typically, it is trans (180
degrees) although cis-bonds (0 degrees) are also occasionally observed
(especially near Proline).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>AtomGroup</em> – 4-atom selection in the correct order. If no C’ found in the
previous residue (by resid) then this method returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionchanged:: 1.0.0</em> – Added arguments for flexible atom names and refactored code for
faster atom matching with boolean arrays.</p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have names. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.phi_selection">
<span class="sig-name descname"><span class="pre">phi_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.phi_selection" title="Permalink to this definition"></a></dt>
<dd><p>Select AtomGroup corresponding to the phi protein backbone dihedral
C’-N-CA-C.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>AtomGroup</em> – 4-atom selection in the correct order. If no C’ found in the
previous residue (by resid) then this method returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionchanged:: 1.0.0</em> – Added arguments for flexible atom names and refactored code for
faster atom matching with boolean arrays.</p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have names. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.psi_selection">
<span class="sig-name descname"><span class="pre">psi_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.psi_selection" title="Permalink to this definition"></a></dt>
<dd><p>Select AtomGroup corresponding to the psi protein backbone dihedral
N-CA-C-N’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>c_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone C atom</p></li>
<li><p><strong>n_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the backbone N atom</p></li>
<li><p><strong>ca_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – name for the alpha-carbon atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>AtomGroup</em> – 4-atom selection in the correct order. If no N’ found in the
following residue (by resid) then this method returns <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><em>.. versionchanged:: 1.0.0</em> – Added arguments for flexible atom names and refactored code for
faster atom matching with boolean arrays.</p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have names. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Residue.segment">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">segment</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Residue.segment" title="Permalink to this definition"></a></dt>
<dd><p>The <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> this <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> belongs to.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Segment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">MDAnalysis.core.groups.</span></span><span class="sig-name descname"><span class="pre">Segment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/core/groups.html#Segment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#MDAnalysis.core.groups.Segment" title="Permalink to this definition"></a></dt>
<dd><p><a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> base class.</p>
<p>This class is used by a <a class="reference internal" href="universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> for
generating its Topology-specific <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a> class. All the
<a class="reference internal" href="topologyattrs.html#MDAnalysis.core.topologyattrs.TopologyAttr" title="MDAnalysis.core.topologyattrs.TopologyAttr"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyAttr</span></code></a> components are obtained
from <code class="xref py py-class docutils literal notranslate"><span class="pre">ComponentBase</span></code>, so this class only includes ad-hoc methods
specific to <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segments</span></code></a>.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.16.2: </span><em>Instant selectors</em> of <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segments</span></code></a> will be removed in the
1.0 release.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.0.0: </span>Removed instant selectors, use either segment.residues[…] to select
residue by number, or segment.residues[segment.residue.resnames = …]
to select by resname.</p>
</div>
<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Segment.atoms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">atoms</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Segment.atoms" title="Permalink to this definition"></a></dt>
<dd><p>An <a class="reference internal" href="#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> of <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atoms</span></code></a> present in this
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Segment.bfactors">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bfactors</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Segment.bfactors" title="Permalink to this definition"></a></dt>
<dd><p>Bfactor alias with warning</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires the underlying topology to have tempfactors. Otherwise, a
<a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.NoDataError" title="MDAnalysis.exceptions.NoDataError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">NoDataError</span></code></a> is raised.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Segment.get_connections">
<span class="sig-name descname"><span class="pre">get_connections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">typename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outside</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#MDAnalysis.core.groups.Segment.get_connections" title="Permalink to this definition"></a></dt>
<dd><p>Get bonded connections between atoms as a
<a class="reference internal" href="topologyobjects.html#MDAnalysis.core.topologyobjects.TopologyGroup" title="MDAnalysis.core.topologyobjects.TopologyGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyGroup</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>typename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.11)"><em>str</em></a>) – group name. One of {“bonds”, “angles”, “dihedrals”,
“impropers”, “ureybradleys”, “cmaps”}</p></li>
<li><p><strong>outside</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.11)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Whether to include connections involving atoms outside
this group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>TopologyGroup</em> – containing the bonded group of choice, i.e. bonds, angles,
dihedrals, impropers, ureybradleys or cmaps.</p></li>
<li><p><em>.. versionadded:: 1.1.0</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Segment.ix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Segment.ix" title="Permalink to this definition"></a></dt>
<dd><p>Unique index of this component.</p>
<p>If this component is an <a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>, this is the index of the
<a class="reference internal" href="#MDAnalysis.core.groups.Atom" title="MDAnalysis.core.groups.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>.
If it is a <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>, this is the index of the <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a>.
If it is a <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>, this is the index of the <a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Segment.ix_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ix_array</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Segment.ix_array" title="Permalink to this definition"></a></dt>
<dd><p>Unique index of this component as an array.</p>
<p>This method gives a consistent API between components and groups.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.core.groups.Segment.ix" title="MDAnalysis.core.groups.Segment.ix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ix</span></code></a></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="MDAnalysis.core.groups.Segment.residues">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">residues</span></span><a class="headerlink" href="#MDAnalysis.core.groups.Segment.residues" title="Permalink to this definition"></a></dt>
<dd><p>A <a class="reference internal" href="#MDAnalysis.core.groups.ResidueGroup" title="MDAnalysis.core.groups.ResidueGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueGroup</span></code></a> of <a class="reference internal" href="#MDAnalysis.core.groups.Residue" title="MDAnalysis.core.groups.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residues</span></code></a> present in this
<a class="reference internal" href="#MDAnalysis.core.groups.Segment" title="MDAnalysis.core.groups.Segment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Segment</span></code></a>.</p>
</dd></dl>

</dd></dl>

</section>
<section id="levels">
<h3><span class="section-number">11.2.2.1.3. </span>Levels<a class="headerlink" href="#levels" title="Permalink to this heading"></a></h3>
<p>Each of the above classes has a <em>level</em> attribute.  This can be used to verify
that two objects are of the same level, or to access a particular class:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">u</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">()</span>

<span class="n">ag</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
<span class="n">at</span> <span class="o">=</span> <span class="n">u</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>

<span class="n">ag</span><span class="o">.</span><span class="n">level</span> <span class="o">==</span> <span class="n">at</span><span class="o">.</span><span class="n">level</span>  <span class="c1"># Returns True</span>

<span class="n">ag</span><span class="o">.</span><span class="n">level</span><span class="o">.</span><span class="n">singular</span>  <span class="c1"># Returns Atom class</span>
<span class="n">at</span><span class="o">.</span><span class="n">level</span><span class="o">.</span><span class="n">plural</span>  <span class="c1"># Returns AtomGroup class</span>
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="universe.html" class="btn btn-neutral float-left" title="11.2.1. Core object: Universe — MDAnalysis.core.universe" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="topology.html" class="btn btn-neutral float-right" title="11.3.1. Core Topology object — MDAnalysis.core.topology" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2005-2023, Naveen Michaud-Agrawal, Elizabeth J. Denning, Christian Beckstein (logo), Joshua L. Adelman, Henok Ademtew, Shobhit Agarwal, Aya M. Alaa, Irfan Alibay, Kazi Shudipto Amin, Anshul Angaria, Luís Pedro Borges Araújo, Balasubramanian, Utkarsh Bansal, Patricio Barletta, Leonardo Barneschi, Jonathan Barnoud, Estefania Barreto-Ojeda, Tone Bengtsen, Alejandro Bernardin, Ninad Bhat, Mateusz Bieniek, Kavya Bisht, Wouter Boomsma, Jose Borreguero, Cédric Bouysset, Kevin Boyd, Meet Brijwani, Bart Bruininks, Sébastien Buchoux, Sören von Bülow, Yantong Cai, David Caplan, Yuanyu Chang, Pratham Chauhan, Matthieu Chavent, Haochuan Chen, Xu Hong Chen, Kathleen Clark, Jennifer A Clark, Orion Cohen, Charlie Cook, Ruggero Cortini, Nicholas Craven, Ramon Crehuet, Davide Cruz, Robert Delgado, John Detlefs, Xavier Deupi, Jan Domanski, David L. Dotson, Mark D. Driver, Ali Ehlen, Daniel J. Evans, Shujie Fan, Bjarne Feddersen, Lennard van der Feltz, Jake Fennick, Philip Fowler, Guillaume Fraux, Anirvinya G, Ahmed Salah Ghoneim, Mikhail Glagolev, William Glass, Joseph Goose, Alexander Gorfer, Richard J. Gowers, Lukas Grossar, Abhinav Gupta, Akshay Gupta, Pratik Gupta, Benjamin Hall, Ameya Harmalkar, Ivan Hristov, Eugen Hruska, Kyle J. Huston, Siddharth Jain, Edis Jakupovic, Joe Jordan, Henrik Jäger, Uma D Kadam, Aditya Kamath, Jon Kapla, Haleema Khan, Navya Khare, Utsav Khatu, Andrew William King, Henry Kobin, Abhishek A. Kognole, Kosuke Kudo, Atharva Kulkarni, Manish Kumar, Alia Lescoulie, Max Linke, Philip Loche, Jinju Lu, Hugo MacDermott-Opeskin, Shaivi Malik, Egor Marin, Domenico Marson, Micaela Matta, Andrew R. McCluskey, Robert McGibbon, Rocco Meli, Manuel Nuno Melo, Marcelo C. R. Melo, Dominik &#39;Rathann&#39; Mierzejewski, Henry Mull, Morgan L. Nance, Fiona B. Naughton, Alex Nesterenko, Hai Nguyen, Sang Young Noh, Meghan Osato, Daniele Padula, Nabarun Pal, Mattia F. Palermo, Dimitrios Papageorgiou, Rafael R. Pappalardo, Vishal Parmar, Danny Parton, Shakul Pathak, Christian Pfaendner, Joshua L. Phillips, Marcelo D. Poleto, Hannah Pollak, Kashish Punjani, Michael Quevillon, Vedant Rathore, Tyler Reddy, Pedro Reis, Paul Rigor, Andrea Rizzi, Xiaoxu Ruan, Carlos Yanez S., Utkarsh Saxena, Moritz Schaeffler, Alexander Schlaich, Marcello Sega, Ricky Sexton, Sean L. Seyler, Faraaz Shah, Sulay Shah, Abhishek Shandilya, Shubham Sharma, Rishabh Shukla, Karthikeyan Singaravelan, Tamandeep Singh, Paul Smith, Andy Somogyi, Caio S. Souza, Shantanu Srivastava, Lukas Stelzl, Jan Stevens, Gorman Stock, Fenil Suchak, Ayush Suhane, Filip T. Szczypiński, Sukeerti T, Matthijs Tadema, Joao Miguel Correia Teixeira, Paarth Thadani, Matthew W. Thompson, Hao Tian, Matteo Tiberti, Wiep van der Toorn, Mieczyslaw Torchala, Aditi Tripathi, Mark Verma, Josh Vermaas, Isaac Virshup, Lily Wang, Nestor Wendt, Zhiyi Wu, Tengyu Xie, Zhuyi Xue, Mingyi Xue, Alexander Yang, Juan Eiros Zamora, Johannes Zeman, Yibo Zhang, Raymond Zhao, Yuxuan Zhuang, and Oliver Beckstein.</p>
  </div>

  
 
<div class="footer"><p>Please see
    our <a href="https://www.mdanalysis.org/pages/privacy/">Privacy Policy</a>
    to learn how <a href="https://www.mdanalysis.org">MDAnalysis</a> collects data.</p>
    <script data-goatcounter="https://mdanalysis.goatcounter.com/count"
        async src="//gc.zgo.at/count.js"></script>
</div>


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
    var versions_json_url = 'https://docs.mdanalysis.org/versions.json'
</script>

<div class="rst-versions" data-toggle="rst-versions" role="note"
     aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      <span class="fa fa-book"></span>
        2.5.0
      <span class="fa fa-caret-down"></span>
    </span>

    <div class="rst-other-versions">
        <dl id="versionselector">
            <dt>Other Versions</dt>
        </dl>

    </div>
</div><script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>