<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>3.1.1. Coordinate fitting and alignment — MDAnalysis.analysis.align &mdash; MDAnalysis 0.11.0 documentation</title>
    
    <link rel="stylesheet" href="../../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.11.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="search" type="application/opensearchdescription+xml"
          title="Search within MDAnalysis 0.11.0 documentation"
          href="../../_static/opensearch.xml"/>
    <link rel="shortcut icon" href="../../_static/mdanalysis-logo.ico"/>
    <link rel="top" title="MDAnalysis 0.11.0 documentation" href="../../index.html" />
    <link rel="up" title="3. Analysis modules" href="../analysis_modules.html" />
    <link rel="next" title="3.1.2. Native contacts analysis — MDAnalysis.analysis.contacts" href="contacts.html" />
    <link rel="prev" title="3. Analysis modules" href="../analysis_modules.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="contacts.html" title="3.1.2. Native contacts analysis — MDAnalysis.analysis.contacts"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../analysis_modules.html" title="3. Analysis modules"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">MDAnalysis 0.11.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../analysis_modules.html" accesskey="U">3. Analysis modules</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../index.html">
              <img class="logo" src="../../_static/mdanalysis-logo-200x150.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">3.1.1. Coordinate fitting and alignment &#8212; <code class="docutils literal"><span class="pre">MDAnalysis.analysis.align</span></code></a><ul>
<li><a class="reference internal" href="#rms-fitting-tutorial">3.1.1.1. RMS-fitting tutorial</a></li>
<li><a class="reference internal" href="#common-usage">3.1.1.2. Common usage</a></li>
<li><a class="reference internal" href="#functions">3.1.1.3. Functions</a></li>
<li><a class="reference internal" href="#helper-functions">3.1.1.4. Helper functions</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../analysis_modules.html"
                        title="previous chapter">3. Analysis modules</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="contacts.html"
                        title="next chapter">3.1.2. Native contacts analysis &#8212; <code class="docutils literal"><span class="pre">MDAnalysis.analysis.contacts</span></code></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../../_sources/documentation_pages/analysis/align.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-MDAnalysis.analysis.align"></span><div class="section" id="coordinate-fitting-and-alignment-mdanalysis-analysis-align">
<h1>3.1.1. Coordinate fitting and alignment &#8212; <a class="reference internal" href="#module-MDAnalysis.analysis.align" title="MDAnalysis.analysis.align"><code class="xref py py-mod docutils literal"><span class="pre">MDAnalysis.analysis.align</span></code></a><a class="headerlink" href="#coordinate-fitting-and-alignment-mdanalysis-analysis-align" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Oliver Beckstein, Joshua Adelman</td>
</tr>
<tr class="field-even field"><th class="field-name">Year:</th><td class="field-body">2010&#8211;2013</td>
</tr>
<tr class="field-odd field"><th class="field-name">Copyright:</th><td class="field-body">GNU Public License v3</td>
</tr>
</tbody>
</table>
<p>The module contains functions to fit a target structure to a reference
structure. They use the fast QCP algorithm to calculate the root mean
square distance (RMSD) between two coordinate sets <a class="reference internal" href="../lib/qcprot.html#theobald2005" id="id1">[Theobald2005]</a> and
the rotation matrix <em>R</em> that minimizes the RMSD <a class="reference internal" href="../lib/qcprot.html#liu2010" id="id2">[Liu2010]</a>. (Please
cite these references when using this module.).</p>
<p>Typically, one selects a group of atoms (such as the C-alphas),
calculates the RMSD and transformation matrix, and applys the
transformation to the current frame of a trajectory to obtain the
rotated structure. The <a class="reference internal" href="#MDAnalysis.analysis.align.alignto" title="MDAnalysis.analysis.align.alignto"><code class="xref py py-func docutils literal"><span class="pre">alignto()</span></code></a> and <a class="reference internal" href="#MDAnalysis.analysis.align.rms_fit_trj" title="MDAnalysis.analysis.align.rms_fit_trj"><code class="xref py py-func docutils literal"><span class="pre">rms_fit_trj()</span></code></a>
functions can be used to do this for individual frames and
trajectories respectively.</p>
<p>The <a class="reference internal" href="#rms-fitting-tutorial"><span>RMS-fitting tutorial</span></a> shows how to do the individual steps
manually and explains the intermediate steps.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<dl class="last docutils">
<dt><a class="reference internal" href="rms.html#module-MDAnalysis.analysis.rms" title="MDAnalysis.analysis.rms"><code class="xref py py-mod docutils literal"><span class="pre">MDAnalysis.analysis.rms</span></code></a></dt>
<dd>contains functions to compute RMSD (when structural alignment is not
required)</dd>
<dt><a class="reference internal" href="../lib/qcprot.html#module-MDAnalysis.lib.qcprot" title="MDAnalysis.lib.qcprot"><code class="xref py py-mod docutils literal"><span class="pre">MDAnalysis.lib.qcprot</span></code></a></dt>
<dd>implements the fast RMSD algorithm.</dd>
</dl>
</div>
<div class="section" id="rms-fitting-tutorial">
<span id="id3"></span><h2>3.1.1.1. RMS-fitting tutorial<a class="headerlink" href="#rms-fitting-tutorial" title="Permalink to this headline">¶</a></h2>
<p>The example uses files provided as part of the MDAnalysis test suite
(in the variables <code class="xref py py-data docutils literal"><span class="pre">PSF</span></code>,
<code class="xref py py-data docutils literal"><span class="pre">DCD</span></code>, and
<code class="xref py py-data docutils literal"><span class="pre">PDB_small</span></code>). For all further
examples execute first</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis</span> <span class="kn">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.analysis.align</span> <span class="kn">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.analysis.rms</span> <span class="kn">import</span> <span class="n">rmsd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">,</span> <span class="n">PDB_small</span>
</pre></div>
</div>
<p>In the simplest case, we can simply calculate the C-alpha RMSD between
two structures, using <code class="xref py py-func docutils literal"><span class="pre">rmsd()</span></code>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ref</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PDB_small</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rmsd</span><span class="p">(</span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">CA</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span> <span class="n">ref</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">CA</span><span class="o">.</span><span class="n">positions</span><span class="p">)</span>
<span class="go">18.858259026820352</span>
</pre></div>
</div>
<p>Note that in this example translations have not been removed. In order
to look at the pure rotation one needs to superimpose the centres of
mass (or geometry) first:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ref0</span> <span class="o">=</span>  <span class="n">ref</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">CA</span><span class="o">.</span><span class="n">positions</span> <span class="o">-</span> <span class="n">ref</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">CA</span><span class="o">.</span><span class="n">center_of_mass</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile0</span> <span class="o">=</span> <span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">CA</span><span class="o">.</span><span class="n">positions</span> <span class="o">-</span> <span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">CA</span><span class="o">.</span><span class="n">center_of_mass</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rmsd</span><span class="p">(</span><span class="n">mobile0</span><span class="p">,</span> <span class="n">ref0</span><span class="p">)</span>
<span class="go"> 6.8093965864717951</span>
</pre></div>
</div>
<p>The rotation matrix that superimposes <em>mobile</em> on <em>ref</em> while
minimizing the CA-RMSD is obtained with the <a class="reference internal" href="#MDAnalysis.analysis.align.rotation_matrix" title="MDAnalysis.analysis.align.rotation_matrix"><code class="xref py py-func docutils literal"><span class="pre">rotation_matrix()</span></code></a>
function</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">R</span><span class="p">,</span> <span class="n">rmsd</span> <span class="o">=</span> <span class="n">rotation_matrix</span><span class="p">(</span><span class="n">mobile0</span><span class="p">,</span> <span class="n">ref0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">rmsd</span>
<span class="go">6.8093965864717951</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">R</span>
<span class="go">[[ 0.14514539 -0.27259113  0.95111876]</span>
<span class="go"> [ 0.88652593  0.46267112 -0.00268642]</span>
<span class="go"> [-0.43932289  0.84358136  0.30881368]]</span>
</pre></div>
</div>
<p>Putting all this together one can superimpose all of <em>mobile</em> onto <em>ref</em>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="o">-</span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">CA</span><span class="o">.</span><span class="n">center_of_mass</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">R</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">ref</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">CA</span><span class="o">.</span><span class="n">center_of_mass</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;mobile_on_ref.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="common-usage">
<h2>3.1.1.2. Common usage<a class="headerlink" href="#common-usage" title="Permalink to this headline">¶</a></h2>
<p>To <strong>fit a single structure</strong> with <a class="reference internal" href="#MDAnalysis.analysis.align.alignto" title="MDAnalysis.analysis.align.alignto"><code class="xref py py-func docutils literal"><span class="pre">alignto()</span></code></a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ref</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">PDB_small</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">)</span>     <span class="c"># we use the first frame</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignto</span><span class="p">(</span><span class="n">mobile</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">select</span><span class="o">=</span><span class="s">&quot;protein and name CA&quot;</span><span class="p">,</span> <span class="n">mass_weighted</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>This will change <em>all</em> coordinates in <em>mobile</em> so that the protein
C-alpha atoms are optimally superimposed (translation and rotation).</p>
<p>To <strong>fit a whole trajectory</strong> to a reference structure with the
<a class="reference internal" href="#MDAnalysis.analysis.align.rms_fit_trj" title="MDAnalysis.analysis.align.rms_fit_trj"><code class="xref py py-func docutils literal"><span class="pre">rms_fit_trj()</span></code></a> function:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ref</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">PDB_small</span><span class="p">)</span>   <span class="c"># reference structure 1AKE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">trj</span> <span class="o">=</span> <span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">)</span>         <span class="c"># trajectory of change 1AKE-&gt;4AKE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rms_fit_trj</span><span class="p">(</span><span class="n">trj</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s">&#39;rmsfit.dcd&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>It is also possible to align two arbitrary structures by providing a
mapping between atoms based on a sequence alignment. This allows
fitting of structural homologs or wild type and mutant.</p>
<p>If a alignment was provided as &#8220;sequences.aln&#8221; one would first produce
the appropriate MDAnalysis selections with the <a class="reference internal" href="#MDAnalysis.analysis.align.fasta2select" title="MDAnalysis.analysis.align.fasta2select"><code class="xref py py-func docutils literal"><span class="pre">fasta2select()</span></code></a>
function and then feed the resulting dictionary to <a class="reference internal" href="#MDAnalysis.analysis.align.rms_fit_trj" title="MDAnalysis.analysis.align.rms_fit_trj"><code class="xref py py-func docutils literal"><span class="pre">rms_fit_trj()</span></code></a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">seldict</span> <span class="o">=</span> <span class="n">fasta2select</span><span class="p">(</span><span class="s">&#39;sequences.aln&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rms_fit_trj</span><span class="p">(</span><span class="n">trj</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s">&#39;rmsfit.dcd&#39;</span><span class="p">,</span> <span class="n">select</span><span class="o">=</span><span class="n">seldict</span><span class="p">)</span>
</pre></div>
</div>
<p>(See the documentation of the functions for this advanced usage.)</p>
</div>
<div class="section" id="functions">
<h2>3.1.1.3. Functions<a class="headerlink" href="#functions" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="MDAnalysis.analysis.align.alignto">
<code class="descclassname">MDAnalysis.analysis.align.</code><code class="descname">alignto</code><span class="sig-paren">(</span><em>mobile</em>, <em>reference</em>, <em>select='all'</em>, <em>mass_weighted=False</em>, <em>subselection=None</em>, <em>tol_mass=0.1</em>, <em>strict=False</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#alignto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.alignto" title="Permalink to this definition">¶</a></dt>
<dd><p>Spatially align <em>mobile</em> to <em>reference</em> by doing a RMSD fit on <em>select</em> atoms.</p>
<p>The superposition is done in the following way:</p>
<ol class="arabic simple">
<li>A rotation matrix is computed that minimizes the RMSD between
the coordinates of <cite>mobile.select_atoms(sel1)</cite> and
<cite>reference.select_atoms(sel2)</cite>; before the rotation, <em>mobile</em> is
translated so that its center of geometry (or center of mass)
coincides with the one of <em>reference</em>. (See below for explanation of
how <em>sel1</em> and <em>sel2</em> are derived from <em>select</em>.)</li>
<li>All atoms in <a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a> that
contains <em>mobile</em> are shifted and rotated. (See below for how
to change this behavior through the <em>subselection</em> keyword.)</li>
</ol>
<p>The <em>mobile</em> and <em>reference</em> atom groups can be constructed so that they
already match atom by atom. In this case, <em>select</em> should be set to &#8220;all&#8221;
(or <code class="docutils literal"><span class="pre">None</span></code>) so that no further selections are applied to <em>mobile</em> and
<em>reference</em>, therefore preserving the exact atom ordering (see
<a class="reference internal" href="../selections.html#ordered-selections-label"><span>Ordered selections</span></a>).</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The atom order for <em>mobile</em> and <em>reference</em> is <em>only</em>
preserved when <em>select</em> is either &#8220;all&#8221; or <code class="docutils literal"><span class="pre">None</span></code>. In any other case,
a new selection will be made that will sort the resulting AtomGroup by
index and therefore destroy the correspondence between the two groups. <strong>It
is safest not to mix ordered AtomGroups with selection strings.</strong></p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first docutils">
<dt><em>mobile</em></dt>
<dd><p class="first last">structure to be aligned, a <a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
or a whole <a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a></p>
</dd>
<dt><em>reference</em></dt>
<dd><p class="first last">reference structure, a <a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
or a whole <a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.Universe" title="MDAnalysis.core.AtomGroup.Universe"><code class="xref py py-class docutils literal"><span class="pre">Universe</span></code></a></p>
</dd>
<dt><em>select</em></dt>
<dd><ol class="first arabic simple">
<li>any valid selection string for
<a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup.select_atoms" title="MDAnalysis.core.AtomGroup.AtomGroup.select_atoms"><code class="xref py py-meth docutils literal"><span class="pre">select_atoms()</span></code></a> that produces identical
selections in <em>mobile</em> and <em>reference</em>; or</li>
<li>dictionary <code class="docutils literal"><span class="pre">{'mobile':sel1,</span> <span class="pre">'reference':sel2}</span></code>.
(the <a class="reference internal" href="#MDAnalysis.analysis.align.fasta2select" title="MDAnalysis.analysis.align.fasta2select"><code class="xref py py-func docutils literal"><span class="pre">fasta2select()</span></code></a> function returns such a
dictionary based on a <a class="reference external" href="http://www.clustal.org/">ClustalW</a> or <a class="reference external" href="http://www.compbio.dundee.ac.uk/manuals/stamp.4.2/">STAMP</a> sequence alignment); or</li>
<li>tuple <code class="docutils literal"><span class="pre">(sel1,</span> <span class="pre">sel2)</span></code></li>
</ol>
<p class="last">When using 2. or 3. with <em>sel1</em> and <em>sel2</em> then these selections can also each be
a list of selection strings (to generate a AtomGroup with defined atom order as
described under <a class="reference internal" href="../selections.html#ordered-selections-label"><span>Ordered selections</span></a>).</p>
</dd>
<dt><em>mass_weighted</em> <span class="classifier-delimiter">:</span> <span class="classifier">boolean</span></dt>
<dd><p class="first last"><code class="docutils literal"><span class="pre">True</span></code> uses the masses <code class="xref py py-meth docutils literal"><span class="pre">reference.masses()</span></code> as weights for the
RMSD fit.</p>
</dd>
<dt><em>tol_mass</em></dt>
<dd><p class="first last">Reject match if the atomic masses for matched atoms differ by more than
<em>tol_mass</em> [0.1]</p>
</dd>
<dt><em>strict</em></dt>
<dd><dl class="first last docutils">
<dt><code class="docutils literal"><span class="pre">True</span></code></dt>
<dd><p class="first last">Will raise <code class="xref py py-exc docutils literal"><span class="pre">SelectioError</span></code> if a single atom does not
match between the two selections.</p>
</dd>
<dt><code class="docutils literal"><span class="pre">False</span></code> [default]</dt>
<dd><p class="first last">Will try to prepare a matching selection by dropping
residues with non-matching atoms. See <a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal"><span class="pre">get_matching_atoms()</span></code></a>
for details.</p>
</dd>
</dl>
</dd>
<dt><em>subselection</em></dt>
<dd><p class="first">Apply the transformation only to this selection.</p>
<dl class="last docutils">
<dt><code class="docutils literal"><span class="pre">None</span></code> [default]</dt>
<dd><p class="first last">Apply to <cite>mobile.universe.atoms</cite> (i.e. all atoms in the
context of the selection from <em>mobile</em> such as the rest of a
protein, ligands and the surrounding water)</p>
</dd>
<dt><em>selection-string</em></dt>
<dd><p class="first last">Apply to <cite>mobile.select_atoms(selection-string)</cite></p>
</dd>
<dt><a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a></dt>
<dd><p class="first last">Apply to the arbitrary group of atoms</p>
</dd>
</dl>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">RMSD before and after spatial alignment.</p>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For RMSD-fitting trajectories it is more efficient to
use <a class="reference internal" href="#MDAnalysis.analysis.align.rms_fit_trj" title="MDAnalysis.analysis.align.rms_fit_trj"><code class="xref py py-func docutils literal"><span class="pre">rms_fit_trj()</span></code></a>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added check that the two groups describe the same atoms including
the new <em>tol_mass</em> keyword.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.10.0: </span>Uses <a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal"><span class="pre">get_matching_atoms()</span></code></a> to work with incomplete selections
and new <em>strict</em> keyword. The new default is to be lenient whereas
the old behavior was the equivalent of <em>strict</em> = <code class="docutils literal"><span class="pre">True</span></code>.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="MDAnalysis.analysis.align.rms_fit_trj">
<code class="descclassname">MDAnalysis.analysis.align.</code><code class="descname">rms_fit_trj</code><span class="sig-paren">(</span><em>traj</em>, <em>reference</em>, <em>select='all'</em>, <em>filename=None</em>, <em>rmsdfile=None</em>, <em>prefix='rmsfit_'</em>, <em>mass_weighted=False</em>, <em>tol_mass=0.1</em>, <em>strict=False</em>, <em>force=True</em>, <em>quiet=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#rms_fit_trj"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.rms_fit_trj" title="Permalink to this definition">¶</a></dt>
<dd><p>RMS-fit trajectory to a reference structure using a selection.</p>
<p>Both reference <em>ref</em> and trajectory <em>traj</em> must be
<code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.Universe</span></code> instances. If they contain a
trajectory then it is used. The output file format is determined
by the file extension of <em>filename</em>. One can also use the same
universe if one wants to fit to the current frame.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first docutils">
<dt><em>traj</em></dt>
<dd><p class="first last">trajectory, <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.Universe</span></code> object</p>
</dd>
<dt><em>reference</em></dt>
<dd><p class="first last">reference coordinates; <code class="xref py py-class docutils literal"><span class="pre">MDAnalysis.Universe</span></code> object
(uses the current time step of the object)</p>
</dd>
<dt><em>select</em></dt>
<dd><ol class="first arabic simple">
<li>any valid selection string for
<a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup.select_atoms" title="MDAnalysis.core.AtomGroup.AtomGroup.select_atoms"><code class="xref py py-meth docutils literal"><span class="pre">select_atoms()</span></code></a> that produces identical
selections in <em>mobile</em> and <em>reference</em>; or</li>
<li>a dictionary <code class="docutils literal"><span class="pre">{'mobile':sel1,</span> <span class="pre">'reference':sel2}</span></code> (the
<a class="reference internal" href="#MDAnalysis.analysis.align.fasta2select" title="MDAnalysis.analysis.align.fasta2select"><code class="xref py py-func docutils literal"><span class="pre">fasta2select()</span></code></a> function returns such a
dictionary based on a <a class="reference external" href="http://www.clustal.org/">ClustalW</a> or <a class="reference external" href="http://www.compbio.dundee.ac.uk/manuals/stamp.4.2/">STAMP</a> sequence alignment); or</li>
<li>a tuple <code class="docutils literal"><span class="pre">(sel1,</span> <span class="pre">sel2)</span></code></li>
</ol>
<p class="last">When using 2. or 3. with <em>sel1</em> and <em>sel2</em> then these selections can also each be
a list of selection strings (to generate a AtomGroup with defined atom order as
described under <a class="reference internal" href="../selections.html#ordered-selections-label"><span>Ordered selections</span></a>).</p>
</dd>
<dt><em>filename</em></dt>
<dd><p class="first last">file name for the RMS-fitted trajectory or pdb; defaults to the
original trajectory filename (from <em>traj</em>) with <em>prefix</em> prepended</p>
</dd>
<dt><em>rmsdfile</em></dt>
<dd><p class="first last">file name for writing the RMSD timeseries [<code class="docutils literal"><span class="pre">None</span></code>]</p>
</dd>
<dt><em>prefix</em></dt>
<dd><p class="first last">prefix for autogenerating the new output filename</p>
</dd>
<dt><em>mass_weighted</em></dt>
<dd><p class="first last">do a mass-weighted RMSD fit</p>
</dd>
<dt><em>tol_mass</em></dt>
<dd><p class="first last">Reject match if the atomic masses for matched atoms differ by more than
<em>tol_mass</em> [0.1]</p>
</dd>
<dt><em>strict</em></dt>
<dd><p class="first">Default: <code class="docutils literal"><span class="pre">False</span></code>
- <code class="docutils literal"><span class="pre">True</span></code>: Will raise <code class="xref py py-exc docutils literal"><span class="pre">SelectioError</span></code> if a single atom does not</p>
<blockquote>
<div><p>match between the two selections.</p>
</div></blockquote>
<ul class="last simple">
<li><code class="docutils literal"><span class="pre">False</span></code>: Will try to prepare a matching selection by dropping
residues with non-matching atoms. See <a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal"><span class="pre">get_matching_atoms()</span></code></a>
for details.</li>
</ul>
</dd>
<dt><em>force</em></dt>
<dd><ul class="first last simple">
<li><code class="docutils literal"><span class="pre">True</span></code>: Overwrite an existing output trajectory (default)</li>
<li><code class="docutils literal"><span class="pre">False</span></code>: simply return if the file already exists</li>
</ul>
</dd>
<dt><em>quiet</em></dt>
<dd><ul class="first simple">
<li><code class="docutils literal"><span class="pre">True</span></code>: suppress progress and logging for levels INFO and below.</li>
<li><code class="docutils literal"><span class="pre">False</span></code>: show all status messages and do not change the the logging
level (default)</li>
</ul>
<div class="last admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If</p>
</div>
</dd>
<dt><em>kwargs</em></dt>
<dd><p class="first last">All other keyword arguments are passed on the trajectory
<a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Writer" title="MDAnalysis.coordinates.base.Writer"><code class="xref py py-class docutils literal"><span class="pre">Writer</span></code></a>; this allows manipulating/fixing
trajectories on the fly (e.g. change the output format by changing the extension of <em>filename</em>
and setting different parameters as described for the corresponding writer).</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><em>filename</em> (either provided or auto-generated)</p>
</td>
</tr>
</tbody>
</table>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.8: </span>Added <em>kwargs</em> to be passed to the trajectory <a class="reference internal" href="../coordinates/base.html#MDAnalysis.coordinates.base.Writer" title="MDAnalysis.coordinates.base.Writer"><code class="xref py py-class docutils literal"><span class="pre">Writer</span></code></a> and
<em>filename</em> is returned.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.10.0: </span>Uses <a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal"><span class="pre">get_matching_atoms()</span></code></a> to work with incomplete selections
and new <em>strict</em> keyword. The new default is to be lenient whereas
the old behavior was the equivalent of <em>strict</em> = <code class="docutils literal"><span class="pre">True</span></code>.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="MDAnalysis.analysis.align.rotation_matrix">
<code class="descclassname">MDAnalysis.analysis.align.</code><code class="descname">rotation_matrix</code><span class="sig-paren">(</span><em>a</em>, <em>b</em>, <em>weights=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#rotation_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.rotation_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the 3x3 rotation matrix for RMSD fitting coordinate sets <em>a</em> and <em>b</em>.</p>
<p>The rotation matrix <em>R</em> transforms <em>a</em> to overlap with <em>b</em> (i.e. <em>b</em> is the
reference structure):</p>
<blockquote>
<div><em>b</em> = <em>R</em> . <em>a</em></div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first docutils">
<dt><em>a</em></dt>
<dd><p class="first last">coordinates that are to be rotated (&#8220;mobile set&#8221;); array of N atoms
of shape N*3 as generated by, e.g.,
<a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup.coordinates" title="MDAnalysis.core.AtomGroup.AtomGroup.coordinates"><code class="xref py py-meth docutils literal"><span class="pre">MDAnalysis.core.AtomGroup.AtomGroup.coordinates()</span></code></a>.</p>
</dd>
<dt><em>b</em></dt>
<dd><p class="first last">reference coordinates; array of N atoms of shape N*3 as generated by,
e.g., <a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup.coordinates" title="MDAnalysis.core.AtomGroup.AtomGroup.coordinates"><code class="xref py py-meth docutils literal"><span class="pre">MDAnalysis.core.AtomGroup.AtomGroup.coordinates()</span></code></a>.</p>
</dd>
<dt><em>weights</em></dt>
<dd><p class="first last">array of floats of size N for doing weighted RMSD fitting (e.g. the
masses of the atoms)</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">(R,</span> <span class="pre">rmsd)</span></code> rmsd and rotation matrix <em>R</em></p>
</td>
</tr>
</tbody>
</table>
<p><em>R</em> can be used as an argument for
<a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup.rotate" title="MDAnalysis.core.AtomGroup.AtomGroup.rotate"><code class="xref py py-meth docutils literal"><span class="pre">MDAnalysis.core.AtomGroup.AtomGroup.rotate()</span></code></a> to generate a rotated
selection, e.g.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">R</span> <span class="o">=</span> <span class="n">rotation_matrix</span><span class="p">(</span><span class="n">A</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s">&#39;backbone&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">coordinates</span><span class="p">(),</span> <span class="n">B</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s">&#39;backbone&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">coordinates</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">A</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">R</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">A</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;rotated.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Note that the function does <em>not</em> shift the centers of mass or geometry;
this needs to be done by the user.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-func docutils literal"><span class="pre">rmsd()</span></code> calculates the RMSD between <em>a</em> and <em>b</em>; for
fitting a whole trajectory it is more efficient to use
<a class="reference internal" href="#MDAnalysis.analysis.align.rms_fit_trj" title="MDAnalysis.analysis.align.rms_fit_trj"><code class="xref py py-func docutils literal"><span class="pre">rms_fit_trj()</span></code></a>. A complete fit of two structures can be
done with <a class="reference internal" href="#MDAnalysis.analysis.align.alignto" title="MDAnalysis.analysis.align.alignto"><code class="xref py py-func docutils literal"><span class="pre">alignto()</span></code></a>.</p>
</div>
</dd></dl>

<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.10.0: </span>Function <a class="reference internal" href="rms.html#MDAnalysis.analysis.rms.rmsd" title="MDAnalysis.analysis.rms.rmsd"><code class="xref py py-func docutils literal"><span class="pre">rmsd()</span></code></a> was removed from
this module and is now exclusively accessible as
<a class="reference internal" href="rms.html#MDAnalysis.analysis.rms.rmsd" title="MDAnalysis.analysis.rms.rmsd"><code class="xref py py-func docutils literal"><span class="pre">rmsd()</span></code></a>.</p>
</div>
</div>
<div class="section" id="helper-functions">
<h2>3.1.1.4. Helper functions<a class="headerlink" href="#helper-functions" title="Permalink to this headline">¶</a></h2>
<p>The following functions are used by the other functions in this
module. They are probably of more interest to developers than to
normal users.</p>
<dl class="function">
<dt id="MDAnalysis.analysis.align.fasta2select">
<code class="descclassname">MDAnalysis.analysis.align.</code><code class="descname">fasta2select</code><span class="sig-paren">(</span><em>fastafilename</em>, <em>is_aligned=False</em>, <em>ref_resids=None</em>, <em>target_resids=None</em>, <em>ref_offset=0</em>, <em>target_offset=0</em>, <em>verbosity=3</em>, <em>alnfilename=None</em>, <em>treefilename=None</em>, <em>clustalw='clustalw2'</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#fasta2select"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.fasta2select" title="Permalink to this definition">¶</a></dt>
<dd><p>Return selection strings that will select equivalent residues.</p>
<p>The function aligns two sequences provided in a FASTA file and
constructs MDAnalysis selection strings of the common atoms. When
these two strings are applied to the two different proteins they
will generate AtomGroups of the aligned residues.</p>
<p><em>fastafilename</em> contains the two un-aligned sequences in FASTA
format. The reference is assumed to be the first sequence, the
target the second. <a class="reference external" href="http://www.clustal.org/">ClustalW</a> produces a pairwise
alignment (which is written to a file with suffix .aln).  The
output contains atom selection strings that select the same atoms
in the two structures.</p>
<p>Unless <em>ref_offset</em> and/or <em>target_offset</em> are specified, the resids
in the structure are assumed to correspond to the positions in the
un-aligned sequence, namely the first residue has resid == 1.</p>
<p>In more complicated cases (e.g. when the resid numbering in the
structure/psf has gaps due to missing parts), simply provide the
sequence of resids as they appear in the psf in <em>ref_resids</em> or
<em>target_resids</em>, e.g.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">target_resids</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span><span class="o">.</span><span class="n">resid</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">trj</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s">&#39;name CA&#39;</span><span class="p">)]</span>
</pre></div>
</div>
<p>(This translation table <em>is</em> combined with any value for <em>xxx_offset</em>!)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first docutils">
<dt><em>fastafilename</em></dt>
<dd><p class="first last">FASTA file with first sequence as reference and
second the one to be aligned (ORDER IS IMPORTANT!)</p>
</dd>
<dt><em>is_aligned</em></dt>
<dd><p class="first last">False: run clustalw for sequence alignment; True: use
the alignment in the file (e.g. from STAMP) [<code class="docutils literal"><span class="pre">False</span></code>]</p>
</dd>
<dt><em>ref_offset</em></dt>
<dd><p class="first last">add this number to the column number in the FASTA file
to get the original residue number</p>
</dd>
<dt><em>target_offset</em></dt>
<dd><p class="first last">same for the target</p>
</dd>
<dt><em>ref_resids</em></dt>
<dd><p class="first last">sequence of resids as they appear in the reference structure</p>
</dd>
<dt><em>target_resids</em></dt>
<dd><p class="first last">sequence of resids as they appear in the target</p>
</dd>
<dt><em>alnfilename</em></dt>
<dd><p class="first last">filename of ClustalW alignment (clustal format) that is
produced by <em>clustalw</em> when <em>is_aligned</em> = <code class="docutils literal"><span class="pre">False</span></code>.
<code class="docutils literal"><span class="pre">None</span></code> uses the name and path of <em>fastafilename</em> and
subsititutes the suffix with &#8216;.aln&#8217;.[<code class="docutils literal"><span class="pre">None</span></code>]</p>
</dd>
<dt><em>treefilename</em></dt>
<dd><p class="first last">filename of ClustalW guide tree (Newick format);
if <code class="docutils literal"><span class="pre">None</span></code>  the the filename is generated from <em>alnfilename</em>
with the suffix &#8216;.dnd&#8217; instead of &#8216;.aln&#8217; [<code class="docutils literal"><span class="pre">None</span></code>]</p>
</dd>
<dt><em>clustalw</em></dt>
<dd><p class="first last">path to the ClustalW (or ClustalW2) binary; only
needed for <em>is_aligned</em> = <code class="docutils literal"><span class="pre">False</span></code> [&#8220;clustalw2&#8221;]</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><em>select_dict</em></dt>
<dd><p class="first last">dictionary with &#8216;reference&#8217; and &#8216;mobile&#8217; selection string
that can be used immediately in <a class="reference internal" href="#MDAnalysis.analysis.align.rms_fit_trj" title="MDAnalysis.analysis.align.rms_fit_trj"><code class="xref py py-func docutils literal"><span class="pre">rms_fit_trj()</span></code></a> as
<code class="docutils literal"><span class="pre">select=select_dict</span></code>.</p>
</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="MDAnalysis.analysis.align.get_matching_atoms">
<code class="descclassname">MDAnalysis.analysis.align.</code><code class="descname">get_matching_atoms</code><span class="sig-paren">(</span><em>ag1</em>, <em>ag2</em>, <em>tol_mass=0.1</em>, <em>strict=False</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#get_matching_atoms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.get_matching_atoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Return two atom groups with one-to-one matched atoms.</p>
<p>The function takes two <a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>
instances <em>ag1</em> and <em>ag2</em> and returns two atom groups <em>g1</em> and <em>g2</em> that
consist of atoms so that the mass of atom <code class="docutils literal"><span class="pre">g1[0]</span></code> is the same as the mass
of atom <code class="docutils literal"><span class="pre">g2[0]</span></code>, <code class="docutils literal"><span class="pre">g1[1]</span></code> and <code class="docutils literal"><span class="pre">g2[1]</span></code> etc.</p>
<p>The current implementation is very simplistic and works on a per-residue basis:</p>
<ol class="arabic simple">
<li>The two groups must contain the same number of residues.</li>
<li>Any residues in each group that have differing number of atoms are discarded.</li>
<li>The masses of corresponding atoms are compared. and if any masses differ
by more than <em>tol_mass</em> the test is considered failed and a
<code class="xref py py-exc docutils literal"><span class="pre">SelectionError</span></code> is raised.</li>
</ol>
<p>The log file (see <a class="reference internal" href="../lib/log.html#MDAnalysis.start_logging" title="MDAnalysis.start_logging"><code class="xref py py-func docutils literal"><span class="pre">MDAnalysis.start_logging()</span></code></a>) will contain detailed
information about mismatches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Arguments:</th><td class="field-body"><dl class="first docutils">
<dt><em>ag1</em>, <em>ag2</em></dt>
<dd><p class="first last"><a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> instances that are compared</p>
</dd>
</dl>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Keywords:</th><td class="field-body"><dl class="first docutils">
<dt><em>tol_mass</em></dt>
<dd><p class="first last">Reject if the atomic masses for matched atoms differ by more than
<em>tol_mass</em> [0.1]</p>
</dd>
<dt><em>strict</em></dt>
<dd><dl class="first last docutils">
<dt><code class="docutils literal"><span class="pre">True</span></code></dt>
<dd><p class="first last">Will raise <code class="xref py py-exc docutils literal"><span class="pre">SelectioError</span></code> if a single atom does not
match between the two selections.</p>
</dd>
<dt><code class="docutils literal"><span class="pre">False</span></code> [default]</dt>
<dd><p class="first last">Will try to prepare a matching selection by dropping
residues with non-matching atoms. See <a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal"><span class="pre">get_matching_atoms()</span></code></a>
for details.</p>
</dd>
</dl>
</dd>
</dl>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Tuple <code class="docutils literal"><span class="pre">(g1,</span> <span class="pre">g2)</span></code> with <a class="reference internal" href="../core/AtomGroup.html#MDAnalysis.core.AtomGroup.AtomGroup" title="MDAnalysis.core.AtomGroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> instances
that match, atom by atom. The groups are either the original groups if all matches
or slices of the original groups.</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><code class="xref py py-exc docutils literal"><span class="pre">SelectionError</span></code> if the number of residues does not match or if in the final
matching masses differ by more than <em>tol</em>.</p>
</td>
</tr>
</tbody>
</table>
<p>The algorithm could be improved by using e.g. the Needleman-Wunsch
algorithm in <code class="xref py py-mod docutils literal"><span class="pre">Bio.profile2</span></code> to align atoms in each residue (doing a
global alignment is too expensive).</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.8.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 0.10.0: </span>Renamed from <code class="xref py py-func docutils literal"><span class="pre">check_same_atoms()</span></code> to <a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal"><span class="pre">get_matching_atoms()</span></code></a> and now returns
matching atomgroups (possibly with residues removed)</p>
</div>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="contacts.html" title="3.1.2. Native contacts analysis — MDAnalysis.analysis.contacts"
             >next</a> |</li>
        <li class="right" >
          <a href="../analysis_modules.html" title="3. Analysis modules"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">MDAnalysis 0.11.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../analysis_modules.html" >3. Analysis modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2005-2015, Naveen Michaud-Agrawal, Elizabeth J. Denning, Joshua Adelman,
    Christian Beckstein (logo), Alejandro Bernardin, Sébastien Buchoux,
    David Caplan, Matthieu Chavent, Xavier Deupi, Jan Domański, David L. Dotson
    Lennard van der Feltz, Philip Fowler, Joseph Goose, Richard J. Gowers, Lukas Grossar,
    Benjamin Hall, Joe Jordan, Jinju Lu, Robert McGibbon, Alex Nesterenko,
    Manuel Nuno Melo, Caio S. Souza, Danny Parton, Joshua L. Phillips, Tyler Reddy,
    Paul Rigor, Sean L. Seyler, Andy Somogyi, Lukas Stelzl, Gorman Stock, Isaac Virshup,
    Zhuyi Xue, Carlos Yáñez S.,
    and Oliver Beckstein.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>