
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>4.2.1. Coordinate fitting and alignment — MDAnalysis.analysis.align &#8212; MDAnalysis 0.20.1 documentation</title>
    <link rel="stylesheet" href="../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/language_data.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="search" type="application/opensearchdescription+xml"
          title="Search within MDAnalysis 0.20.1 documentation"
          href="../../_static/opensearch.xml"/>
    <link rel="shortcut icon" href="../../_static/mdanalysis-logo.ico"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="4.2.2. Native contacts analysis — MDAnalysis.analysis.contacts" href="contacts.html" />
    <link rel="prev" title="4.1.1. Analysis building blocks — MDAnalysis.analysis.base" href="base.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-MDAnalysis.analysis.align"></span><div class="section" id="coordinate-fitting-and-alignment-mdanalysis-analysis-align">
<h1><span class="section-number">4.2.1. </span>Coordinate fitting and alignment — <a class="reference internal" href="#module-MDAnalysis.analysis.align" title="MDAnalysis.analysis.align"><code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.align</span></code></a><a class="headerlink" href="#coordinate-fitting-and-alignment-mdanalysis-analysis-align" title="Permalink to this headline">¶</a></h1>
<dl class="field-list simple">
<dt class="field-odd">Author</dt>
<dd class="field-odd"><p>Oliver Beckstein, Joshua Adelman</p>
</dd>
<dt class="field-even">Year</dt>
<dd class="field-even"><p>2010–2013</p>
</dd>
<dt class="field-odd">Copyright</dt>
<dd class="field-odd"><p>GNU Public License v3</p>
</dd>
</dl>
<p>The module contains functions to fit a target structure to a reference
structure. They use the fast QCP algorithm to calculate the root mean
square distance (RMSD) between two coordinate sets <a class="reference internal" href="../lib/qcprot.html#theobald2005" id="id1"><span>[Theobald2005]</span></a> and
the rotation matrix <em>R</em> that minimizes the RMSD <a class="reference internal" href="../lib/qcprot.html#liu2010" id="id2"><span>[Liu2010]</span></a>. (Please
cite these references when using this module.).</p>
<p>Typically, one selects a group of atoms (such as the C-alphas),
calculates the RMSD and transformation matrix, and applys the
transformation to the current frame of a trajectory to obtain the
rotated structure. The <a class="reference internal" href="#MDAnalysis.analysis.align.alignto" title="MDAnalysis.analysis.align.alignto"><code class="xref py py-func docutils literal notranslate"><span class="pre">alignto()</span></code></a> and <a class="reference internal" href="#MDAnalysis.analysis.align.AlignTraj" title="MDAnalysis.analysis.align.AlignTraj"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignTraj</span></code></a>
functions can be used to do this for individual frames and
trajectories respectively.</p>
<p>The <a class="reference internal" href="#rms-fitting-tutorial"><span class="std std-ref">RMS-fitting tutorial</span></a> shows how to do the individual steps
manually and explains the intermediate steps.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="rms.html#module-MDAnalysis.analysis.rms" title="MDAnalysis.analysis.rms"><code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.rms</span></code></a></dt><dd><p>contains functions to compute RMSD (when structural alignment is not required)</p>
</dd>
<dt><a class="reference internal" href="../lib/qcprot.html#module-MDAnalysis.lib.qcprot" title="MDAnalysis.lib.qcprot"><code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.lib.qcprot</span></code></a></dt><dd><p>implements the fast RMSD algorithm.</p>
</dd>
</dl>
</div>
<div class="section" id="rms-fitting-tutorial">
<span id="id3"></span><h2><span class="section-number">4.2.1.1. </span>RMS-fitting tutorial<a class="headerlink" href="#rms-fitting-tutorial" title="Permalink to this headline">¶</a></h2>
<p>The example uses files provided as part of the MDAnalysis test suite
(in the variables <code class="xref py py-data docutils literal notranslate"><span class="pre">PSF</span></code>,
<code class="xref py py-data docutils literal notranslate"><span class="pre">DCD</span></code>, and
<code class="xref py py-data docutils literal notranslate"><span class="pre">PDB_small</span></code>). For all further
examples execute first</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">MDAnalysis</span> <span class="k">as</span> <span class="nn">mda</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.analysis</span> <span class="kn">import</span> <span class="n">align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.analysis.rms</span> <span class="kn">import</span> <span class="n">rmsd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">,</span> <span class="n">PDB_small</span>
</pre></div>
</div>
<p>In the simplest case, we can simply calculate the C-alpha RMSD between
two structures, using <code class="xref py py-func docutils literal notranslate"><span class="pre">rmsd()</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ref</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PDB_small</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span><span class="n">DCD</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rmsd</span><span class="p">(</span><span class="n">mobile</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span> <span class="n">ref</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span><span class="p">)</span>
<span class="go">16.282308620224068</span>
</pre></div>
</div>
<p>Note that in this example translations have not been removed. In order
to look at the pure rotation one needs to superimpose the centres of
mass (or geometry) first:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rmsd</span><span class="p">(</span><span class="n">mobile</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span> <span class="n">ref</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">12.639693690256898</span>
</pre></div>
</div>
<p>This has only done a translational superposition. If you want to also do a
rotational superposition use the superposition keyword. This will calculate a
minimized RMSD between the reference and mobile structure.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rmsd</span><span class="p">(</span><span class="n">mobile</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span> <span class="n">ref</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>     <span class="n">superposition</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">6.8093965864717951</span>
</pre></div>
</div>
<p>The rotation matrix that superimposes <em>mobile</em> on <em>ref</em> while
minimizing the CA-RMSD is obtained with the <a class="reference internal" href="#MDAnalysis.analysis.align.rotation_matrix" title="MDAnalysis.analysis.align.rotation_matrix"><code class="xref py py-func docutils literal notranslate"><span class="pre">rotation_matrix()</span></code></a>
function</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mobile0</span> <span class="o">=</span> <span class="n">mobile</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span> <span class="o">-</span> <span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">center_of_mass</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ref0</span> <span class="o">=</span> <span class="n">ref</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span> <span class="o">-</span> <span class="n">ref</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">center_of_mass</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">R</span><span class="p">,</span> <span class="n">rmsd</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">rotation_matrix</span><span class="p">(</span><span class="n">mobile0</span><span class="p">,</span> <span class="n">ref0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">rmsd</span>
<span class="go">6.8093965864717951</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">R</span>
<span class="go">[[ 0.14514539 -0.27259113  0.95111876]</span>
<span class="go"> [ 0.88652593  0.46267112 -0.00268642]</span>
<span class="go"> [-0.43932289  0.84358136  0.30881368]]</span>
</pre></div>
</div>
<p>Putting all this together one can superimpose all of <em>mobile</em> onto <em>ref</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="o">-</span><span class="n">mobile</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">center_of_mass</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">R</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">ref</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">center_of_mass</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;mobile_on_ref.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="common-usage">
<h2><span class="section-number">4.2.1.2. </span>Common usage<a class="headerlink" href="#common-usage" title="Permalink to this headline">¶</a></h2>
<p>To <strong>fit a single structure</strong> with <a class="reference internal" href="#MDAnalysis.analysis.align.alignto" title="MDAnalysis.analysis.align.alignto"><code class="xref py py-func docutils literal notranslate"><span class="pre">alignto()</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ref</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">PDB_small</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mobile</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">)</span>     <span class="c1"># we use the first frame</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="o">.</span><span class="n">alignto</span><span class="p">(</span><span class="n">mobile</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">select</span><span class="o">=</span><span class="s2">&quot;protein and name CA&quot;</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="s2">&quot;mass&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>This will change <em>all</em> coordinates in <em>mobile</em> so that the protein
C-alpha atoms are optimally superimposed (translation and rotation).</p>
<p>To <strong>fit a whole trajectory</strong> to a reference structure with the
<a class="reference internal" href="#MDAnalysis.analysis.align.AlignTraj" title="MDAnalysis.analysis.align.AlignTraj"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignTraj</span></code></a> class:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ref</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">PDB_small</span><span class="p">)</span>   <span class="c1"># reference structure 1AKE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">trj</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PSF</span><span class="p">,</span> <span class="n">DCD</span><span class="p">)</span>         <span class="c1"># trajectory of change 1AKE-&gt;4AKE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">AlignTraj</span><span class="p">(</span><span class="n">trj</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s1">&#39;rmsfit.dcd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>It is also possible to align two arbitrary structures by providing a
mapping between atoms based on a sequence alignment. This allows
fitting of structural homologs or wild type and mutant.</p>
<p>If a alignment was provided as “sequences.aln” one would first produce
the appropriate MDAnalysis selections with the <a class="reference internal" href="#MDAnalysis.analysis.align.fasta2select" title="MDAnalysis.analysis.align.fasta2select"><code class="xref py py-func docutils literal notranslate"><span class="pre">fasta2select()</span></code></a>
function and then feed the resulting dictionary to <a class="reference internal" href="#MDAnalysis.analysis.align.AlignTraj" title="MDAnalysis.analysis.align.AlignTraj"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignTraj</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seldict</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">fasta2select</span><span class="p">(</span><span class="s1">&#39;sequences.aln&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">AlignTraj</span><span class="p">(</span><span class="n">trj</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s1">&#39;rmsfit.dcd&#39;</span><span class="p">,</span> <span class="n">select</span><span class="o">=</span><span class="n">seldict</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>(See the documentation of the functions for this advanced usage.)</p>
</div>
<div class="section" id="functions-and-classes">
<h2><span class="section-number">4.2.1.3. </span>Functions and Classes<a class="headerlink" href="#functions-and-classes" title="Permalink to this headline">¶</a></h2>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.10.0: </span>Function <a class="reference internal" href="rms.html#MDAnalysis.analysis.rms.rmsd" title="MDAnalysis.analysis.rms.rmsd"><code class="xref py py-func docutils literal notranslate"><span class="pre">rmsd()</span></code></a> was removed from
this module and is now exclusively accessible as
<a class="reference internal" href="rms.html#MDAnalysis.analysis.rms.rmsd" title="MDAnalysis.analysis.rms.rmsd"><code class="xref py py-func docutils literal notranslate"><span class="pre">rmsd()</span></code></a>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.16.0: </span>Function <code class="xref py py-func docutils literal notranslate"><span class="pre">rms_fit_trj()</span></code> deprecated
in favor of <a class="reference internal" href="#MDAnalysis.analysis.align.AlignTraj" title="MDAnalysis.analysis.align.AlignTraj"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignTraj</span></code></a> class.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.17.0: </span>removed deprecated <code class="xref py py-func docutils literal notranslate"><span class="pre">rms_fit_trj()</span></code></p>
</div>
<dl class="function">
<dt id="MDAnalysis.analysis.align.alignto">
<code class="sig-prename descclassname">MDAnalysis.analysis.align.</code><code class="sig-name descname">alignto</code><span class="sig-paren">(</span><em class="sig-param">mobile</em>, <em class="sig-param">reference</em>, <em class="sig-param">select='all'</em>, <em class="sig-param">weights=None</em>, <em class="sig-param">subselection=None</em>, <em class="sig-param">tol_mass=0.1</em>, <em class="sig-param">strict=False</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#alignto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.alignto" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a spatial superposition by minimizing the RMSD.</p>
<p>Spatially align the group of atoms <cite>mobile</cite> to <cite>reference</cite> by
doing a RMSD fit on <cite>select</cite> atoms.</p>
<p>The superposition is done in the following way:</p>
<ol class="arabic simple">
<li><p>A rotation matrix is computed that minimizes the RMSD between
the coordinates of <cite>mobile.select_atoms(sel1)</cite> and
<cite>reference.select_atoms(sel2)</cite>; before the rotation, <cite>mobile</cite> is
translated so that its center of geometry (or center of mass)
coincides with the one of <cite>reference</cite>. (See below for explanation of
how <em>sel1</em> and <em>sel2</em> are derived from <cite>select</cite>.)</p></li>
<li><p>All atoms in <a class="reference internal" href="../core/universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a> that
contain <cite>mobile</cite> are shifted and rotated. (See below for how
to change this behavior through the <cite>subselection</cite> keyword.)</p></li>
</ol>
<p>The <cite>mobile</cite> and <cite>reference</cite> atom groups can be constructed so that they
already match atom by atom. In this case, <cite>select</cite> should be set to “all”
(or <code class="docutils literal notranslate"><span class="pre">None</span></code>) so that no further selections are applied to <cite>mobile</cite> and
<cite>reference</cite>, therefore preserving the exact atom ordering (see
<a class="reference internal" href="../selections.html#ordered-selections-label"><span class="std std-ref">Ordered selections</span></a>).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The atom order for <cite>mobile</cite> and <cite>reference</cite> is <em>only</em>
preserved when <cite>select</cite> is either “all” or <code class="docutils literal notranslate"><span class="pre">None</span></code>. In any other case,
a new selection will be made that will sort the resulting AtomGroup by
index and therefore destroy the correspondence between the two groups.
<strong>It is safest not to mix ordered AtomGroups with selection strings.</strong></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mobile</strong> (<a class="reference internal" href="../core/universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><em>Universe</em></a><em> or </em><a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><em>AtomGroup</em></a>) – structure to be aligned, a
<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> or a whole
<a class="reference internal" href="../core/universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a></p></li>
<li><p><strong>reference</strong> (<a class="reference internal" href="../core/universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><em>Universe</em></a><em> or </em><a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><em>AtomGroup</em></a>) – reference structure, a <a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>
or a whole <a class="reference internal" href="../core/universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><code class="xref py py-class docutils literal notranslate"><span class="pre">Universe</span></code></a></p></li>
<li><p><strong>select</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a><em> (</em><em>optional</em><em>)</em>) – <p>The selection to operate on; can be one of:</p>
<ol class="arabic simple">
<li><p>any valid selection string for
<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup.select_atoms" title="MDAnalysis.core.groups.AtomGroup.select_atoms"><code class="xref py py-meth docutils literal notranslate"><span class="pre">select_atoms()</span></code></a> that
produces identical selections in <cite>mobile</cite> and <cite>reference</cite>; or</p></li>
<li><p>a dictionary <code class="docutils literal notranslate"><span class="pre">{'mobile':</span> <span class="pre">sel1,</span> <span class="pre">'reference':</span> <span class="pre">sel2}</span></code> where <em>sel1</em>
and <em>sel2</em> are valid selection strings that are applied to
<cite>mobile</cite> and <cite>reference</cite> respectively (the
<a class="reference internal" href="#MDAnalysis.analysis.align.fasta2select" title="MDAnalysis.analysis.align.fasta2select"><code class="xref py py-func docutils literal notranslate"><span class="pre">MDAnalysis.analysis.align.fasta2select()</span></code></a> function returns such
a dictionary based on a <a class="reference external" href="http://www.clustal.org/">ClustalW</a> or <a class="reference external" href="http://www.compbio.dundee.ac.uk/manuals/stamp.4.2/">STAMP</a> sequence alignment); or</p></li>
<li><p>a tuple <code class="docutils literal notranslate"><span class="pre">(sel1,</span> <span class="pre">sel2)</span></code></p></li>
</ol>
<p>When using 2. or 3. with <em>sel1</em> and <em>sel2</em> then these selection strings
are applied to <cite>atomgroup</cite> and <cite>reference</cite> respectively and should
generate <em>groups of equivalent atoms</em>.  <em>sel1</em> and <em>sel2</em> can each also
be a <em>list of selection strings</em> to generate a
<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> with defined atom order as
described under <a class="reference internal" href="../selections.html#ordered-selections-label"><span class="std std-ref">Ordered selections</span></a>).</p>
</p></li>
<li><p><strong>weights</strong> ({“mass”, <code class="docutils literal notranslate"><span class="pre">None</span></code>} or array_like (optional)) – choose weights. With <code class="docutils literal notranslate"><span class="pre">&quot;mass&quot;</span></code> uses masses as weights; with <code class="docutils literal notranslate"><span class="pre">None</span></code>
weigh each atom equally. If a float array of the same length as
<cite>mobile</cite> is provided, use each element of the <cite>array_like</cite> as a
weight for the corresponding atom in <cite>mobile</cite>.</p></li>
<li><p><strong>tol_mass</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a><em> (</em><em>optional</em><em>)</em>) – Reject match if the atomic masses for matched atoms differ by more than
<cite>tol_mass</cite>, default [0.1]</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – <dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">True</span></code></dt><dd><p>Will raise <code class="xref py py-exc docutils literal notranslate"><span class="pre">SelectionError</span></code> if a single atom does not
match between the two selections.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">False</span></code> [default]</dt><dd><p>Will try to prepare a matching selection by dropping
residues with non-matching atoms. See <a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_matching_atoms()</span></code></a>
for details.</p>
</dd>
</dl>
</p></li>
<li><p><strong>subselection</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> or </em><a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><em>AtomGroup</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a><em> (</em><em>optional</em><em>)</em>) – <p>Apply the transformation only to this selection.</p>
<dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">None</span></code> [default]</dt><dd><p>Apply to <code class="docutils literal notranslate"><span class="pre">mobile.universe.atoms</span></code> (i.e., all atoms in the
context of the selection from <cite>mobile</cite> such as the rest of a
protein, ligands and the surrounding water)</p>
</dd>
<dt><em>selection-string</em></dt><dd><p>Apply to <code class="docutils literal notranslate"><span class="pre">mobile.select_atoms(selection-string)</span></code></p>
</dd>
<dt><a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a></dt><dd><p>Apply to the arbitrary group of atoms</p>
</dd>
</dl>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>old_rmsd</strong> (<em>float</em>) – RMSD before spatial alignment</p></li>
<li><p><strong>new_rmsd</strong> (<em>float</em>) – RMSD after spatial alignment</p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#MDAnalysis.analysis.align.AlignTraj" title="MDAnalysis.analysis.align.AlignTraj"><code class="xref py py-func docutils literal notranslate"><span class="pre">AlignTraj()</span></code></a></dt><dd><p>More efficient method for RMSD-fitting trajectories.</p>
</dd>
</dl>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8: </span>Added check that the two groups describe the same atoms including
the new <em>tol_mass</em> keyword.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.10.0: </span>Uses <a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_matching_atoms()</span></code></a> to work with incomplete selections
and new <cite>strict</cite> keyword. The new default is to be lenient whereas
the old behavior was the equivalent of <code class="docutils literal notranslate"><span class="pre">strict</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.16.0: </span>new general ‘weights’ kwarg replace <cite>mass_weighted</cite>, deprecated <cite>mass_weighted</cite></p>
</div>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.16.0: </span>Instead of <code class="docutils literal notranslate"><span class="pre">mass_weighted=True</span></code> use new <code class="docutils literal notranslate"><span class="pre">weights='mass'</span></code></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.17.0: </span>Deprecated keyword <cite>mass_weighted</cite> was removed.</p>
</div>
</dd></dl>

<dl class="class">
<dt id="MDAnalysis.analysis.align.AlignTraj">
<em class="property">class </em><code class="sig-prename descclassname">MDAnalysis.analysis.align.</code><code class="sig-name descname">AlignTraj</code><span class="sig-paren">(</span><em class="sig-param">mobile</em>, <em class="sig-param">reference</em>, <em class="sig-param">select='all'</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">prefix='rmsfit_'</em>, <em class="sig-param">weights=None</em>, <em class="sig-param">tol_mass=0.1</em>, <em class="sig-param">strict=False</em>, <em class="sig-param">force=True</em>, <em class="sig-param">in_memory=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#AlignTraj"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.AlignTraj" title="Permalink to this definition">¶</a></dt>
<dd><p>RMS-align trajectory to a reference structure using a selection.</p>
<p>Both the reference <cite>reference</cite> and the trajectory <cite>mobile</cite> must be
<code class="xref py py-class docutils literal notranslate"><span class="pre">MDAnalysis.Universe</span></code> instances. If they contain a trajectory then
it is used. The output file format is determined by the file extension of
<cite>filename</cite>. One can also use the same universe if one wants to fit to the
current frame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mobile</strong> (<a class="reference internal" href="../core/universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><em>Universe</em></a>) – Universe containing trajectory to be fitted to reference</p></li>
<li><p><strong>reference</strong> (<a class="reference internal" href="../core/universe.html#MDAnalysis.core.universe.Universe" title="MDAnalysis.core.universe.Universe"><em>Universe</em></a>) – Universe containing trajectory frame to be used as reference</p></li>
<li><p><strong>select</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – Set as default to all, is used for Universe.select_atoms to choose
subdomain to be fitted against</p></li>
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – Provide a filename for results to be written to</p></li>
<li><p><strong>prefix</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – Provide a string to prepend to filename for results to be written
to</p></li>
<li><p><strong>weights</strong> ({“mass”, <code class="docutils literal notranslate"><span class="pre">None</span></code>} or array_like (optional)) – choose weights. With <code class="docutils literal notranslate"><span class="pre">&quot;mass&quot;</span></code> uses masses of <cite>reference</cite> as
weights; with <code class="docutils literal notranslate"><span class="pre">None</span></code> weigh each atom equally. If a float array of
the same length as the selection is provided, use each element of
the <cite>array_like</cite> as a weight for the corresponding atom in the
selection.</p></li>
<li><p><strong>tol_mass</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a><em> (</em><em>optional</em><em>)</em>) – Tolerance given to <cite>get_matching_atoms</cite> to find appropriate atoms</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Force <cite>get_matching_atoms</cite> to fail if atoms can’t be found using
exact methods</p></li>
<li><p><strong>force</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Force overwrite of filename for rmsd-fitting</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em> (</em><em>optional</em><em>)</em>) – First frame of trajectory to analyse, Default: 0</p></li>
<li><p><strong>stop</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em> (</em><em>optional</em><em>)</em>) – Last frame of trajectory to analyse, Default: -1</p></li>
<li><p><strong>step</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em> (</em><em>optional</em><em>)</em>) – Step between frames to analyse, Default: 1</p></li>
<li><p><strong>in_memory</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – <em>Permanently</em> switch <cite>mobile</cite> to an in-memory trajectory
so that alignment can be done in-place, which can improve
performance substantially in some cases. In this case, no file
is written out (<cite>filename</cite> and <cite>prefix</cite> are ignored) and only
the coordinates of <cite>mobile</cite> are <em>changed in memory</em>.</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – Set logger to show more information and show detailed progress of
the calculation if set to <code class="docutils literal notranslate"><span class="pre">True</span></code>; the default is <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="MDAnalysis.analysis.align.AlignTraj.reference_atoms">
<code class="sig-name descname">reference_atoms</code><a class="headerlink" href="#MDAnalysis.analysis.align.AlignTraj.reference_atoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Atoms of the reference structure to be aligned against</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup">AtomGroup</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.analysis.align.AlignTraj.mobile_atoms">
<code class="sig-name descname">mobile_atoms</code><a class="headerlink" href="#MDAnalysis.analysis.align.AlignTraj.mobile_atoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Atoms inside each trajectory frame to be rmsd_aligned</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup">AtomGroup</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.analysis.align.AlignTraj.rmsd">
<code class="sig-name descname">rmsd</code><a class="headerlink" href="#MDAnalysis.analysis.align.AlignTraj.rmsd" title="Permalink to this definition">¶</a></dt>
<dd><p>Array of the rmsd values of the least rmsd between the mobile_atoms
and reference_atoms after superposition and minimimization of rmsd</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Array</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="MDAnalysis.analysis.align.AlignTraj.filename">
<code class="sig-name descname">filename</code><a class="headerlink" href="#MDAnalysis.analysis.align.AlignTraj.filename" title="Permalink to this definition">¶</a></dt>
<dd><p>String reflecting the filename of the file where mobile_atoms
positions will be written to upon running RMSD alignment</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Notes</p>
<ul class="simple">
<li><p>If set to <code class="docutils literal notranslate"><span class="pre">verbose=False</span></code>, it is recommended to wrap the statement
in a <code class="docutils literal notranslate"><span class="pre">try</span> <span class="pre">...</span>&#160; <span class="pre">finally</span></code> to guarantee restoring of the log level in
the case of an exception.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">in_memory</span></code> option changes the <cite>mobile</cite> universe to an
in-memory representation (see <a class="reference internal" href="../coordinates/memory.html#module-MDAnalysis.coordinates.memory" title="MDAnalysis.coordinates.memory"><code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.coordinates.memory</span></code></a>)
for the remainder of the Python session. If <code class="docutils literal notranslate"><span class="pre">mobile.trajectory</span></code> is
already a <code class="xref py py-class docutils literal notranslate"><span class="pre">MemoryReader</span></code> then it is <em>always</em> treated as if
<code class="docutils literal notranslate"><span class="pre">in_memory</span></code> had been set to <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p></li>
</ul>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.19.1: </span>Default <code class="docutils literal notranslate"><span class="pre">filename</span></code> directory will change in 1.0 to the current directory.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.16.0: </span>new general <code class="docutils literal notranslate"><span class="pre">weights</span></code> kwarg replace <code class="docutils literal notranslate"><span class="pre">mass_weights</span></code></p>
</div>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.16.0: </span>Instead of <code class="docutils literal notranslate"><span class="pre">mass_weighted=True</span></code> use new <code class="docutils literal notranslate"><span class="pre">weights='mass'</span></code></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.17.0: </span>removed deprecated <cite>mass_weighted</cite> keyword</p>
</div>
</dd></dl>

<dl class="function">
<dt id="MDAnalysis.analysis.align.rotation_matrix">
<code class="sig-prename descclassname">MDAnalysis.analysis.align.</code><code class="sig-name descname">rotation_matrix</code><span class="sig-paren">(</span><em class="sig-param">a</em>, <em class="sig-param">b</em>, <em class="sig-param">weights=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#rotation_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.rotation_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the 3x3 rotation matrix <cite>R</cite> for RMSD fitting coordinate
sets <cite>a</cite> and <cite>b</cite>.</p>
<p>The rotation matrix <cite>R</cite> transforms vector <cite>a</cite> to overlap with
vector <cite>b</cite> (i.e., <cite>b</cite> is the reference structure):</p>
<div class="math notranslate nohighlight">
\[\mathbf{b} = \mathsf{R} \cdot \mathbf{a}\]</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<em>array_like</em>) – coordinates that are to be rotated (“mobile set”); array of N atoms
of shape N*3 as generated by, e.g.,
<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup.positions" title="MDAnalysis.core.groups.AtomGroup.positions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup.positions</span></code></a>.</p></li>
<li><p><strong>b</strong> (<em>array_like</em>) – reference coordinates; array of N atoms of shape N*3 as generated by,
e.g., <a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup.positions" title="MDAnalysis.core.groups.AtomGroup.positions"><code class="xref py py-attr docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup.positions</span></code></a>.</p></li>
<li><p><strong>weights</strong> (<em>array_like</em><em> (</em><em>optional</em><em>)</em>) – array of floats of size N for doing weighted RMSD fitting (e.g. the
masses of the atoms)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>R</strong> (<em>ndarray</em>) – rotation matrix</p></li>
<li><p><strong>rmsd</strong> (<em>float</em>) – RMSD between <cite>a</cite> and <cite>b</cite> before rotation</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">(R,</span> <span class="pre">rmsd)</span></code> rmsd and rotation matrix <em>R</em></p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p><cite>R</cite> can be used as an argument for
<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup.rotate" title="MDAnalysis.core.groups.AtomGroup.rotate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MDAnalysis.core.groups.AtomGroup.rotate()</span></code></a> to generate a rotated
selection, e.g.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">R</span> <span class="o">=</span> <span class="n">rotation_matrix</span><span class="p">(</span><span class="n">A</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;backbone&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>                    <span class="n">B</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;backbone&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">positions</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">A</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">rotate</span><span class="p">(</span><span class="n">R</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">A</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;rotated.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>The function does <em>not</em> shift the centers of mass or geometry;
this needs to be done by the user.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="rms.html#MDAnalysis.analysis.rms.rmsd" title="MDAnalysis.analysis.rms.rmsd"><code class="xref py py-func docutils literal notranslate"><span class="pre">MDAnalysis.analysis.rms.rmsd()</span></code></a></dt><dd><p>Calculates the RMSD between <em>a</em> and <em>b</em>.</p>
</dd>
<dt><a class="reference internal" href="#MDAnalysis.analysis.align.alignto" title="MDAnalysis.analysis.align.alignto"><code class="xref py py-func docutils literal notranslate"><span class="pre">alignto()</span></code></a></dt><dd><p>A complete fit of two structures.</p>
</dd>
<dt><a class="reference internal" href="#MDAnalysis.analysis.align.AlignTraj" title="MDAnalysis.analysis.align.AlignTraj"><code class="xref py py-func docutils literal notranslate"><span class="pre">AlignTraj()</span></code></a></dt><dd><p>Fit a whole trajectory.</p>
</dd>
</dl>
</div>
</dd></dl>

</div>
<div class="section" id="helper-functions">
<h2><span class="section-number">4.2.1.4. </span>Helper functions<a class="headerlink" href="#helper-functions" title="Permalink to this headline">¶</a></h2>
<p>The following functions are used by the other functions in this
module. They are probably of more interest to developers than to
normal users.</p>
<dl class="function">
<dt id="MDAnalysis.analysis.align._fit_to">
<code class="sig-prename descclassname">MDAnalysis.analysis.align.</code><code class="sig-name descname">_fit_to</code><span class="sig-paren">(</span><em class="sig-param">mobile_coordinates</em>, <em class="sig-param">ref_coordinates</em>, <em class="sig-param">mobile_atoms</em>, <em class="sig-param">mobile_com</em>, <em class="sig-param">ref_com</em>, <em class="sig-param">weights=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#_fit_to"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align._fit_to" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform an rmsd-fitting to determine rotation matrix and align atoms</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mobile_coordinates</strong> (<em>ndarray</em>) – Coordinates of atoms to be aligned</p></li>
<li><p><strong>ref_coordinates</strong> (<em>ndarray</em>) – Coordinates of atoms to be fit against</p></li>
<li><p><strong>mobile_atoms</strong> (<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><em>AtomGroup</em></a>) – Atoms to be translated</p></li>
<li><p><strong>mobile_com</strong> (<em>ndarray</em>) – array of xyz coordinate of mobile center of mass</p></li>
<li><p><strong>ref_com</strong> (<em>ndarray</em>) – array of xyz coordinate of reference center of mass</p></li>
<li><p><strong>weights</strong> (<em>array_like</em><em> (</em><em>optional</em><em>)</em>) – choose weights. With <code class="docutils literal notranslate"><span class="pre">None</span></code> weigh each atom equally. If a float array
of the same length as <cite>mobile_coordinates</cite> is provided, use each element
of the <cite>array_like</cite> as a weight for the corresponding atom in
<cite>mobile_coordinates</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>mobile_atoms</strong> (<em>AtomGroup</em>) – AtomGroup of translated and rotated atoms</p></li>
<li><p><strong>min_rmsd</strong> (<em>float</em>) – Minimum rmsd of coordinates</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function assumes that <cite>mobile_coordinates</cite> and <cite>ref_coordinates</cite> have
already been shifted so that their centers of geometry (or centers of mass,
depending on <cite>weights</cite>) coincide at the origin. <cite>mobile_com</cite> and <cite>ref_com</cite>
are the centers <em>before</em> this shift.</p>
<ol class="arabic">
<li><p>The rotation matrix <span class="math notranslate nohighlight">\(\mathsf{R}\)</span> is determined with
<a class="reference internal" href="#MDAnalysis.analysis.align.rotation_matrix" title="MDAnalysis.analysis.align.rotation_matrix"><code class="xref py py-func docutils literal notranslate"><span class="pre">rotation_matrix()</span></code></a> directly from <cite>mobile_coordinates</cite> and
<cite>ref_coordinates</cite>.</p></li>
<li><p><cite>mobile_atoms</cite> <span class="math notranslate nohighlight">\(X\)</span> is rotated according to the
rotation matrix and the centers according to</p>
<div class="math notranslate nohighlight">
\[X' = \mathsf{R}(X - \bar{X}) + \bar{X}_{\text{ref}}\]</div>
<p>where <span class="math notranslate nohighlight">\(\bar{X}\)</span> is the center.</p>
</li>
</ol>
</dd></dl>

<dl class="function">
<dt id="MDAnalysis.analysis.align.fasta2select">
<code class="sig-prename descclassname">MDAnalysis.analysis.align.</code><code class="sig-name descname">fasta2select</code><span class="sig-paren">(</span><em class="sig-param">fastafilename</em>, <em class="sig-param">is_aligned=False</em>, <em class="sig-param">ref_resids=None</em>, <em class="sig-param">target_resids=None</em>, <em class="sig-param">ref_offset=0</em>, <em class="sig-param">target_offset=0</em>, <em class="sig-param">verbosity=3</em>, <em class="sig-param">alnfilename=None</em>, <em class="sig-param">treefilename=None</em>, <em class="sig-param">clustalw='clustalw2'</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#fasta2select"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.fasta2select" title="Permalink to this definition">¶</a></dt>
<dd><p>Return selection strings that will select equivalent residues.</p>
<p>The function aligns two sequences provided in a FASTA file and
constructs MDAnalysis selection strings of the common atoms. When
these two strings are applied to the two different proteins they
will generate AtomGroups of the aligned residues.</p>
<p><cite>fastafilename</cite> contains the two un-aligned sequences in FASTA
format. The reference is assumed to be the first sequence, the
target the second. <a class="reference external" href="http://www.clustal.org/">ClustalW</a> produces a pairwise
alignment (which is written to a file with suffix <code class="docutils literal notranslate"><span class="pre">.aln</span></code>).  The
output contains atom selection strings that select the same atoms
in the two structures.</p>
<p>Unless <cite>ref_offset</cite> and/or <cite>target_offset</cite> are specified, the resids
in the structure are assumed to correspond to the positions in the
un-aligned sequence, namely the first residue has resid == 1.</p>
<p>In more complicated cases (e.g., when the resid numbering in the
input structure has gaps due to missing parts), simply provide the
sequence of resids as they appear in the topology in <cite>ref_resids</cite> or
<cite>target_resids</cite>, e.g.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">target_resids</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span><span class="o">.</span><span class="n">resid</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">trj</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;name CA&#39;</span><span class="p">)]</span>
</pre></div>
</div>
<p>(This translation table <em>is</em> combined with any value for
<cite>ref_offset</cite> or <cite>target_offset</cite>!)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fastafilename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>, </em><em>path to filename</em>) – FASTA file with first sequence as reference and
second the one to be aligned (ORDER IS IMPORTANT!)</p></li>
<li><p><strong>is_aligned</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – <dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">False</span></code> (default)</dt><dd><p>run clustalw for sequence alignment;</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">True</span></code></dt><dd><p>use the alignment in the file (e.g. from STAMP) [<code class="docutils literal notranslate"><span class="pre">False</span></code>]</p>
</dd>
</dl>
</p></li>
<li><p><strong>ref_offset</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em> (</em><em>optional</em><em>)</em>) – add this number to the column number in the FASTA file
to get the original residue number, default: 0</p></li>
<li><p><strong>target_offset</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em> (</em><em>optional</em><em>)</em>) – add this number to the column number in the FASTA file
to get the original residue number, default: 0</p></li>
<li><p><strong>ref_resids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – sequence of resids as they appear in the reference structure</p></li>
<li><p><strong>target_resids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – sequence of resids as they appear in the target</p></li>
<li><p><strong>alnfilename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – filename of ClustalW alignment (clustal format) that is
produced by <em>clustalw</em> when <em>is_aligned</em> = <code class="docutils literal notranslate"><span class="pre">False</span></code>.
default <code class="docutils literal notranslate"><span class="pre">None</span></code> uses the name and path of <em>fastafilename</em> and
subsititutes the suffix with ‘.aln’.</p></li>
<li><p><strong>treefilename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – filename of ClustalW guide tree (Newick format);
if default <code class="docutils literal notranslate"><span class="pre">None</span></code>  the the filename is generated from <em>alnfilename</em>
with the suffix ‘.dnd’ instead of ‘.aln’</p></li>
<li><p><strong>clustalw</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> (</em><em>optional</em><em>)</em>) – path to the ClustalW (or ClustalW2) binary; only
needed for <cite>is_aligned</cite> = <code class="docutils literal notranslate"><span class="pre">False</span></code>, default: “ClustalW2”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>select_dict</strong> – dictionary with ‘reference’ and ‘mobile’ selection string
that can be used immediately in <a class="reference internal" href="#MDAnalysis.analysis.align.AlignTraj" title="MDAnalysis.analysis.align.AlignTraj"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignTraj</span></code></a> as
<code class="docutils literal notranslate"><span class="pre">select=select_dict</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)">dict</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#MDAnalysis.analysis.align.sequence_alignment" title="MDAnalysis.analysis.align.sequence_alignment"><code class="xref py py-func docutils literal notranslate"><span class="pre">sequence_alignment()</span></code></a>, <code class="xref py py-func docutils literal notranslate"><span class="pre">which()</span></code>, <code class="xref py py-func docutils literal notranslate"><span class="pre">programs.()</span></code></p>
</div>
</dd></dl>

<dl class="function">
<dt id="MDAnalysis.analysis.align.sequence_alignment">
<code class="sig-prename descclassname">MDAnalysis.analysis.align.</code><code class="sig-name descname">sequence_alignment</code><span class="sig-paren">(</span><em class="sig-param">mobile</em>, <em class="sig-param">reference</em>, <em class="sig-param">match_score=2</em>, <em class="sig-param">mismatch_penalty=-1</em>, <em class="sig-param">gap_penalty=-2</em>, <em class="sig-param">gapextension_penalty=-0.1</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#sequence_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.sequence_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a global sequence alignment between two residue groups.</p>
<p>The residues in <cite>reference</cite> and <cite>mobile</cite> will be globally aligned.
The global alignment uses the Needleman-Wunsch algorithm as
implemented in <code class="xref py py-mod docutils literal notranslate"><span class="pre">Bio.pairwise2</span></code>. The parameters of the dynamic
programming algorithm can be tuned with the keywords. The defaults
should be suitable for two similar sequences. For sequences with
low sequence identity, more specialized tools such as clustalw,
muscle, tcoffee, or similar should be used.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mobile</strong> (<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><em>AtomGroup</em></a>) – Atom group to be aligned</p></li>
<li><p><strong>reference</strong> (<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><em>AtomGroup</em></a>) – Atom group to be aligned against</p></li>
<li><p><strong>match_score</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a><em> (</em><em>optional</em><em>)</em><em>, </em><em>default 2</em>) – score for matching residues, default 2</p></li>
<li><p><strong>mismatch_penalty</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a><em> (</em><em>optional</em><em>)</em><em>, </em><em>default -1</em>) – penalty for residues that do not match , default : -1</p></li>
<li><p><strong>gap_penalty</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a><em> (</em><em>optional</em><em>)</em><em>, </em><em>default -2</em>) – penalty for opening a gap; the high default value creates compact
alignments for highly identical sequences but might not be suitable
for sequences with low identity, default : -2</p></li>
<li><p><strong>gapextension_penalty</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a><em> (</em><em>optional</em><em>)</em><em>, </em><em>default -0.1</em>) – penalty for extending a gap, default: -0.1</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>alignment</strong> – Tuple of top sequence matching output <cite>(‘Sequence A’, ‘Sequence B’, score,
begin, end)</cite></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>BioPython documentation for <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.pairwise2-module.html">pairwise2</a>. Alternatively, use
<a class="reference internal" href="#MDAnalysis.analysis.align.fasta2select" title="MDAnalysis.analysis.align.fasta2select"><code class="xref py py-func docutils literal notranslate"><span class="pre">fasta2select()</span></code></a> with <strong class="program">clustalw2</strong> and the option
<code class="docutils literal notranslate"><span class="pre">is_aligned=False</span></code>.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.10.0.</span></p>
</div>
</div>
</dd></dl>

<dl class="function">
<dt id="MDAnalysis.analysis.align.get_matching_atoms">
<code class="sig-prename descclassname">MDAnalysis.analysis.align.</code><code class="sig-name descname">get_matching_atoms</code><span class="sig-paren">(</span><em class="sig-param">ag1</em>, <em class="sig-param">ag2</em>, <em class="sig-param">tol_mass=0.1</em>, <em class="sig-param">strict=False</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/MDAnalysis/analysis/align.html#get_matching_atoms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MDAnalysis.analysis.align.get_matching_atoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Return two atom groups with one-to-one matched atoms.</p>
<p>The function takes two <a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>
instances <cite>ag1</cite> and <cite>ag2</cite> and returns two atom groups <cite>g1</cite> and <cite>g2</cite> that
consist of atoms so that the mass of atom <code class="docutils literal notranslate"><span class="pre">g1[0]</span></code> is the same as the mass
of atom <code class="docutils literal notranslate"><span class="pre">g2[0]</span></code>, <code class="docutils literal notranslate"><span class="pre">g1[1]</span></code> and <code class="docutils literal notranslate"><span class="pre">g2[1]</span></code> etc.</p>
<p>The current implementation is very simplistic and works on a per-residue basis:</p>
<ol class="arabic simple">
<li><p>The two groups must contain the same number of residues.</p></li>
<li><p>Any residues in each group that have differing number of atoms are discarded.</p></li>
<li><p>The masses of corresponding atoms are compared. and if any masses differ
by more than <cite>tol_mass</cite> the test is considered failed and a
<code class="xref py py-exc docutils literal notranslate"><span class="pre">SelectionError</span></code> is raised.</p></li>
</ol>
<p>The log file (see <a class="reference internal" href="../lib/log.html#MDAnalysis.start_logging" title="MDAnalysis.start_logging"><code class="xref py py-func docutils literal notranslate"><span class="pre">MDAnalysis.start_logging()</span></code></a>) will contain detailed
information about mismatches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ag1</strong> (<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><em>AtomGroup</em></a>) – First <a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> instance that is
compared</p></li>
<li><p><strong>ag2</strong> (<a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><em>AtomGroup</em></a>) – Second <a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a> instance that is
compared</p></li>
<li><p><strong>tol_mass</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a><em> (</em><em>optional</em><em>)</em>) – Reject if the atomic masses for matched atoms differ by more than
<cite>tol_mass</cite> [0.1]</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em> (</em><em>optional</em><em>)</em>) – <dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">True</span></code></dt><dd><p>Will raise <code class="xref py py-exc docutils literal notranslate"><span class="pre">SelectionError</span></code> if a single atom does not
match between the two selections.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">False</span></code> [default]</dt><dd><p>Will try to prepare a matching selection by dropping
residues with non-matching atoms. See <a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_matching_atoms()</span></code></a>
for details.</p>
</dd>
</dl>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>(g1, g2)</strong> – Tuple with <a class="reference internal" href="../core/groups.html#MDAnalysis.core.groups.AtomGroup" title="MDAnalysis.core.groups.AtomGroup"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomGroup</span></code></a>
instances that match, atom by atom. The groups are either the
original groups if all matched or slices of the original
groups.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference internal" href="../exceptions.html#MDAnalysis.exceptions.SelectionError" title="MDAnalysis.exceptions.SelectionError"><strong>SelectionError</strong></a> – Error raised if the number of residues does not match or if in the final
    matching masses differ by more than <em>tol</em>.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The algorithm could be improved by using e.g. the Needleman-Wunsch
algorithm in <code class="xref py py-mod docutils literal notranslate"><span class="pre">Bio.profile2</span></code> to align atoms in each residue (doing a
global alignment is too expensive).</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.8.</span></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.10.0: </span>Renamed from <code class="xref py py-func docutils literal notranslate"><span class="pre">check_same_atoms()</span></code> to
<a class="reference internal" href="#MDAnalysis.analysis.align.get_matching_atoms" title="MDAnalysis.analysis.align.get_matching_atoms"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_matching_atoms()</span></code></a> and now returns matching atomgroups
(possibly with residues removed)</p>
</div>
</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../../index.html">
    <img class="logo" src="../../_static/logos/mdanalysis-logo-200x150.png" alt="Logo"/>
    
  </a>
</p>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=MDAnalysis&repo=mdanalysis&type=star&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../overview.html">1. Overview over MDAnalysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../topology.html">2. The topology system</a></li>
<li class="toctree-l1"><a class="reference internal" href="../selections.html">3. Selection commands</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../analysis_modules.html">4. Analysis modules</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#building-blocks-for-analysis">4.1. Building blocks for Analysis</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../analysis_modules.html#distances-and-contacts">4.2. Distances and contacts</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">4.2.1. Coordinate fitting and alignment — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.align</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#rms-fitting-tutorial">4.2.1.1. RMS-fitting tutorial</a></li>
<li class="toctree-l4"><a class="reference internal" href="#common-usage">4.2.1.2. Common usage</a></li>
<li class="toctree-l4"><a class="reference internal" href="#functions-and-classes">4.2.1.3. Functions and Classes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#helper-functions">4.2.1.4. Helper functions</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="contacts.html">4.2.2. Native contacts analysis — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.contacts</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="distances.html">4.2.3. Distance analysis — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.distances</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="rms.html">4.2.4. Calculating root mean square quantities — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.rms</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="psa.html">4.2.5. Calculating path similarity — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.psa</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="encore.html">4.2.6. ENCORE Ensemble Similarity Calculations — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.encore</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#hydrogen-bonding">4.3. Hydrogen bonding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#membranes-and-membrane-proteins">4.4. Membranes and membrane proteins</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#nucleic-acids">4.5. Nucleic acids</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#polymers">4.6. Polymers</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#structure">4.7. Structure</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#volumetric-analysis">4.8. Volumetric analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#dimensionality-reduction">4.9. Dimensionality Reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#legacy-analysis-modules">4.10. Legacy analysis modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analysis_modules.html#data">4.11. Data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../topology_modules.html">5. Topology modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../coordinates_modules.html">6. Coordinates modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectory_transformations.html">7. Trajectory transformations (“on-the-fly” transformations)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../selections_modules.html">8. Selection exporters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../auxiliary_modules.html">9. Auxiliary modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core_modules.html">10. Core modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../visualization_modules.html">11. Visualization modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../lib_modules.html">12. Library functions — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.lib</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../version.html">13. Version information for MDAnalysis - <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.version</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../units.html">14. Constants and unit conversion — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.units</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../exceptions.html">15. Custom exceptions and warnings — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.exceptions</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">16. References</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../analysis_modules.html"><span class="section-number">4. </span>Analysis modules</a><ul>
      <li>Previous: <a href="base.html" title="previous chapter"><span class="section-number">4.1.1. </span>Analysis building blocks — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.base</span></code></a></li>
      <li>Next: <a href="contacts.html" title="next chapter"><span class="section-number">4.2.2. </span>Native contacts analysis — <code class="xref py py-mod docutils literal notranslate"><span class="pre">MDAnalysis.analysis.contacts</span></code></a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2005-2017, Naveen Michaud-Agrawal, Elizabeth J. Denning, Christian Beckstein (logo), Joshua L. Adelman, Shobhit Agarwal, Irfan Alibay, Luís Pedro Borges Araújo, Balasubramanian, Utkarsh Bansal, Jonathan Barnoud, Tone Bengtsen, Alejandro Bernardin, Ninad Bhat, Mateusz Bieniek, Wouter Boomsma, Jose Borreguero, Bart Bruininks, Sébastien Buchoux, Sören von Bülow, David Caplan, Matthieu Chavent, Kathleen Clark, Ruggero Cortini, Davide Cruz, Robert Delgado, John Detlefs, Xavier Deupi, Jan Domanski, David L. Dotson, Ali Ehlen, Shujie Fan, Lennard van der Feltz, Philip Fowler, Joseph Goose, Richard J. Gowers, Lukas Grossar, Abhinav Gupta, Akshay Gupta, Benjamin Hall, Eugen Hruska, Kyle J. Huston, Joe Jordan, Jon Kapla, Navya Khare, Andrew William King, Abhishek A. Kognole, Max Linke, Philip Loche, Jinju Lu, Micaela Matta, Andrew R. McCluskey, Robert McGibbon, Rocco Meli, Manuel Nuno Melo, Dominik 'Rathann' Mierzejewski, Henry Mull, Fiona B. Naughton, Alex Nesterenko, Hai Nguyen, Sang Young Noh, Daniele Padula, Nabarun Pal, Mattia F. Palermo, Danny Parton, Joshua L. Phillips, Kashish Punjani, Vedant Rathore, Tyler Reddy, Pedro Reis, Paul Rigor, Carlos Yanez S., Utkarsh Saxena, Sean L. Seyler, Paul Smith, Andy Somogyi, Caio S. Souza, Shantanu Srivastava, Lukas Stelzl, Gorman Stock, Fenil Suchak, Ayush Suhane, Xiki Tempula, Matthew W. Thompson, Matteo Tiberti, Isaac Virshup, Lily Wang, Nestor Wendt, Zhiyi Wu, Zhuyi Xue, Juan Eiros Zamora, Johannes Zeman, Yibo Zhang, and Oliver Beckstein.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../../_sources/documentation_pages/analysis/align.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/MDAnalysis/mdanalysis" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>